---
title: "P458"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
library(Signac)
library(Seurat)
library(SeuratWrappers)
library(monocle3)
library(Matrix)
library(ggplot2)
library(patchwork)
library(xlsx)
library(knitr)
library(dplyr)
library(ggplot2); theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill=NA, size=1),
          axis.text=element_text(colour="black"),
          axis.ticks=element_line(colour="black"),
          legend.key = element_blank(),
          strip.text.x = element_text(size = 14,margin = margin( b = 2, t = 2) ),
            strip.background = element_rect(fill="white", colour="black")))

library(ggthemes)
library(ggbeeswarm)
library(viridis)
library(stringr)
library(readxl)
library(kableExtra)
library(RColorBrewer)
library(plotly)
library(tidyr)
library(gtools)
#library(devtools)  #if needed to obtain github packages
# install_github('mjdufort/TCRtools') #if needed to get Matt Dufort's package
#library(TCRtools) 
#library(annotables)
#devtools::install_github("stephenturner/annotables")
#library(RNAseQC) #install_github('benaroyaresearch/RNAseQC')
#library(data.table)
#library(edgeR)
#library(ggrepel)
library(ComplexHeatmap)
#library(geneSetTools) # For barcode plots
library(egg) #For ggarrange
library(ggpubr) #Also for ggarrange
#library(DGETools) #install_github('benaroyaresearch/DGETools')
library(inlmisc) #For colors
library(umap)
#library(tcrGraph)
#library(igraph)
library(forcats)
#library(ggalluvial)
library(Seurat)
#library(scRepertoire)
library(randomcoloR)
library(apird)
library(GGally)
library(MASS)

opts_chunk$set(fig.width=6, fig.height=4.0, cache = TRUE, echo=FALSE, warning=FALSE, message=FALSE, cache.lazy = FALSE, results='hide')
opts_knit$set(root.dir = "/Users/hdeberg/Box/P458_SLICC_Mikacenic")

options(stringsAsFactors = FALSE)
set.seed(42)

```

```{r set_up_directories, cache = TRUE}

baseDir <- "~/Library/CloudStorage/Box-Box/P458_SLICC_Mikacenic"
dataDir <- file.path(baseDir, "data")
plotDir <- file.path(baseDir, "plots/")
tableDir <- file.path(baseDir, "tables")
annoDir <- file.path(baseDir, "anno")

```

```{r color_setup}
set.seed(42)

sampleColors <- big_colorblind_pal(12)

studyGroupColors <- c("ARDS" = "red",
                      "No ARDS" = "darkcyan")

studyGroupColors2 <- c("ARDS" = "red",
                      "No_ARDS" = "darkcyan")

studyGroupVisitColors <- c("v1ARDS" = "red",
                      "v1No_ARDS" = "skyblue",
                      "v2No_ARDS" = "gray",
                      "v2ARDS" = "gray")

donorColors <- RColorBrewer::brewer.pal(8,"Set1")

names(donorColors) <- c("SLK12",
                        "SLK14",
                        "SLK21",
                        "SLK22",
                        "SLK23",
                        "SLK3",
                        "SLK7",
                        "SLK8")

visitColors <- c("v1" = "orange",
                 "v2" = "blue")

```

```{r get_anno}

#Use apird function to get project data
# project <- "P458"
# 
# getGcqProjectInfo(project)
# libids <- getProjectLibs(project, searchType = "regex")
# 
# anno <- getAnno(libids)
# #Save for offline analysis
# save(anno,
#      file = file.path(annoDir,"P458Anno.Rdata"))

load(file.path(annoDir,
               "P458Anno.Rdata"))


```

#Read in data from droplets called cells and perform QC

```{r read_10x_gex_data_raw}

#Read in samples individually 
gexAbFiles <- list.files(file.path(dataDir, "count_per_sample"),
                        full.names = T,
                        pattern = "filtered_feature_bc_matrix.h5")

gexAbObjects <- lapply(gexAbFiles, Read10X_h5)

getGEXorAb <- function(gexAbList, type){
        res <- list(length(gexAbList))
        
        if(type == "gex"){
          for(i in 1:length(gexAbList)) {
            res[[i]] <- gexAbList[[i]]$`Gene Expression`
          }
        }
        
        if(type == "ab"){
          for(i in 1:length(gexAbList)) {
            res[[i]] <- gexAbList[[i]]$`Antibody Capture`
          }
        }
        return(res)
}

gexObjects <- getGEXorAb(gexAbObjects, type = "gex")

abObjects  <- getGEXorAb(gexAbObjects, type = "ab")

#Make a list of Seurat gene expression count objects
#Set parameters to define cells 
seuratObjects <- lapply(gexObjects,
                         function(x) CreateSeuratObject(counts = x,
                                                        min.features = 100,
                                                        min.cells = 3))


```

#Add feature barcode data

```{r add_fb_data_raw}

#Create a separate "FB" assay slot within the Seurat object
#This is preferable over adding the ab data as metadata because
#it allows for seurat's normalization routine to be run on it. 

for(i in 1:length(seuratObjects)){
  seurat <- seuratObjects[[i]]
  
  fb <- abObjects[[i]]
  
  keepCells <- colnames(seurat)

  fb <- fb[,keepCells]
  
  #Rename some of the rows
  tcrABIdx <- which(rownames(fb) == "TCR a/b")
  rownames(fb)[tcrABIdx] <- "TCR a_b"
  
  cd158Idx <- which(rownames(fb) == "CD158 (KIR2DL1/S1/S3/S5)")
  rownames(fb)[cd158Idx] <- "CD158 (KIR2DL1_S1_S3_S5)"
  
  cd158bIdx <- which(rownames(fb) == "CD158b (KIR2DL2/L3,  NKAT2)")
  rownames(fb)[cd158bIdx] <- "CD158b (KIR2DL2_L3,  NKAT2)"
  
  hamsterIdx <- which(rownames(fb) == "Armenian Hamster IgG Isotype Ctrl")
  rownames(fb)[hamsterIdx] <- "Hamster IgG"
  
  #Rename some of the rows
  seurat[["FB"]] <- CreateAssayObject(fb)
  
  seuratObjects[[i]] <- seurat

}

```

```{r add_qc_metrics_anno}
##Add % mito information, library name info
donors <- str_extract(gexAbFiles, "slk[0-9]+") %>% str_to_upper()
visits <- str_extract(gexAbFiles, "v[1,2]")
samples <- paste0(donors, visits)

for(i in 1:length(seuratObjects)){
  seurat <- seuratObjects[[i]]
  seurat[["percent_mt"]] <- PercentageFeatureSet(seurat, pattern = "^MT-")
  seurat[["donorID"]] <- donors[i]
  seurat[["visit"]] <- visits[i]
  seurat[["sample"]] <- paste0(donors[i],visits[i])
  seuratObjects[[i]] <- seurat
}

```

#Gene expression quality control

##Plots of the number of features detected per library

The following plots show the distribution of features (genes) detected for cells within a sample. High features may indicate a multi-plet rather than a single cell; low feature counts may indicate a low quality cell. 

```{r plot_n_features, fig.height = 3, fig.width=3}

#Select QC cutoffs

nFeatureLow <-500 
nFeatureHigh <- 4500
pctMt <- 12.5


#Look at distributions of quality features
nFeatureHistograms <- lapply(seuratObjects, function(x) hist(x$nFeature_RNA, 200, xlim= c(0,5000), main = unique(x$sample), xlab = "nFeature RNA"))

pdf(file.path(plotDir,
              "NFeatureHistograms.pdf"),
    height = 6,
    width = 8)

par(mfrow=c(4,3))
lapply(seuratObjects, function(x) {hist(x$nFeature_RNA, 50, xlim= c(0,6000), main = unique(x$sample) , xlab = "nFeature RNA")
  abline(v = nFeatureHigh, col = "red")
  abline(v = nFeatureLow, col = "red")})

invisible(dev.off())
```

```{r plot_n_features_abs, fig.height = 3, fig.width=3}

#Look at distributions of quality features
nFeatureHistograms <- lapply(seuratObjects, function(x) hist(x$nFeature_FB, 200, xlim= c(0,200), main = unique(x$sample), xlab = "nFeature FB"))

pdf(file.path(plotDir,
              "NFeatureHistogramsFB.pdf"),
    height = 6,
    width = 8)

par(mfrow=c(4,3))
lapply(seuratObjects, function(x) {hist(x$nFeature_FB, 50, xlim= c(0,200), main = unique(x$sample) , xlab = "nFeature Abs")})

invisible(dev.off())

#Make a single plot of n feature distributions

nFeatureFBList <- NULL

for(i in 1:length(seuratObjects)){
  seurat <- seuratObjects[[i]]
    
  nFeatureFBList[[i]] <- data.frame(sample = seurat$sample,
                                nFeatureFB = seurat$nFeature_FB)

}

nFeatureFB <- bind_rows(nFeatureFBList)

gNFeatureAb <- nFeatureFB %>%
  ggplot() +
  geom_density(aes(x = nFeatureFB,
                   color = sample))+
  scale_color_manual(values = sampleColors) +
  labs(x = "Number of unique antibody features",
       y = "Density (area under each curve = 1)",
       color = "Sample")

pdf(file.path(plotDir,
              "NFeatureDensityPlotsFB.pdf"),
    height = 6,
    width = 8)

print(gNFeatureAb)

invisible(dev.off())

```

##Plots of the percent mitochondrial reads per library

The following plots show the percent of features that mapped to mitochondrial genes. High percentages can indicate a poor quality or dying cell. 

```{r plot_pct_mito, fig.height = 3, fig.width=3}

pctMitoHistograms <- lapply(seuratObjects, function(x) hist(x$percent_mt, 50, xlim= c(0,20), main = unique(x$sample), xlab = "% mitochondrial reads"))

pdf(file.path(plotDir,
              "PctMitoHistograms.pdf"),
    height = 6,
    width = 8)

par(mfrow=c(4,3))
lapply(seuratObjects, function(x) {hist(x$percent_mt, 100, xlim= c(0,25),main = unique(x$sample), xlab = "% mitochondrial reads")
  abline(v = pctMt, col = "red")})

invisible(dev.off())
```

## Quality metric scatterplots with cutoffs

The plots below explore the relationship between percent mitochonridal reads and the number of features detected. Each point represents a cell. Lines on the plots indicate quality cutoffs for features (both a lower and an upper bound) and mitochrondial reads (an upper bound). 

```{r inspect_qc_cutoffs, fig.width=8, fig.height=2}

plotLayers <- list(geom_vline(xintercept = nFeatureLow),
                    geom_vline(xintercept = nFeatureHigh),
                    geom_hline(yintercept = pctMt),
                    labs(x = "Number of genes",
                         y = "% mito reads",
                         title = ""),
                    xlim(c(0, 7000)),
                    ylim(c(0, 80)),
                    theme(legend.position = 'none'))

featureScatterPlots <- lapply(seuratObjects, function(x) FeatureScatter(x, 
                          feature1 = "nFeature_RNA", 
                          feature2 = "percent_mt") +
  plotLayers +
  ggtitle(unique(x$sample)))


# ggpubr::ggarrange(plotlist = featureScatterPlots,
#                   ncol = 4,
#                   nrow = 3)

png(file.path(plotDir,
              "QCscatterplots.png"),
    height = 600,
    width = 800)

ggpubr::ggarrange(plotlist = featureScatterPlots,
                  ncol = 4,
                  nrow = 3)

invisible(dev.off())

```

##Table of quality cutoff results

The table below summarizes the number of cells per sample before and after quality control cutoffs. 

```{r make_qc_cuts, results = "show"}

seuratQC <- NULL

for(i in 1:length(seuratObjects)){
  seurat <- seuratObjects[[i]]
    
  seuratQC[[i]] <- subset(seurat, 
                      subset = nFeature_RNA > nFeatureLow & nFeature_RNA < nFeatureHigh)

}

#Tally how many cells pass QC
sampleAnno <- data.frame("sample" = sort(paste0(donors, visits)))

sampleAnno$nCells <- NA
sampleAnno$nCellsQC <- NA

for(i in 1:length(seuratObjects)){
  
  selectedSample <- unique(seuratObjects[[i]]$sample) %>%
    as.character()
    
  sampleAnno$nCells[sampleAnno$sample == selectedSample] <- dim(seuratObjects[[i]])[2]
  
  sampleAnno$nCellsQC[sampleAnno$sample == selectedSample] <- dim(seuratQC[[i]])[2]

}

sampleAnno$pctPass <- round(sampleAnno$nCellsQC/sampleAnno$nCells*100,2)

sampleAnno %>%
  dplyr::select(sample, nCells, nCellsQC, pctPass) %>%
  dplyr::rename(Sample = sample,
                `Number of cells` = nCells,
                `Number of high quality cells` = nCellsQC,
                `Percent of cells that pass QC` = pctPass) %>%
  kable(row.names = F,
        caption = "Quality analysis summary") %>%
  kable_styling(full_width = F, 
                position = "left")

```

#Merge and visualize data without integration 

```{r merge_no_integration}

seuratMerged <- merge(seuratQC[[1]], 
                       y = c(seuratQC[2:12]),
                       add.cell.ids = samples,
                       merge.data = TRUE)

 #Norm and scale FB data
  DefaultAssay(seuratMerged) <- "FB"
  
  seuratMerged <- NormalizeData(object = seuratMerged, 
                                    verbose = FALSE,
                                    normalization.method = "CLR", 
                                    margin = 2, 
                                    assay = "FB")
  
  allAbs <- rownames(seuratMerged)  
  
  controlAbs <- c("Hamster IgG",
                  "Mouse IgG1",
                  "Mouse IgG2a",
                  "Mouse IgG2b",
                  "Rat IgG1",
                  "Rat IgG2a",
                  "Rat IgG2b")
  
  citeSeqAbs <- allAbs[!(allAbs %in% controlAbs)]
  allAbs <- gtools::mixedsort(allAbs)
  controlAbs <- sort(controlAbs)
  citeSeqAbs <- sort(citeSeqAbs)

  
  VariableFeatures(seuratMerged) <- citeSeqAbs
  
  seuratMerged <- ScaleData(object = seuratMerged, verbose = FALSE)
  seuratMerged <-  RunPCA(seuratMerged, reduction.name = 'apca')
  
  seuratMerged <- RunUMAP(object = seuratMerged, 
                           reduction = "apca",
                           dims = 1:30,
                           reduction.name = "aumap")


DefaultAssay(seuratMerged) <- "RNA"

seuratMerged <- NormalizeData(object = seuratMerged, verbose = FALSE)
seuratMerged <- FindVariableFeatures(object = seuratMerged, 
        selection.method = "vst", nfeatures = 2000, verbose = FALSE)

seuratMerged <- ScaleData(object = seuratMerged, verbose = FALSE)
seuratMerged <- RunPCA(object = seuratMerged, npcs = 30, verbose = FALSE)
#ElbowPlot(seuratMerged)

seuratMerged <- FindNeighbors(seuratMerged, dims = 1:30)
seuratMerged <- FindClusters(seuratMerged, resolution = 0.5)

seuratMerged <- RunUMAP(object = seuratMerged, 
                        reduction = "pca",
                        dims = 1:30,
                        reduction.name = "umap")
```

```{r add_additional_anno}

anno$sample <- paste0(anno$donorId, "v", anno$visit)
annoGEX <- anno[anno$libraryProtocolId == "next_gem_single_cell_5_prime_v2_dual_index_10x_genomics_gex_lib_prep",] 

seuratMerged$studyGroup <- annoGEX$studyGroup[match(seuratMerged$sample,
                                                    annoGEX$sample)]


```

```{r merged_plots}

gMergedSample <- DimPlot(object = seuratMerged, 
                         reduction = "umap", 
                         group.by = "sample",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = sampleColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Merged_Sample.png"),
    height = 400, 
    width = 600)

print(gMergedSample)

invisible(dev.off())

gMergedSampleFaceted <- DimPlot(object = seuratMerged, 
                         reduction = "umap", 
                         group.by = "sample",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = sampleColors) +
  facet_wrap(~sample) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Merged_SampleFaceted.png"),
    height = 800, 
    width = 1200)

print(gMergedSampleFaceted)

invisible(dev.off())

gMergedDonor <- DimPlot(object = seuratMerged, 
                         reduction = "umap", 
                         group.by = "donorID",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = donorColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Merged_Donor.png"),
    height = 400, 
    width = 600)

print(gMergedDonor)

invisible(dev.off())

gMergedVisit <- DimPlot(object = seuratMerged, 
                         reduction = "umap", 
                         group.by = "visit",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = visitColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Merged_Visit.png"),
    height = 400, 
    width = 600)

print(gMergedVisit)

invisible(dev.off())

gMergedStudyGroup <- DimPlot(object = seuratMerged, 
                         reduction = "umap", 
                         group.by = "studyGroup",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = studyGroupColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Merged_StudyGroup.png"),
    height = 400, 
    width = 600)

print(gMergedStudyGroup)

invisible(dev.off())

gMergedClusters <- DimPlot(object = seuratMerged, 
                         reduction = "umap", 
                         group.by = "seurat_clusters",
                         label = TRUE,
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Merged_Cluster.png"),
    height = 400, 
    width = 600)

print(gMergedClusters)

invisible(dev.off())

gQC <- FeaturePlot(object = seuratMerged, 
                   reduction = "umap", 
                   features = c("percent_mt", "nFeature_RNA")) +
  labs(x = "UMAP 1", 
       y ="UMAP 2",
       color = "") +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Merged_QCVars.png"),
    height = 400,
    width = 900)

print(gQC)

invisible(dev.off())

```

```{r map_multimodal_ref_mapping}

#Multimodal reference - use to predict cell types
#Ref is PBMCs, but it might work okay for BAL cells

# data10xMultimodalReference <-
#   SeuratDisk::LoadH5Seurat(
#     file.path("/Users/hdeberg/Box/P453_TotalSeqCTitration_LinsleyLong",
#               "multimodalRefData",
#               "pbmc_multimodal.h5seurat"))
# 
# #Takes 2-3 min
# transferAnchorsData10xMultimodalReference <- FindTransferAnchors(
#   reference = data10xMultimodalReference,
#   query = seuratMerged,
#   normalization.method = "SCT",
#   reference.reduction = "spca",
#   dims = 1:50
# )
# 
# seuratMerged <- MapQuery(
#   anchorset = transferAnchorsData10xMultimodalReference,
#   query = seuratMerged,
#   reference = data10xMultimodalReference,
#   refdata = list(
#     celltype.l1 = "celltype.l1",
#     celltype.l2 = "celltype.l2",
#     celltype.l3 = "celltype.l3",
#     predicted_ADT = "ADT"
#   ),
#   reference.reduction = "spca", 
#   reduction.model = "wnn.umap"
# )
# 


```


```{r clean_up}

#Remove things from memory that are no longer needed
remove(data10xMultimodalReference)
remove(seuratObjects)
remove(seuratQC)
remove(gexObjects)
remove(abObjects)
remove(gexAbObjects)
remove(seurat)

```

```{r cell_type_prediction}

# DimPlot(
#   object = seuratMerged, 
#   reduction = "ref.umap", 
#   group.by = "predicted.celltype.l2",
#   label = TRUE, label.size = 3, repel = TRUE) +
#   NoLegend()
# options(ggrepel.max.overlaps = Inf)
# nL2States <- length(unique(seuratMerged$predicted.celltype.l2))
# set.seed(20220304)
# L2Colors <- randomcoloR::distinctColorPalette(nL2States)
# 
# nL1States <- length(unique(seuratMerged$predicted.celltype.l1))
# set.seed(20220304)
# L1Colors <- randomcoloR::distinctColorPalette(nL1States)
# 
# gL1 <- DimPlot(object = seuratMerged, 
#                reduction = "umap", 
#                group = "predicted.celltype.l1",
#                label = TRUE,
#                repel = TRUE) +
#   scale_color_manual(values = L1Colors) +
#   labs(x = "UMAP 1", 
#        y ="UMAP 2", 
#        color = "") +
#   theme(aspect.ratio=1)
# 
# png(file.path(plotDir, "UMAP_Merged_L1CellType.png"),
#     height = 400, 
#     width = 600)
# 
# print(gL1)
# 
# invisible(dev.off())
# 
# gL2 <- DimPlot(object = seuratMerged, 
#                reduction = "umap", 
#                group = "predicted.celltype.l2",
#                label = TRUE,
#                repel = TRUE) +
#   scale_color_manual(values = L2Colors) +
#   labs(x = "UMAP 1", 
#        y ="UMAP 2", 
#        color = "") +
#   theme(aspect.ratio=1)
# 
# png(file.path(plotDir, "UMAP_Merged_L2CellType.png"),
#     height = 400, 
#     width = 600)
# 
# print(gL2)
# 
# invisible(dev.off())

```


```{r cell_type_comp_by_sample}
# 
# cellTypeCounts <- seuratMerged@meta.data %>%
#   dplyr::group_by(sample, predicted.celltype.l1, studyGroup) %>%
#   dplyr::summarize(nCells = n()) %>%
#   dplyr::ungroup() %>%
#   dplyr::group_by(sample) %>%
#   dplyr::mutate(nCellsTot = sum(nCells),
#                 pctCells = nCells/nCellsTot) %>%
#   dplyr::arrange(sample)
# 
# gCellTypeCounts <- cellTypeCounts %>%
#   ggplot(aes(x = sample,
#              y = pctCells,
#              color = studyGroup,
#              fill = predicted.celltype.l1)) +
#   geom_col() +
#   scale_color_manual(values = studyGroupColors) +
#   scale_fill_manual(values = L1Colors) +
#   scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
#   labs(x = "",
#        y = "% of cells",
#        color = "Study group",
#        fill = "Cell type") +
#   theme(axis.text.x=element_text(angle=60, hjust=1))
# 
# pdf(file.path(plotDir, "Barplot_MergedBySampleCellType.pdf"),
#     height = 5.5,
#     width = 12)
# 
# print(gCellTypeCounts)
# 
# invisible(dev.off())

clusterCounts <- seuratMerged@meta.data %>%
  dplyr::group_by(sample, seurat_clusters, studyGroup) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(sample) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(sample)

gClusterCounts <- clusterCounts %>%
  ggplot(aes(x = sample,
             y = pctCells,
             color = studyGroup,
             fill = seurat_clusters)) +
  geom_col() +
  scale_color_manual(values = studyGroupColors) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "",
       y = "% of cells",
       color = "Study group",
       fill = "Cluster") +
  theme(axis.text.x=element_text(angle=60, hjust=1))

pdf(file.path(plotDir, "Barplot_MergedBySampleCluster.pdf"),
    height = 6,
    width = 12)

print(gClusterCounts)

invisible(dev.off())

```

```{r wnn_clustering}

seuratMerged <- FindMultiModalNeighbors(
  seuratMerged, reduction.list = list("pca", "apca"), 
  dims.list = list(1:30, 1:18), modality.weight.name = "RNA.weight"
)

seuratMerged <- RunUMAP(seuratMerged, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")

seuratMerged <- FindClusters(seuratMerged, graph.name = "wsnn", algorithm = 3, resolution = 2, verbose = FALSE)

```

```{r merged_plots}

gMergedSample <- DimPlot(object = seuratMerged, 
                         reduction = "wnn.umap", 
                         group.by = "sample",
                         shuffle = TRUE)+
  labs(x = "wnnUMAP 1", 
       y ="wnnUMAP 2") +
  scale_color_manual(values = sampleColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAPwnn_Merged_Sample.png"),
    height = 400, 
    width = 600)

print(gMergedSample)

invisible(dev.off())

gMergedSampleFaceted <- DimPlot(object = seuratMerged, 
                         reduction = "wnn.umap", 
                         group.by = "sample",
                         shuffle = TRUE)+
  labs(x = "wnnUMAP 1", 
       y ="wnnUMAP 2") +
  scale_color_manual(values = sampleColors) +
  facet_wrap(~sample) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAPwnn_Merged_SampleFaceted.png"),
    height = 800, 
    width = 1200)

print(gMergedSampleFaceted)

invisible(dev.off())

gMergedDonor <- DimPlot(object = seuratMerged, 
                         reduction = "wnn.umap", 
                         group.by = "donorID",
                         shuffle = TRUE)+
  labs(x = "wnnUMAP 1", 
       y ="wnnUMAP 2") +
  scale_color_manual(values = donorColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAPwnn_Merged_Donor.png"),
    height = 400, 
    width = 600)

print(gMergedDonor)

invisible(dev.off())

gMergedVisit <- DimPlot(object = seuratMerged, 
                         reduction = "wnn.umap", 
                         group.by = "visit",
                         shuffle = TRUE)+
  labs(x = "wnnUMAP 1", 
       y ="wnnUMAP 2") +
  scale_color_manual(values = visitColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAPwnn_Merged_Visit.png"),
    height = 400, 
    width = 600)

print(gMergedVisit)

invisible(dev.off())

gMergedStudyGroup <- DimPlot(object = seuratMerged, 
                         reduction = "wnn.umap", 
                         group.by = "studyGroup",
                         shuffle = TRUE)+
  labs(x = "wnnUMAP 1", 
       y ="wnnUMAP 2") +
  scale_color_manual(values = studyGroupColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAPwnn_Merged_StudyGroup.png"),
    height = 400, 
    width = 600)

print(gMergedStudyGroup)

invisible(dev.off())

# gMergedClusters <- DimPlot(object = seuratMerged, 
#                          reduction = "wnn.umap", 
#                          group.by = "seurat_clusters",
#                          label = TRUE,
#                          shuffle = TRUE)+
#   labs(x = "UMAP 1", 
#        y ="UMAP 2") +
#   theme(aspect.ratio=1)
# 
# png(file.path(plotDir, "UMAPwnn_Merged_Cluster.png"),
#     height = 400, 
#     width = 600)
# 
# print(gMergedClusters)
# 
# invisible(dev.off())

gQC <- FeaturePlot(object = seuratMerged, 
                   reduction = "wnn.umap", 
                   features = c("percent_mt", "nFeature_RNA")) +
  labs(x = "wnnUMAP 1", 
       y ="wnnUMAP 2",
       color = "") +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAPwnn_Merged_QCVars.png"),
    height = 400,
    width = 900)

print(gQC)

invisible(dev.off())

# gL1 <- DimPlot(object = seuratMerged, 
#                reduction = "wnn.umap", 
#                group = "predicted.celltype.l1",
#                label = TRUE,
#                repel = TRUE) +
#   scale_color_manual(values = L1Colors) +
#   labs(x = "wnnUMAP 1", 
#        y ="wnnUMAP 2", 
#        color = "") +
#   theme(aspect.ratio=1)
# 
# png(file.path(plotDir, "UMAPwnn_Merged_L1CellType.png"),
#     height = 400, 
#     width = 600)
# 
# print(gL1)
# 
# invisible(dev.off())
# 
# gL2 <- DimPlot(object = seuratMerged, 
#                reduction = "wnn.umap", 
#                group = "predicted.celltype.l2",
#                label = TRUE,
#                repel = TRUE) +
#   scale_color_manual(values = L2Colors) +
#   labs(x = "wnnUMAP 1", 
#        y ="wnnUMAP 2", 
#        color = "") +
#   theme(aspect.ratio=1)
# 
# png(file.path(plotDir, "UMAPwnn_Merged_L2CellType.png"),
#     height = 400, 
#     width = 600)
# 
# print(gL2)
# 
# invisible(dev.off())

```

```{r split_data}

seuratSplit <- SplitObject(seuratMerged, split.by = "sample")

# normalize and identify variable features for each dataset independently
seuratSplit <- lapply(X = seuratSplit, FUN = function(x) {
  
  x <- NormalizeData(object = x,
                     assay = "FB",
                     verbose = FALSE,
                     normalization.method = "CLR", 
                     margin = 2,)
  
  #VariableFeatures(x) <- citeSeqAbs
  
  
  x <- NormalizeData(object = x, assay = "RNA", verbose = FALSE)
  x <- FindVariableFeatures(object = x, 
                            selection.method = "vst", nfeatures = 200, verbose = FALSE)
  
})

```

```{r integrate}

features <- SelectIntegrationFeatures(object.list = seuratSplit)
seuratSplit <- lapply(X = seuratSplit, FUN = function(x) {
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
})


fastIntAnchors <- FindIntegrationAnchors(object.list = seuratSplit, 
                                         anchor.features = features,
                                         reduction = "rpca")

seuratIntegrated <- IntegrateData(anchorset = fastIntAnchors)

DefaultAssay(seuratIntegrated) <- "integrated"

```

```{r process_integrated_data}

# Run the standard workflow for visualization and clustering
seuratIntegrated <- ScaleData(seuratIntegrated, verbose = FALSE)
seuratIntegrated <- RunPCA(seuratIntegrated, npcs = 30, verbose = FALSE)
seuratIntegrated <- RunUMAP(seuratIntegrated, reduction = "pca", dims = 1:30, return.model=TRUE)
seuratIntegrated <- FindNeighbors(seuratIntegrated, reduction = "pca", dims = 1:30)
seuratIntegrated <- FindClusters(seuratIntegrated, resolution = 0.4)

save(seuratIntegrated,
     file = file.path(dataDir,"20220318_SeuratIntegratedBySample.RData"))

```

```{r integrated_plots}

gIntegratedSample <- DimPlot(object = seuratIntegrated, 
                         reduction = "umap", 
                         group.by = "sample",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = sampleColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Integrated_Sample.png"),
    height = 400, 
    width = 600)

print(gIntegratedSample)

invisible(dev.off())

gIntegratedSampleFaceted <- DimPlot(object = seuratIntegrated, 
                         reduction = "umap", 
                         group.by = "sample",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = sampleColors) +
  facet_wrap(~sample) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Integrated_SampleFaceted.png"),
    height = 800, 
    width = 1200)

print(gIntegratedSampleFaceted)

invisible(dev.off())

gIntegratedDonor <- DimPlot(object = seuratIntegrated, 
                         reduction = "umap", 
                         group.by = "donorID",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = donorColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Integrated_Donor.png"),
    height = 400, 
    width = 600)

print(gIntegratedDonor)

invisible(dev.off())

gIntegratedVisit <- DimPlot(object = seuratIntegrated, 
                         reduction = "umap", 
                         group.by = "visit",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = visitColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Integrated_Visit.png"),
    height = 400, 
    width = 600)

print(gIntegratedVisit)

invisible(dev.off())

gIntegratedStudyGroup <- DimPlot(object = seuratIntegrated, 
                         reduction = "umap", 
                         group.by = "studyGroup",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = studyGroupColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Integrated_StudyGroup.png"),
    height = 400, 
    width = 600)

print(gIntegratedStudyGroup)

invisible(dev.off())

set.seed(42)
clusterColors <- randomcoloR::distinctColorPalette(length(unique(seuratIntegrated$seurat_clusters)))
names(clusterColors) <- as.character(seq(0,length(clusterColors)-1))


gIntegratedClusters <- DimPlot(object = seuratIntegrated,
                         reduction = "umap",
                         group.by = "seurat_clusters",
                         label = TRUE,
                         shuffle = TRUE)+
  scale_color_manual(values = clusterColors) +
  labs(x = "UMAP 1",
       y ="UMAP 2") +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Integrated_Cluster.png"),
    height = 400,
    width = 600)

print(gIntegratedClusters)

invisible(dev.off())

gQC <- FeaturePlot(object = seuratIntegrated, 
                   reduction = "umap", 
                   features = c("percent_mt", "nFeature_RNA")) +
  labs(x = "UMAP 1", 
       y ="UMAP 2",
       color = "") +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_Integrated_QCVars.png"),
    height = 400,
    width = 900)

print(gQC)

invisible(dev.off())

# gL1 <- DimPlot(object = seuratIntegrated, 
#                reduction = "umap", 
#                group = "predicted.celltype.l1",
#                label = TRUE,
#                repel = TRUE) +
#   scale_color_manual(values = L1Colors) +
#   labs(x = "UMAP 1", 
#        y ="UMAP 2", 
#        color = "") +
#   theme(aspect.ratio=1)
# 
# png(file.path(plotDir, "UMAP_Integrated_L1CellType.png"),
#     height = 400, 
#     width = 600)
# 
# print(gL1)
# 
# invisible(dev.off())
# 
# gL2 <- DimPlot(object = seuratIntegrated, 
#                reduction = "umap", 
#                group = "predicted.celltype.l2",
#                label = TRUE,
#                repel = TRUE) +
#   scale_color_manual(values = L2Colors) +
#   labs(x = "UMAP 1", 
#        y ="UMAP 2", 
#        color = "") +
#   theme(aspect.ratio=1)
# 
# png(file.path(plotDir, "UMAP_Integrated_L2CellType.png"),
#     height = 400, 
#     width = 600)
# 
# print(gL2)
# 
# invisible(dev.off())

```

```{r integrated_cluster_counts}


clusterCounts <- seuratIntegrated@meta.data %>%
  dplyr::group_by(sample, seurat_clusters, studyGroup) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(sample) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(sample)

gClusterCounts <- clusterCounts %>%
  ggplot(aes(x = sample,
             y = pctCells,
             color = studyGroup,
             fill = seurat_clusters)) +
  geom_col() +
  scale_color_manual(values = studyGroupColors) +
  scale_fill_manual(values = clusterColors) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "",
       y = "% of cells",
       color = "Study group",
       fill = "Cluster") +
  theme(axis.text.x=element_text(angle=60, hjust=1))

pdf(file.path(plotDir, "Barplot_IntegratedBySampleCluster.pdf"),
    height = 6.5,
    width = 12)

print(gClusterCounts)

invisible(dev.off())

```

```{r cluster_markers}

Idents(seuratIntegrated) <- "seurat_clusters"

clusterMarkers <- FindAllMarkers(seuratIntegrated)

topMarkersLong <- clusterMarkers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n=10) %>%
  dplyr::slice_max(avg_log2FC, n = 10) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

topMarkers <- topMarkersLong %>%
  pivot_wider(id_cols = "fc_order",
              names_from = "cluster",
              values_from = "gene")

write.csv(topMarkers,
          file.path(tableDir, "TopMarkersMerged.csv"),
          row.names = F,
          quote = F)

#Get top all top markers in wide format

topMarkersWide <- clusterMarkers %>%
  dplyr::filter(!str_detect(gene, "Trav|Trbv")) %>%
  dplyr::mutate(cluster_id = paste0("Cluster ",cluster),
                cluster_id = factor(cluster_id, levels = mixedsort(unique(cluster_id)))) %>%
  dplyr::group_by(cluster_id) %>%
  dplyr::mutate(row_order = order(p_val_adj)) %>%
  pivot_wider(id_cols = row_order,
              names_from = "cluster_id",
              names_glue = "{cluster_id}_{.value}",
              values_from = c( "cluster_id",
                               "gene",
                               "avg_log2FC",
                               "p_val_adj"))
  
clusters <- unique(clusterMarkers$cluster) %>% mixedsort()
clusterLabs <- paste0("Cluster ",rep(clusters, each = 4))

columnLabs <- c("cluster_id",
                 "gene",
                 "avg_log2FC",
                "p_val_adj")

nClust <- length(unique(seuratIntegrated$seurat_clusters))
columnLabs <- rep(columnLabs, nClust)

columnLabs <- paste0(clusterLabs, "_", columnLabs)

topMarkersWide <- topMarkersWide %>%
  dplyr::select(columnLabs)

write.csv(topMarkersWide,
          file.path(tableDir, "SeuratIntegrated_TopMarkersByCluster.csv"),
          row.names = F,
          quote = F)

```

```{r marker_dotplot}

topMarkersDotPlot <- clusterMarkers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n=2) %>%
  dplyr::slice_max(avg_log2FC, n = 2) %>%
  dplyr::mutate(fc_order = order(avg_log2FC))


pdf(file.path(plotDir, "DotPlot_Integrated_Top2ClusterMarkers.pdf"),
    height = 6.5,
    width = 12)

DotPlot(seuratIntegrated, 
        features = unique(topMarkersDotPlot$gene)) + 
  RotatedAxis()

invisible(dev.off())


```

```{r integrated_hmap}

#Average top marker expr by cluster.

topMarkersHmap <- clusterMarkers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 4) %>%
  dplyr::slice_max(avg_log2FC, n = 4) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

selectedFeatures <- unique(topMarkersHmap$gene)

clusterAverages <- AverageExpression(seuratIntegrated, 
                                    assays = "RNA",
                                    features = selectedFeatures,
                                    group.by = "seurat_clusters",
                                    return.seurat = T) 

avgExprMat <- clusterAverages[["RNA"]]@data %>% as.matrix()

scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- rownames(clusterAverages@meta.data)

columnAnno <- rowAnnotation(Cluster = clusterAnno,
                              col = list(Cluster = clusterColors),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf(file.path(plotDir, 
              "Heatmap_Integrated_TopClusterMarkers.pdf"),
    height = 4,
    width = 10)

Heatmap(t(scaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        left_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())

```

```{r de_functions}
save_DE_table <- function(top_genes, out_file){
  top_genes$Gene <- rownames(top_genes)
  top_genes <- top_genes %>%
    dplyr::rename(log2FoldChange = logFC,
                  pValue = P.Value,
                  MultipleTestingCorrectedFDR = adj.P.Val) %>%
    dplyr::select(Gene, log2FoldChange, pValue, MultipleTestingCorrectedFDR)
  
  write.csv(top_genes,
            file.path(tableDir,
                      out_file),
            row.names = F,
            quote=F)
}

save_sig_table <- function(top_genes, 
                           out_file,
                           fdr_cut = 0.05,
                           lfc_cut = 1){
  
  top_genes$Gene <- rownames(top_genes)
  sig_genes <- top_genes %>%
    dplyr::rename(log2FoldChange = logFC,
                  pValue = P.Value,
                  MultipleTestingCorrectedFDR = adj.P.Val) %>%
    dplyr::filter(MultipleTestingCorrectedFDR <= fdr_cut,
                  abs(log2FoldChange) >= lfc_cut) %>%
    dplyr::select(Gene, log2FoldChange, pValue, MultipleTestingCorrectedFDR)
  
  write.csv(sig_genes,
            file.path(tableDir,
                      out_file),
            row.names = F,
            quote=F)
}

get_sig_genes <- function(top_genes,
                           fdr_cut = 0.05,
                           lfc_cut = 1){
  
  top_genes$gene <- rownames(top_genes)
  
  sig_table <- top_genes %>%
    dplyr::filter(abs(logFC) > lfc_cut,
                  adj.P.Val < fdr_cut)
  
  sig_genes <- sig_table$gene
  return(sig_genes)

}

get_top_n_genes <- function(top_genes,
                            n = 20){
  
  top_genes$gene <- rownames(top_genes)
  
  top_n_table <- top_genes %>%
    dplyr::arrange(adj.P.Val)
  
  top_n_table <- top_n_table[1:n,]
  
  top_n_genes <- top_n_table$gene
  
  return(top_n_genes)

}

add_gene_names <- function(x){
  x$gene <- rownames(x)
  return(x)
}

combine_top_tables <- function(table_list, 
                               first_table, 
                               first_table_string){
  
  first_table <- add_gene_names(first_table)
  table_list <- lapply(table_list, add_gene_names)
  
  de_table1 <- first_table %>%
    rename_with(.fn = ~paste0(first_table_string, "_",.), .cols = c("logFC", "adj.P.Val") )
  
  de_table <- safejoin::eat(de_table1, table_list,
    .by = c("gene"), .mode= "full",.check="") %>%
  dplyr::select(gene,
                ends_with("logFC"),
                ends_with("adj.P.Val"))
  
  return(de_table)
  
}

```

```{r aggregate_counts_by_cluster}
library(edgeR)
library(limma)
#selCluster <- 0

clusters <- sort(unique(seuratIntegrated$seurat_clusters))

for(selCluster in clusters){
  
  seuratCluster <- subset(seuratIntegrated, 
                          subset = seurat_clusters == selCluster)
  
  aggCounts <- AggregateExpression(seuratCluster, 
                                   assays = "RNA",
                                   slot = "counts",
                                   group.by = "sample",
                                   return.seurat = F) 
  
  aggCounts <- aggCounts$RNA
  
  aggMetadata <- data.frame("sample" = colnames(aggCounts))
  annoIdxMatch <- match(aggMetadata$sample, anno$sample)
  aggMetadata$donorID <- anno$donorId[annoIdxMatch]
  aggMetadata$visit <- anno$visit[annoIdxMatch]
  aggMetadata$studyGroup <- anno$studyGroup[annoIdxMatch]
  aggMetadata$studyGroup[aggMetadata$studyGroup == "No ARDS"] <- "No_ARDS"
  aggMetadata$studyGroup <- factor(aggMetadata$studyGroup,
                                   levels = c("No_ARDS",
                                              "ARDS")) 
  
  #Normalize for limma
  #Normalize using the TMM algorithm 
  dge <- DGEList(aggCounts)
  dge <- calcNormFactors(dge)
  normCounts <- cpm(dge, normalized.lib.sizes=TRUE)
  
  #Run PCA
  pca = prcomp(log2(as.data.frame(t(normCounts))+1), center=TRUE, scale=FALSE)
  
  pcaSummary = summary(pca)
  pcaScores= as.data.frame(pca$x)
  
  pdatscores <- merge(aggMetadata, pcaScores, by.x = "sample", by.y="row.names")
  
  pc1Lab <- paste("PC1, (", round(100*pcaSummary$importance[2, 1], 1),  "%)", sep="")
  pc2Lab <- paste("PC2, (", round(100*pcaSummary$importance[2, 2], 1),  "%)", sep="")
  
  gPCA <- ggplot(data = pdatscores) + 
    geom_point(aes(x=PC1, 
                   y=PC2, 
                   color = studyGroup,
                   shape = visit)) + 
    scale_color_manual(values = studyGroupColors,
                       name = "") +
    labs(x = pc1Lab,
         y = pc2Lab) +
    theme(aspect.ratio = 1)
  
  pcaFile <- paste0("PCA_Cluster_", selCluster, ".pdf")
  
  pdf(file.path(plotDir, pcaFile),
      height = 3,
      width = 5)
  
  print(gPCA)
  
  invisible(dev.off())
  
  #Run limma
  #Make sure counts and metadata are in the same order
  aggCounts <- aggCounts[ ,aggMetadata$sample]
  
  designMat <- model.matrix(~aggMetadata$studyGroup)
  
  #simplify columns names so contrasts are easier to make later
  colnames(designMat) <- colnames(designMat) %>%
    str_remove_all("aggMetadata|studyGroup") %>%
    str_remove_all("\\$|\\(|\\)") %>%
    str_replace_all("\\:", "_")
  
  voom <- voom(aggCounts, design= designMat, plot=T, span=0.1)
  
  corfit <-
    duplicateCorrelation(voom,
                         design=designMat,
                         block=aggMetadata$donorID)
  
  # #Run voom again. Recommended by G. Smyth
  # #https://support.bioconductor.org/p/59700/
  voom <- voomWithQualityWeights(aggCounts,
                                 design = designMat,
                                 block = aggMetadata$donorID,
                                 correlation =
                                   corfit$consensus,
                                 plot = F)
  
  # fit model, including random effect
  vfit <-
    lmFit(voom,
          block=aggMetadata$donorID,
          correlation=corfit$consensus.correlation)
  
  vfitEb <- eBayes(vfit)
  
  topGenes <- topTable(vfitEb, coef = "ARDS", sort.by = "P", number = Inf)
  
  deTableFile <- paste0("DEStats_Cluster_", selCluster, ".csv")
  sigTableFile <- paste0("SigGeneTable_Cluster_", selCluster, ".csv")
  
  save_DE_table(topGenes,
                deTableFile)
  
  save_sig_table(topGenes,
                 sigTableFile)
  
}

```

```{r split_for_dual_rna_ab_integration}

seuratSplit <- SplitObject(seuratMerged, split.by = "sample")

# normalize and identify variable features for each dataset independently
seuratSplit <- lapply(X = seuratSplit, FUN = function(x) {
  
  x <- NormalizeData(object = x, 
                     verbose = FALSE,
                     normalization.method = "CLR", 
                     margin = 2, 
                     assay = "FB")
  
  x <- NormalizeData(object = x, assay = "RNA", verbose = FALSE)
  x <- FindVariableFeatures(object = x, 
                            selection.method = "vst", nfeatures = 2000, verbose = FALSE)
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
  
})

```

```{r integrate}


integrationAnchorsBySample.tmp <-
  FindIntegrationAnchors(seuratSplit, 
                         assay = rep("RNA", length(seuratSplit)),
                         reduction = "rpca", anchor.features = features)

seuratMergedWnn <- seuratMerged
seuratMergedWnn[["RNAinteg"]] <-
  IntegrateData(integrationAnchorsBySample.tmp, new.assay.name = "RNAinteg")[["RNAinteg"]]


integrationAnchorsBySample.tmp <-
  FindIntegrationAnchors(seuratSplit, assay = rep("FB", length(seuratSplit)))

seuratMergedWnn[["ADTinteg"]] <-
  IntegrateData(integrationAnchorsBySample.tmp, new.assay.name = "ADTinteg")[["ADTinteg"]]

save(seuratMergedWnn,
     file = file.path(dataDir, "seuratMergedWnn.RData"))

```


```{r investigate_weights}

#How are FB/RNAseq data weighted


gWeights <- seuratMergedWnn@meta.data %>%
  pivot_longer(cols = c("ADTinteg.weight",
                        "RNAinteg.weight"),
               names_to = "data",
               values_to = "weight") %>%
  ggplot(aes(x = weight)) +
  geom_histogram() +
  labs(x = "Weight",
       y = "Number of cells") +
  facet_wrap(~data)

pdf(file.path(plotDir, "Histogram_DataWeights_SeuratIntegratedWnn.pdf"),
    height = 3.5,
    width = 7)

print(gWeights)

invisible(dev.off())

```

```{r wnn_on_integrated_data}


#Process integrated data for WNN
seuratMergedWnn <- ScaleData(seuratMergedWnn, 
                             assay = "RNAinteg",
                             verbose = FALSE)

seuratMergedWnn <- RunPCA(seuratMergedWnn, 
                          assay = "RNAinteg",
                          reduction.name = "pca_RNAinteg", 
                          reduction.key = "pcRNAinteg_",
                          npcs = 30, 
                          verbose = FALSE)

seuratMergedWnn <- ScaleData(seuratMergedWnn, 
                             assay = "ADTinteg",
                             features = citeSeqAbs,
                             verbose = FALSE)

seuratMergedWnn <- RunPCA(seuratMergedWnn, 
                          assay = "ADTinteg",
                          features = citeSeqAbs,
                          reduction.name = "pca_ADTinteg", 
                          reduction.key = "pcADTinteg_",
                          npcs = 30, 
                          verbose = FALSE)

dimsToUse.tmp <-
  c("RNAinteg" = 30, "ADTinteg" = 30)

seuratMergedWnn <-
  FindMultiModalNeighbors(
    seuratMergedWnn, reduction.list = list("pca_RNAinteg", "pca_ADTinteg"), 
    dims.list = list(1:dimsToUse.tmp[["RNAinteg"]], 1:dimsToUse.tmp[["ADTinteg"]]),
    modality.weight.name = c("RNAinteg.weight", "ADTinteg.weight"))

seuratMergedWnn <- RunUMAP(seuratMergedWnn, 
                           nn.name = "weighted.nn",
                           reduction.name = "umap_multimodalInteg",
                           reduction.key = "umapMultimodalInteg_")

seuratMergedWnn <-
  FindClusters(
    seuratMergedWnn, 
    graph.name = "wsnn",
    resolution = 0.2,
    verbose = FALSE)


```


```{r visualize_wnn}

gIntegratedWnnSample <- DimPlot(object = seuratMergedWnn, 
                         reduction = "umap_multimodalInteg", 
                         group.by = "sample",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = sampleColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_IntegratedWnn_Sample.png"),
    height = 400, 
    width = 600)

print(gIntegratedWnnSample)

invisible(dev.off())

gIntegratedWnnSampleFaceted <-  DimPlot(object = seuratMergedWnn, 
                         reduction = "umap_multimodalInteg", 
                         group.by = "sample",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = sampleColors) +
  facet_wrap(~sample) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_IntegratedWnn_SampleFaceted.png"),
    height = 800, 
    width = 1200)

print(gIntegratedWnnSampleFaceted)

invisible(dev.off())

gIntegratedWnnDonor <-  DimPlot(object = seuratMergedWnn, 
                         reduction = "umap_multimodalInteg", 
                         group.by = "donorID",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = donorColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_IntegratedWnn_Donor.png"),
    height = 400, 
    width = 600)

print(gIntegratedWnnDonor)

invisible(dev.off())

gIntegratedWnnVisit <-  DimPlot(object = seuratMergedWnn, 
                         reduction = "umap_multimodalInteg", 
                         group.by = "visit",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = visitColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_IntegratedWnn_Visit.png"),
    height = 400, 
    width = 600)

print(gIntegratedWnnVisit)

invisible(dev.off())

gIntegratedWnnStudyGroup <-  DimPlot(object = seuratMergedWnn, 
                         reduction = "umap_multimodalInteg", 
                         group.by = "studyGroup",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", 
       y ="UMAP 2") +
  scale_color_manual(values = studyGroupColors) +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_IntegratedWnn_StudyGroup.png"),
    height = 400, 
    width = 600)

print(gIntegratedWnnStudyGroup)

invisible(dev.off())

set.seed(42)
clusterColors <- randomcoloR::distinctColorPalette(length(unique(seuratIntegrated$seurat_clusters)))
names(clusterColors) <- as.character(seq(0,length(clusterColors)-1))


gIntegratedWnnClusters <- DimPlot(object = seuratMergedWnn,
                         reduction = "umap_multimodalInteg", 
                         group.by = "seurat_clusters",
                         label = TRUE,
                         shuffle = TRUE)+
  #scale_color_manual(values = clusterColors) +
  labs(x = "UMAP 1",
       y ="UMAP 2") +
  theme(aspect.ratio=1)

png(file.path(plotDir, "UMAP_IntegratedWnn_Cluster.png"),
    height = 400,
    width = 600)

print(gIntegratedWnnClusters)

invisible(dev.off())

# gQC <- FeaturePlot(object = seuratIntegrated, 
#                    reduction = "umap", 
#                    features = c("percent_mt", "nFeature_RNA")) +
#   labs(x = "UMAP 1", 
#        y ="UMAP 2",
#        color = "") +
#   theme(aspect.ratio=1)
# 
# png(file.path(plotDir, "UMAP_Integrated_QCVars.png"),
#     height = 400,
#     width = 900)
# 
# print(gQC)
# 
# invisible(dev.off())
```

```{r cluster_markers_wnn}

Idents(seuratMergedWnn) <- "seurat_clusters"

clusterMarkers <- FindAllMarkers(seuratMergedWnn)

topMarkersLong <- clusterMarkers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n=10) %>%
  dplyr::slice_max(avg_log2FC, n = 10) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

topMarkers <- topMarkersLong %>%
  pivot_wider(id_cols = "fc_order",
              names_from = "cluster",
              values_from = "gene")

write.csv(topMarkers,
          file.path(tableDir, "TopMarkersMergedSeuratIntegratedWnn.csv"),
          row.names = F,
          quote = F)

#Get top all top markers in wide format

topMarkersWide <- clusterMarkers %>%
  dplyr::filter(!str_detect(gene, "Trav|Trbv")) %>%
  dplyr::mutate(cluster_id = paste0("Cluster ",cluster),
                cluster_id = factor(cluster_id, levels = mixedsort(unique(cluster_id)))) %>%
  dplyr::group_by(cluster_id) %>%
  dplyr::mutate(row_order = order(p_val_adj)) %>%
  pivot_wider(id_cols = row_order,
              names_from = "cluster_id",
              names_glue = "{cluster_id}_{.value}",
              values_from = c( "cluster_id",
                               "gene",
                               "avg_log2FC",
                               "p_val_adj"))
  
clusters <- unique(clusterMarkers$cluster) %>% mixedsort()
clusterLabs <- paste0("Cluster ",rep(clusters, each = 4))

columnLabs <- c("cluster_id",
                 "gene",
                 "avg_log2FC",
                "p_val_adj")

nClust <- length(unique(seuratMergedWnn$seurat_clusters))
columnLabs <- rep(columnLabs, nClust)

columnLabs <- paste0(clusterLabs, "_", columnLabs)

topMarkersWide <- topMarkersWide %>%
  dplyr::select(columnLabs)

write.csv(topMarkersWide,
          file.path(tableDir, "SeuratIntegratedWnn_TopMarkersByCluster.csv"),
          row.names = F,
          quote = F)

```

```{r marker_dotplot}

topMarkersDotPlot <- ClusterMarkers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n=2) %>%
  dplyr::slice_max(avg_log2FC, n = 2) %>%
  dplyr::mutate(fc_order = order(avg_log2FC))


pdf(file.path(plotDir, "DotPlot_IntegratedWnn_Top2ClusterMarkers.pdf"),
    height = 6.5,
    width = 12)

DotPlot(MonocytesIntegrated, 
        features = unique(topMarkersDotPlot$gene)) + 
  RotatedAxis()

invisible(dev.off())


```

```{r integrated_hmap}

#Average top marker expr by cluster.

topMarkersHmap <- ClusterMarkers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 4) %>%
  dplyr::slice_max(avg_log2FC, n = 4) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

selectedFeatures <- unique(topMarkersHmap$gene)

clusterAverages <- AverageExpression(MonocytesIntegratedviaCluster, 
                                    features = selectedFeatures,
                                    assays = 'RNA',
                                    group.by = "seurat_clusters",
                                    use.scale = TRUE,
                                    return.seurat = T) 

avgExprMat <- clusterAverages[["RNA"]]@data %>% as.matrix()

scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- rownames(clusterAverages@meta.data)

columnAnno <- rowAnnotation(Cluster = clusterAnno,
                              col = list(Cluster = clusterColors),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf(file.path(plotDir, 
              "Heatmap_Integrated_TopClusterMarkers.pdf"),
    height = 4,
    width = 10)

Heatmap(t(scaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        left_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())

```

```{r supervised analyses}
#set row name identities to the comparison group of interest
Idents(seuratInegrated) <- predicted.celltype.l1
MonocytesIntegrated <- subset(seuratIntegrated, idents = "Mono")

Idents(MonocytesIntegrated) <- 'studyGroup'

#determine the markers of interest between the two groups, for both RNA and ADT -- FB will require a reduced log-fold threshold.
ARDSmarkersRNA <- FindMarkers(MonocytesIntegrated,
                           assay = 'RNA',
                           ident.1 = 'ARDS',
                           ident.2 = 'No ARDS')

ARDSmarkersFB <- FindMarkers(MonocytesIntegrated,
                             assay = 'FB',
                             ident.1 = 'ARDS',
                             ident.2 = 'No ARDS',
                             logfc.threshold = 0.1)

#run PCA for RNA assay with selected markers
DefaultAssay(MonocytesIntegrated) <- 'RNA'
MonocytesIntegrated <- NormalizeData(MonocytesIntegrated) %>% ScaleData() %>% RunPCA(features = rownames(ARDSmarkersRNA),
                                                                                     assay = "RNA",
                                                                                     reduction.name = "RNABestPCA")
#run PCA for FB ssay with selected markers
DefaultAssay(MonocytesIntegrated) <- 'FB'
MonocytesIntegrated <- NormalizeData(MonocytesIntegrated,
                                     normalization.method = 'CLR',
                                     margin = 2) %>% ScaleData() %>% RunPCA(features = rownames(ARDSmarkersFB),
                                                                                     assay = "FB",
                                                                                     reduction.name = "FBBestPCA")

#FindMultiModal Neighbors for WNN

MonocytesIntegrated <- FindMultiModalNeighbors(MonocytesIntegrated,
                                               reduction.list = list("FBBestPCA", "RNABestPCA"),
                                               dims.list = list(1:10, 1:20),
                                               modality.weight.name = list("BestIntegratedFB.weight", "BestIntegratedRNA.weight"))

#run UMAP
MonocytesIntegrated <- RunUMAP(MonocytesIntegrated, 
                               nn.name = "weighted.nn",
                               reduction.name = "wnn.umap", 
                               reduction.key = "wnnUMAP_")

#identify clusters
MonocytesIntegrated <- FindClusters(MonocytesIntegrated,
                                    graph.name = "wsnn",
                                    algorithm = 3,
                                    resolution = .2,
                                    verbose = FALSE)
```

```{r visualization}
#Visualize WNN supervised analyses

BestClusterPlot <- DimPlot(MonocytesIntegrated,
                             reduction = 'wnn.umap',
                             label = TRUE,
                             repel = TRUE,
                             label.size = 2.5) + NoLegend() 

BestSamplePlot <- DimPlot(MonocytesIntegrated,
                          reduction = 'wnn.umap',
                          label = TRUE, repel = TRUE,
                          label.size = 2.5, group.by = "sample") + NoLegend()

BestConditionPlot <- DimPlot(MonocytesIntegrated,
                             reduction = 'wnn.umap',
                             label = TRUE, repel = TRUE,
                             label.size = 2.5,
                             group.by = "studyGroup") + NoLegend() 

LargePlot <- BestConditionPlot + BestClusterPlot + BestSamplePlot
```

```{r markers}

#Find Marks between all clusters of supervised test

ClusterMarkers <- FindMarkers(MonocytesIntegratedviaCluster)

#Extract putative gene names and synonymize gene names; future need to use Mygene API to look up aliases -- even after this converstion, more than half of the genes didn't map to ENTREZ and had to be manually adjusted to their actual gene names
enrichmentSet <- ClusterMarkers[,7] %>% unique() %>% gsub("\\.1", "") %>% gsub(" \\(.*", "")

# Convert gene names to ENTREZ ID
enrichmentSet <- bitr(enrichmentSet, fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = org.Hs.eg.db )

#Perform gene ontology term analysis for CC, BP, and MF terms
egoMF <- enrichGO(gene = enrichmentSet,
                  OrgDB = org.Hs.eg,
                  ont = "MF",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)

egoBP <- enrichGO(gene = enrichmentSet,
                  OrgDB = org.Hs.eg,
                  ont = "BP",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)

egoCC <- enrichGO(gene = enrichmentSet,
                  OrgDB = org.Hs.eg,
                  ont = "CC",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)
```

```{r supervised resubset}
#subset based on cluster identity and rerun the supervised analyses
#MonocytesIntegratedviaCluster <- subset(seuratIntegrated, 
                                      #  idents = c("9","10","13","14", "15"), 
                                       # invert = TRUE)

#set row name identities to the comparison group of interest
Idents(MonocytesIntegratedviaCluster) <- 'studyGroup'

#determine the markers of interest between the two groups, for both RNA and ADT -- FB will require a reduced log-fold threshold.
ARDSmarkersRNA2 <- FindMarkers(MonocytesIntegratedviaCluster,
                           assay = 'RNA',
                           ident.1 = 'ARDS',
                           ident.2 = 'No ARDS')

ARDSmarkersFB2 <- FindMarkers(MonocytesIntegratedviaCluster,
                             assay = 'FB',
                             ident.1 = 'ARDS',
                             ident.2 = 'No ARDS',
                             logfc.threshold = 0.1)

#run PCA for RNA assay with selected markers
DefaultAssay(MonocytesIntegratedviaCluster) <- 'RNA'
MonocytesIntegratedviaCluster <- NormalizeData(MonocytesIntegratedviaCluster) %>% ScaleData() %>% RunPCA(features = rownames(ARDSmarkersRNA2),
                                                                                     assay = "RNA",
                                                                                     reduction.name = "RNABestPCA")
#run PCA for FB ssay with selected markers
DefaultAssay(MonocytesIntegratedviaCluster) <- 'FB'
MonocytesIntegratedviaCluster <- NormalizeData(MonocytesIntegratedviaCluster,
                                     normalization.method = 'CLR',
                                     margin = 2) %>% ScaleData() %>% RunPCA(features = rownames(ARDSmarkersFB2),
                                                                                     assay = "FB",
                                                                                     reduction.name = "FBBestPCA")
#FindMultiModal Neighbors for WNN

MonocytesIntegratedviaCluster <- FindMultiModalNeighbors(MonocytesIntegratedviaCluster,
                                               reduction.list = list("FBBestPCA", "RNABestPCA"),
                                               dims.list = list(1:10, 1:20),
                                               modality.weight.name = list("BestIntegratedFB.weight", "BestIntegratedRNA.weight"))

#run UMAP
MonocytesIntegratedviaCluster <- RunUMAP(MonocytesIntegratedviaCluster, 
                               nn.name = "weighted.nn",
                               reduction.name = "umap", 
                               reduction.key = "UMAP_")

#identify clusters
MonocytesIntegratedviaCluster <- FindClusters(MonocytesIntegratedviaCluster,
                                    graph.name = "wsnn",
                                    algorithm = 3,
                                    resolution = .2,
                                    verbose = FALSE)
```

```{r markers}

#Find Marks between all clusters of supervised test

ClusterMarkers <- FindAllMarkers(MonocytesIntegratedviaCluster)

topMarkersLongMono <- ClusterMarkers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n=10) %>%
  dplyr::slice_max(avg_log2FC, n = 10) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

topMarkersMono <- topMarkersLongMono %>%
  pivot_wider(id_cols = "fc_order",
              names_from = "cluster",
              values_from = "gene")

write.csv(topMarkersMono,
          file.path(tableDir, "TopMarkersMonoSubset.csv"),
          row.names = F,
          quote = F)

#Extract putative gene names and synonymize gene names; future need to use Mygene API to look up aliases -- even after this converstion, more than half of the genes didn't map to ENTREZ and had to be manually adjusted to their actual gene names
enrichmentSet <- ClusterMarkers[,7] %>% unique() 
enrichmentSet <- gsub("\\.1", "", enrichmentSet)
enrichmentSet <- gsub(" \\(.*", "", enrichmentSet)
enrichmentSet <- alias2Symbol(enrichmentSet, species = "Hs")

# Convert gene names to ENTREZ ID
enrichmentSet <- bitr(enrichmentSet, fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = org.Hs.eg.db )
enrichmentSet <- enrichmentSet[,2]
#Perform gene ontology term analysis for CC, BP, and MF terms
egoMF <- enrichGO(gene = enrichmentSet,
                  OrgDb = org.Hs.eg.db,
                  ont = "MF",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)

egoBP <- enrichGO(gene = enrichmentSet,
                  OrgDb = org.Hs.eg.db,
                  ont = "BP",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)

egoCC <- enrichGO(gene = enrichmentSet,
                  OrgDb = org.Hs.eg.db,
                  ont = "CC",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)
```

```{r visualization}
#Visualize WNN supervised analyses

set.seed(42)
clusterColors <- randomcoloR::distinctColorPalette(length(unique(MonocytesIntegratedviaCluster$seurat_clusters)))
names(clusterColors) <- as.character(seq(0,length(clusterColors)-1))

 clustersMonocyteSupervised <- DimPlot(MonocytesIntegratedviaCluster,
         reduction = 'wnn.umap',
         label = TRUE,
         repel = TRUE,
         label.size = 3.5) +
         NoLegend() +
         labs(x = "wnnUMAP 1",
              y ="wnnUMAP 2") +
         scale_color_manual(values = clusterColors) +
         theme(aspect.ratio=1)

png(file.path(plotDir, "SupervisedWNN.png"),
    height = 400, 
    width = 600)

print(clustersMonocyteSupervised)

invisible(dev.off()) 

visitMonocyteSupervised <- DimPlot(MonocytesIntegratedviaCluster,
         reduction = 'wnn.umap',
         repel = TRUE,
         cols = c("SLK12v1" = "blue",
                                                      "SLK12v2" = "orange",
                                                      "SLK21v1" = "blue",
                                                      "SLK21v2" = "orange",
                                                      "SLK23v1" = "blue",
                                                      "SLK23v2" = "orange",
                                                      "SLK7v1" = "blue",
                                                      "SLK7v2" = "orange"),
         group.by = "sample") +
         labs(x = "wnnUMAP 1",
              y ="wnnUMAP 2") +
         theme(aspect.ratio=1)

png(file.path(plotDir, "visitWNNsample.png"),
    height = 400, 
    width = 600)

print(visitMonocyteSupervised)

invisible(dev.off()) 

conditionMonocyteSupervised <- DimPlot(MonocytesIntegratedviaCluster,
         reduction = 'RNABestPCA',
         repel = TRUE,
         group.by = "studyGroup") +
         labs(x = "UMAP 1",
              y ="UMAP 2") +
         scale_color_manual(values = studyGroupColors) +
         theme(aspect.ratio=1)

png(file.path(plotDir, "SupervisedWNNcondition.png"),
    height = 400, 
    width = 600)

print(conditionMonocyteSupervised)

invisible(dev.off()) 
```

```{r heatmap of expression in monocytes}
topMarkersHmap <- ClusterMarkers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 4) %>%
  dplyr::slice_max(avg_log2FC, n = 4) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

selectedFeatures <- unique(topMarkersHmap$gene)

clusterAverages <- AverageExpression(MonocytesIntegratedviaCluster, 
                                    features = selectedFeatures,
                                    assays = 'RNA',
                                    group.by = "seurat_clusters",
                                    use.scale = TRUE,
                                    return.seurat = T) 

avgExprMat <- clusterAverages[["RNA"]]@data %>% as.matrix()

scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- rownames(clusterAverages@meta.data)

columnAnno <- rowAnnotation(Cluster = clusterAnno,
                              col = list(Cluster = clusterColors),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf(file.path(plotDir, 
              "Heatmap_Integrated_TopClusterMarkersMonocytes.pdf"),
    height = 4,
    width = 4)

Heatmap(t(scaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        left_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())

```

```{r find markers between visits}
Idents(MonocytesIntegratedviaCluster) <- "sample"

seuratVisit <- subset(MonocytesIntegratedviaCluster, idents = c("SLK8v1","SLK3v1","SLK22v1","SLK14v1"), invert = TRUE) 

Idents(seuratVisit) <- "visit"
DefaultAssay(seuratVisit) <- "integrated"

visitMarkers <- FindMarkers(seuratVisit, logfc.threshold = 0.20, ident.1 = "v1", ident.2 = "v2")

visitSet <- rownames(visitMarkers) %>% unique() 
visitSet <- gsub("\\.1", "", visitSet)
visitSet <- gsub(" \\(.*", "", visitSet)
visitSet <- alias2Symbol(visitSet, species = "Hs")

visitSet <- bitr(visitSet, fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = org.Hs.eg.db )
visitSet <- visitSet[,2]

visitMF <- enrichGO(gene = visitSet,
                  OrgDb = org.Hs.eg.db,
                  ont = "MF",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)

visitBP <- enrichGO(gene = visitSet,
                  OrgDb = org.Hs.eg.db,
                  ont = "BP",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)

visitCC <- enrichGO(gene = visitSet,
                  OrgDb = org.Hs.eg.db,
                  ont = "CC",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)

write.csv(visitBP,
          file.path(tableDir, "visitGOBP.csv"),
          row.names = F,
          quote = F)
```

```{r ardsGO}
ardsSet <- rownames(ARDSmarkersRNA3) %>% unique() 
ardsSet <- gsub("\\.1", "", ardsSet)
ardsSet <- gsub(" \\(.*", "", ardsSet)
ardsSet <- alias2Symbol(ardsSet, species = "Hs")

ardsSet <- bitr(ardsSet, fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = org.Hs.eg.db )
ardsSet <- ardsSet[,2]

ardsMF <- enrichGO(gene = ardsSet,
                  OrgDb = org.Hs.eg.db,
                  ont = "MF",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)

ardsBP <- enrichGO(gene = ardsSet,
                  OrgDb = org.Hs.eg.db,
                  ont = "BP",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)

ardsCC <- enrichGO(gene = ardsSet,
                  OrgDb = org.Hs.eg.db,
                  ont = "CC",
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  qvalueCutoff = 0.05,
                  readable = TRUE)

write.csv(ardsBP,
          file.path(tableDir, "ardsGOBP.csv"),
          row.names = F,
          quote = F)
```

```{r heatmap between visits}

visitMarkersHmap <- visitMarkers %>%
  #dplyr::group_by(sample) %>%
  #dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 200) %>%
  dplyr::slice_max(avg_log2FC, n = 30) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

visitFeatures <- unique(rownames(visitMarkersHmap))



visitAverages <- AverageExpression(seuratVisit, 
                                    features = visitFeatures,
                                    assays = 'RNA',
                                    group.by = "sample",
                                    use.scale = TRUE,
                                    return.seurat = T) 

visitAvgExprMat <- visitAverages[["RNA"]]@data %>% as.matrix()

visitScaledExprMat <- t(scale(t(visitAvgExprMat)))


visitClusterAnno <- rownames(visitAverages@meta.data)

visitColumnAnno <- rowAnnotation(samples = visitClusterAnno,
                              col =  list(samples = c("SLK12v1" = "blue",
                                                      "SLK12v2" = "orange",
                                                      "SLK21v1" = "blue",
                                                      "SLK21v2" = "orange",
                                                      "SLK23v1" = "blue",
                                                      "SLK23v2" = "orange",
                                                      "SLK7v1" = "blue",
                                                      "SLK7v2" = "orange")),
                              simple_anno_size = unit(0.25, "cm"), 
                              show_legend = FALSE,
                              show_annotation_name = FALSE)

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf(file.path(plotDir, 
              "Heatmap_Integrated_visitShift_Extended_Narrow.pdf"),
    height = 6,
    width = 4)

Heatmap(t(visitScaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = FALSE,
        row_order = c("SLK12v2", "SLK7v2", "SLK23v2", "SLK21v2", "SLK21v1", "SLK12v1", "SLK23v1", "SLK7v1"),
        show_row_dend = TRUE,
        left_annotation = visitColumnAnno,
        height = nrow(t(visitScaledExprMat))*unit(5, "mm"),
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())
```

```{r reinforce cluster variation}
DefaultAssay(MonocytesIntegratedviaCluster) <- 'RNA'
MonocytesIntegratedviaCluster <- NormalizeData(MonocytesIntegratedviaCluster) %>% ScaleData() %>% RunPCA(features = selectedFeatures,
                                                                                     assay = "RNA",
                                                                                     reduction.name = "RNAreinforce")
MonocytesIntegratedviaCluster<- RunUMAP(object = MonocytesIntegratedviaCluster, 
                           reduction = "RNAreinforce",
                           dims = 1:30,
                           reduction.name = "reinforceumap")


DefaultAssay(MonocytesIntegratedviaCluster) <- "RNA"


MonocytesIntegratedviaCluster <- FindNeighbors(MonocytesIntegratedviaCluster, dims = 1:30)
MonocytesIntegratedviaCluster <- FindClusters(MonocytesIntegratedviaCluster, resolution = 0.5)

MonocytesIntegratedviaCluster <- RunUMAP(object = MonocytesIntegratedviaCluster, 
                        reduction = "RNAreinforce",
                        dims = 1:30,
                        reduction.name = "umap")
```
```{r heatmap of ARDS markers}
ardsMarkersHmap <- ARDSmarkersRNA2 %>%
  #dplyr::group_by("sample") %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 30) %>%
  dplyr::slice_max(avg_log2FC, n = 30) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

ardsFeatures <- rownames(ardsMarkersHmap)



ardsAverages <- AverageExpression(MonocytesIntegratedviaCluster, 
                                    features = ardsFeatures,
                                    assays = 'RNA',
                                    group.by = "sample",
                                    use.scale = TRUE,
                                    return.seurat = T) 

ardsAvgExprMat <- ardsAverages[["RNA"]]@data %>% as.matrix()

ardsScaledExprMat <- t(scale(t(ardsAvgExprMat)))


ardsClusterAnno <- rownames(ardsAverages@meta.data)

ardsColumnAnno <- rowAnnotation(sample = ardsClusterAnno,
                              col =  list(sample = c("SLK8v1" = "darkcyan",
                                                     "SLK7v2" = "red",
                                                     "SLK7v1" = "red",
                                                     "SLK12v2" = "darkcyan",
                                                     "SLK12v1" = "darkcyan",
                                                     "SLK14v1" = "darkcyan",
                                                     "SLK21v2" = "darkcyan",
                                                     "SLK21v1" = "darkcyan",
                                                     "SLK22v1" = "red",
                                                     "SLK23v2" = "red",
                                                     "SLK23v1" = "red",
                                                     "SLK3v1" = "red")),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf(file.path(plotDir, 
              "Heatmap_Integrated_ARDS_visit1_Narrow.pdf"),
    height = 4,
    width = 4)

        Heatmap(t(ardsScaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        left_annotation = ardsColumnAnno,
        height = nrow(t(ardsScaledExprMat))*unit(5, "mm"),
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())
```

```{r monocle3}

SLK12 <- MonocytesIntegratedviaCluster[,  MonocytesIntegratedviaCluster$sample %in% c("SLK12v1", "SLK12v2")]
SLK21 <- MonocytesIntegratedviaCluster[,  MonocytesIntegratedviaCluster$sample %in% c("SLK21v1", "SLK21v2")]
SLK7 <- MonocytesIntegratedviaCluster[,  MonocytesIntegratedviaCluster$sample %in% c("SLK7v1", "SLK7v2")]
SLK23 <- MonocytesIntegratedviaCluster[,  MonocytesIntegratedviaCluster$sample %in% c("SLK23v1", "SLK23v2")]
SLK12.cds <- as.cell_data_set(SLK12)
SLK21.cds <- as.cell_data_set(SLK21)
SLK7.cds <- as.cell_data_set(SLK7)
SLK23.cds <- as.cell_data_set(SLK23)

SLK12.cds <- cluster_cells(cds = SLK12.cds, reduction_method = "UMAP")
SLK12.cds <- learn_graph(SLK12.cds, use_partition = TRUE)
SLK12.cds <- order_cells(SLK12.cds, reduction_method = "UMAP")

SLK21.cds <- cluster_cells(cds = SLK21.cds, reduction_method = "UMAP")
SLK21.cds <- learn_graph(SLK21.cds, use_partition = TRUE)
SLK21.cds <- order_cells(SLK21.cds, reduction_method = "UMAP")

SLK23.cds <- cluster_cells(cds = SLK23.cds, reduction_method = "UMAP")
SLK23.cds <- learn_graph(SLK23.cds, use_partition = TRUE)
SLK23.cds <- order_cells(SLK23.cds, reduction_method = "UMAP")

SLK7.cds <- cluster_cells(cds = SLK7.cds, reduction_method = "UMAP")
SLK7.cds <- learn_graph(SLK7.cds, use_partition = TRUE)
SLK7.cds <- order_cells(SLK7.cds, reduction_method = "UMAP")

MonocytesIntegratedviaCluster <- AddMetaData(
  object = MonocytesIntegratedviaCluster,
  metadata = SLK12.cds@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "SLK12"
)
MonocytesIntegratedviaCluster <- AddMetaData(
  object = MonocytesIntegratedviaCluster,
  metadata = SLK21.cds@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "SLK21"
)
MonocytesIntegratedviaCluster <- AddMetaData(
  object = MonocytesIntegratedviaCluster,
  metadata = SLK23.cds@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "SLK23"
)
MonocytesIntegratedviaCluster <- AddMetaData(
  object = MonocytesIntegratedviaCluster,
  metadata = SLK7.cds@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "SLK7"
)

FeaturePlot(MonocytesIntegratedviaCluster, c("SLK12", "SLK21", "SLK23", "SLK7"), pt.size = 0.1) & scale_color_viridis_c()

pdf(file.path(plotDir, 
              "PseudotimeVisits.pdf"),
    height = 4,
    width = 4)

FeaturePlot(MonocytesIntegratedviaCluster, c("SLK12", "SLK21", "SLK23", "SLK7"), pt.size = 0.1) & scale_color_viridis_c()

dev.off()

png(file.path(plotDir, 
              "PseudotimeVisits.png"),
    height = 8,
    width = 8)

FeaturePlot(MonocytesIntegratedviaCluster, c("SLK12", "SLK21", "SLK23", "SLK7"), pt.size = 0.1) & scale_color_viridis_c()

dev.off()

cds <- as.cell_data_set(MonocytesIntegratedviaCluster)
cds <- cluster_cells(cds)
p1 <- plot_cells(cds, show_trajectory_graph = FALSE)
p2 <- plot_cells(cds, color_cells_by = "partition", show_trajectory_graph = FALSE)
wrap_plots(p1, p2)

```

```{r supervised analyses for visit driven differences (time orientation)}
#set row name identities to the comparison group of interest
Idents(seuratVisit) <- "visit"
visitMarkersRNA <- FindMarkers(seuratVisit, logfc.threshold = 0.25, assay = "RNA", ident.1 = "v1", ident.2 = "v2")
visitMarkersFB <- FindMarkers(seuratVisit, logfc.threshold = 0.15, assay = "FB", ident.1 = "v1", ident.2 = "v2")

Idents(MonocytesIntegratedviaCluster) <- 'sample'

#run PCA for integrated assay with selected markers
DefaultAssay(MonocytesIntegratedviaCluster) <- 'RNA'
MonocytesIntegratedviaCluster <- NormalizeData(MonocytesIntegratedviaCluster) %>% ScaleData() %>% RunPCA(
                                        features = rownames(visitMarkersRNA),
                                        assay = "RNA",
                                        reduction.name = "visitRNAPCA")

DefaultAssay(MonocytesIntegratedviaCluster) <- 'FB'
MonocytesIntegratedviaCluster <- NormalizeData(MonocytesIntegratedviaCluster) %>% ScaleData() %>% RunPCA(
                                        features = rownames(visitMarkersFB),
                                        assay = "FB",
                                        reduction.name = "visitFBPCA")

#FindMultiModal Neighbors for WNN

MonocytesIntegratedviaCluster <- FindMultiModalNeighbors(MonocytesIntegratedviaCluster,
                                               reduction.list = list("visitRNAPCA","visitFBPCA"),
                                               dims.list = list(1:20, 1:5),
                                               modality.weight.name = "visit.weight",
                                               knn.graph.name = "wknnvisit",
                                               snn.graph.name = "wsnnvisit",
                                               weighted.nn.name = "weightedvisit.nn"
                                               )

#run UMAP
MonocytesIntegratedviaCluster <- RunUMAP(MonocytesIntegratedviaCluster, 
                               nn.name = "weightedvisit.nn",
                               reduction.name = "umap_visit", 
                               reduction.key = "UMAP_visit_")

#identify clusters
MonocytesIntegratedviaCluster <- FindClusters(MonocytesIntegratedviaCluster,
                                    graph.name = "wsnnvisit",
                                    algorithm = 3,
                                    resolution = .2,
                                    verbose = FALSE)
 visitSampleColored <- DimPlot(MonocytesIntegratedviaCluster,
         reduction = 'umap',
         repel = TRUE,
         cols = c("SLK12v1" = "blue",
                                                      "SLK12v2" = "orange",
                                                      "SLK21v1" = "blue",
                                                      "SLK21v2" = "orange",
                                                      "SLK23v1" = "blue",
                                                      "SLK23v2" = "orange",
                                                      "SLK7v1" = "blue",
                                                      "SLK7v2" = "orange",
                  "SLK8v1" = "gray",
                  "SLK3v1" = "gray",
                  "SLK14v1" = "gray",
                  "SLK22v1" = "gray"),
         group.by = "sample") +
         labs(x = "UMAP 1",
              y ="UMAP 2") +
         theme(aspect.ratio=1) + labs(title = "Visit Timing") + NoLegend()

pdf(file.path(plotDir, 
              "VisitUMAPvsSample.pdf"),
    height = 4,
    width = 6)
 visitSampleColored + sampleOnlyUMAP
dev.off()

S100A12Umap <- FeaturePlot(MonocytesIntegratedviaCluster, feature = "S100A12", reduction = "umap")
S100A8Umap <- FeaturePlot(MonocytesIntegratedviaCluster, feature = "S100A8" , reduction = "umap", min.cutoff = 2)
CCR2Umap <- FeaturePlot(MonocytesIntegratedviaCluster, feature = "CCR2" , reduction = "umap", max.cutoff = 2)

CXCL8Umap <- FeaturePlot(MonocytesIntegratedviaCluster, feature = "CXCL8" , reduction = "umap", min.cutoff = 2.5)

CCL2Umap <- FeaturePlot(MonocytesIntegratedviaCluster, feature = "CCL2" , reduction = "umap", min.cutoff = 1.5)
SPP1Umap <- FeaturePlot(MonocytesIntegratedviaCluster, feature = "SPP1" , reduction = "umap", min.cutoff = 1, max.cutoff = 4)

CCL18Umap <- FeaturePlot(MonocytesIntegratedviaCluster, feature = "CCL18" , reduction = "umap", min.cutoff = 2, max.cutoff = 8)
CD81Umap <- FeaturePlot(MonocytesIntegratedviaCluster, feature = "CD81" , reduction = "umap", min.cutoff = 2, max.cutoff = 4)
CD9Umap <- FeaturePlot(MonocytesIntegratedviaCluster, feature = "CD9" , reduction = "umap", min.cutoff = 1, max.cutoff = 6)
TREM2Umap <- FeaturePlot(MonocytesIntegratedviaCluster, feature = "TREM2" , reduction = "umap", min.cutoff = 1, max.cutoff = 6)
MERTKUmap <- FeaturePlot(MonocytesIntegratedviaCluster, "MERTK", reduction = "umap", min.cutoff = .2, max.cutoff = 1.5)


pdf(file.path(plotDir, 
               "Fibrosis.pdf"),
     height = 4,
     width = 6)
 
 ggarrange( plotlist = list(MERTKUmap, SPP1Umap))
 dev.off()

pdf(file.path(plotDir, 
               "earlyGenesForPsuedo.pdf"),
     height = 4,
     width = 6)
 
 ggarrange( plotlist = list(S100A12Umap, S100A8Umap, CCR2Umap, CXCL8Umap))
 dev.off()

 
 pdf(file.path(plotDir, 
               "midGenesForPsuedo.pdf"),
     height = 4,
     width = 6)
 
 ggarrange(CCL2Umap + SPP1Umap, ncol = 2, nrow = 1)
 dev.off()
 
 
 pdf(file.path(plotDir, 
               "lateGenesForPsuedo.pdf"),
     height = 4,
     width = 6)
 
 ggarrange(CCL18Umap,CD81Umap, CD9Umap, TREM2Umap)
 dev.off()
 
  pdf(file.path(plotDir, 
               "ClustersVisit.pdf"),
     height = 4,
     width = 4)
 
 DimPlot(MonocytesIntegratedviaCluster, group.by = "seurat_clusters", reduction = "umap")
 
 dev.off()
 

 cds <- as.cell_data_set(MonocytesIntegratedviaCluster)
cds <- cluster_cells(cds)
p1 <- plot_cells(cds, show_trajectory_graph = FALSE)
p2 <- plot_cells(cds, color_cells_by = "partition", show_trajectory_graph = FALSE)
wrap_plots(p1, p2)
cds <- learn_graph(cds)
plot_cells(cds, label_groups_by_cluster = FALSE, label_leaves = FALSE, label_branch_points = FALSE)

cds <- order_cells(cds, reduction_method = "UMAP")

  pdf(file.path(plotDir, 
               "PseudoTimeVisit.pdf"),
     height = 4,
     width = 8)
plot_cells(
  cds = MonocyteMonocleObj,,
  color_cells_by = "pseudotime",
  show_trajectory_graph = TRUE
)
dev.off()

  pdf(file.path(plotDir, 
               "CellFates.pdf"),
     height = 4,
     width = 12)
ggarrange(NoCCL2Fate, CCL18Fate)
dev.off()


p$data$PseudotimeTrem2 <- p$data$cell_color
FullDataFramePseudoTime$pseudotime_CCL[ ! rownames(FullDataFramePseudoTime) %in% CCLCells] <- 0

pdf(height = 4, width = 12)
ggplot( DataframeTrem2Pseudotime, aes(x=data_dim_1,y=data_dim_2,color=PseudotimeTrem2)) + geom_point() + scale_color_viridis( option="B") + labs(title = "Trem2 Fated Cells") + NoLegend()
dev.off()

```

```{r DE across selected trajectory subsets}

CCL18High <- choose_graph_segments(monocleObj, clear_cds = FALSE)
#You will have to remove the monocleObj -- these are too big to work on while keeping both in memory.

CCL18High <- order_cells(CCL18High)

pdf(file.path(plotDir, "CCL18HighWithTrajectory.pdf"), width = 12, height = 8)
plot_cells(CCL18High,
           color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=1.5) + labs( title = "High CCL18 Trajectory")
dev.off()

CCL18High <- cluster_cells(CCL18High)

pdf(file.path(plotDir, "CCL18HighWithClusters.pdf"), width = 12, height = 8)
plot_cells(CCL18High,
           color_cells_by = "cluster",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=1.5) + labs(title = "High CCL18 Trajectory")
dev.off()
CCL18High <- learn_graph(CCL18High)
CCL18HighFateGenes <- graph_test(CCL18High, neighbor_graph = "principal_graph")
pr_deg_ids <- row.names(subset(CCL18HighFateGenes, q_value < 0.05))
gene_module_df <- find_gene_modules(CCL18High[pr_deg_ids,], resolution=1e-3)



cell_group_df <- tibble::tibble(cell=row.names(colData(CCL18High)), 
                                cell_group=CCL18High$monocle_cluster)
agg_mat <- aggregate_gene_expression(CCL18High, gene_module_df, cell_group_df)
row.names(agg_mat) <- stringr::str_c("Module ", row.names(agg_mat))
colnames(agg_mat) <- stringr::str_c("Cluster ", colnames(agg_mat))


pdf(file.path(plotDir, "heatmapModulesCCL18High.pdf"), height = 6, width = 6)
pheatmap::pheatmap(agg_mat, cluster_rows=TRUE, cluster_cols=FALSE,
                   scale="column", clustering_method="ward.D2",
                   fontsize=6, main = "High CCL18 Fate Derived Gene Modules vs High Trajectory")
dev.off()

Module1 <- gene_module_df[gene_module_df$module ==1,]
Module2 <- gene_module_df[gene_module_df$module ==2,]
Module3 <- gene_module_df[gene_module_df$module ==3,]
Module4 <- gene_module_df[gene_module_df$module ==4,]
Module5 <- gene_module_df[gene_module_df$module ==5,]
Module6 <- gene_module_df[gene_module_df$module ==6,]
Module7 <- gene_module_df[gene_module_df$module ==7,]
Module8 <- gene_module_df[gene_module_df$module ==8,]
Module9 <- gene_module_df[gene_module_df$module ==9,]
Module10 <- gene_module_df[gene_module_df$module ==10,]
Module11 <- gene_module_df[gene_module_df$module ==11,]
Module12 <- gene_module_df[gene_module_df$module ==12,]
Module13 <- gene_module_df[gene_module_df$module ==13,]
Module14 <- gene_module_df[gene_module_df$module ==14,]

write.table(Module1, "Module1.txt")
write.table(Module2, "Module2.txt")
write.table(Module3, "Module3.txt")
write.table(Module4, "Module4.txt")
write.table(Module5, "Module5.txt")
write.table(Module6, "Module6.txt")
write.table(Module7, "Module7.txt")
write.table(Module8, "Module8.txt")
write.table(Module9, "Module9.txt")
write.table(Module10, "Module10.txt")
write.table(Module11, "Module11.txt")
write.table(Module12, "Module12.txt")
write.table(Module13, "Module13.txt")
write.table(Module14, "Module14.txt")

CCL18Low <- choose_graph_segments(monocleObj, clear_cds = FALSE)
#You will have to remove the monocleObj -- these are too big to work on while keeping both in memory.

CCL18Low <- order_cells(CCL18Low)

pdf(file.path(plotDir, "CCL18LowWithTrajectory.pdf"), width = 12, height = 8)
plot_cells(CCL18Low,
           color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=1.5) + labs(title = "Low CCL18 Trajectory")
dev.off()

CCL18Low <- cluster_cells(CCL18Low)
CCL18Low <- learn_graph(CCL18Low)
CCL18LowFateGenes <- graph_test(CCL18Low, neighbor_graph = "principal_graph")
pr_deg_ids2 <- row.names(subset(CCL18LowFateGenes, q_value < 0.05))
gene_module_df2 <- find_gene_modules(CCL18Low[pr_deg_ids2,], resolution=.1)

pdf(file.path(plotDir, "CCL18LowWithClusters.pdf"), width = 12, height = 8)
plot_cells(CCL18Low,
           color_cells_by = "cluster",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=1.5) + labs(title = "Low CCL18 Trajectory")
dev.off()

cell_group_df2 <- tibble::tibble(cell=row.names(colData(CCL18Low)), 
                                cell_group=CCL18Low$monocle_clusters)
agg_mat2 <- aggregate_gene_expression(CCL18Low, gene_module_df2, cell_group_df2)
row.names(agg_mat2) <- stringr::str_c("Module ", row.names(agg_mat2))
colnames(agg_mat2) <- stringr::str_c("Cluster ", colnames(agg_mat2))


pdf(file.path(plotDir, "heatmapModulesCCL18Low.pdf"), height = 6, width = 6)
pheatmap::pheatmap(agg_mat2, cluster_rows=TRUE, cluster_cols=FALSE,
                   scale="column", clustering_method="ward.D2",
                   fontsize=6,main = "Low CCL18 Fate Derived Gene Models vs Low Trajectory")
dev.off()

LowModule1 <- gene_module_df2[gene_module_df2$module ==1,]
LowModule2 <- gene_module_df2[gene_module_df2$module ==2,]
LowModule3 <- gene_module_df2[gene_module_df2$module ==3,]
LowModule4 <- gene_module_df2[gene_module_df2$module ==4,]
LowModule5 <- gene_module_df2[gene_module_df2$module ==5,]
LowModule6 <- gene_module_df2[gene_module_df2$module ==6,]
LowModule7 <- gene_module_df2[gene_module_df2$module ==7,]
LowModule8 <- gene_module_df2[gene_module_df2$module ==8,]
LowModule9 <- gene_module_df2[gene_module_df2$module ==9,]
LowModule10 <- gene_module_df2[gene_module_df2$module ==10,]
LowModule11 <- gene_module_df2[gene_module_df2$module ==11,]
LowModule12 <- gene_module_df2[gene_module_df2$module ==12,]
LowModule13 <- gene_module_df2[gene_module_df2$module ==13,]


write.table(LowModule1, "Module1.txt")
write.table(LowModule2, "Module2.txt")
write.table(LowModule3, "Module3.txt")
write.table(LowModule4, "Module4.txt")
write.table(LowModule5, "Module5.txt")
write.table(LowModule6, "Module6.txt")
write.table(LowModule7, "Module7.txt")
write.table(LowModule8, "Module8.txt")
write.table(LowModule9, "Module9.txt")
write.table(LowModule10, "Module10.txt")
write.table(LowModule11, "Module11.txt")
write.table(LowModule12, "Module12.txt")
write.table(LowModule13, "Module13.txt")

agg_mat3 <- aggregate_gene_expression(CCL18High, gene_module_df2, cell_group_df)
row.names(agg_mat3) <- stringr::str_c("Module ", row.names(agg_mat3))
colnames(agg_mat3) <- stringr::str_c("Cluster ", colnames(agg_mat3))

col.order <- c("Cluster 4", "Cluster 5", "Cluster 2", "Cluster 3", "Cluster 1")
row.order <-c("Module 7", "Module 8", "Module 11", "Module 3", "Module 13", "Module 2", "Module 4", "Module 9", "Module 12", "Module 5", "Module 10", "Module 1", "Module 6")
agg_mat3 <- agg_mat3[row.order,col.order]
colnames(agg_mat3) <- c("Stage 1", "Stage 2", "Stage 3", "Stage 4", "Stage 5")

agg_mat4 <- aggregate_gene_expression(CCL18Low, gene_module_df, cell_group_df2)
row.names(agg_mat4) <- stringr::str_c("Module ", row.names(agg_mat4))
colnames(agg_mat4) <- stringr::str_c("Cluster ", colnames(agg_mat4))

col.order <- c("Cluster 3", "Cluster 6", "Cluster 1", "Cluster 4", "Cluster 2", "Cluster 5")
row.order <- c("Module 8", "Module 14", "Module 4", "Module 11", "Module 3", "Module 5", "Module 12", "Module 6", "Module 10", "Module 2", "Module 1", "Module 9", "Module 7", "Module 13")
agg_mat4 <- agg_mat4[row.order,col.order]
colnames(agg_mat4) <- c("Stage 1", "Stage 2", "Stage 3", "Stage 4", "Stage 5", "Stage 6")

pdf(file.path(plotDir, "heatmapModulesCCL18LowInHigh.pdf"), height = 6, width = 6)
pheatmap::pheatmap(agg_mat3, cluster_rows=FALSE, cluster_cols=FALSE,
                   scale="column", clustering_method="ward.D2",
                   fontsize=6,main = "Low Fate Derived Gene Models vs High Trajectory")
dev.off()

pdf(file.path(plotDir, "heatmapModulesCCL18HighinLow.pdf"), height = 6, width = 6)
pheatmap::pheatmap(agg_mat4, cluster_rows=FALSE, cluster_cols=FALSE,
                   scale="column", clustering_method="ward.D2",
                   fontsize=6,main = "High  Fate Derived Gene Models vs Low Trajectory")
dev.off()

```

```{r GEO incorporation}

library(GEOquery)

gds <- getGEO("GSE151928")

sampleAnno <- gds$GSE151928_series_matrix.txt.gz@phenoData@data

#Test use of seurat to read in geo data
library(Seurat)
library(stringr)
library(dplyr)
library(ggplot2); theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill=NA, size=1),
          axis.text=element_text(colour="black"),
          axis.ticks=element_line(colour="black"),
          legend.key = element_blank(),
          strip.text.x = element_text(size = 14,margin = margin( b = 2, t = 2) ),
          strip.background = element_rect(fill="white", colour="black")))
library(kableExtra)

baseDir <- "/Users/mlawrance/Library/CloudStorage/Box-Box/P458_SLICC_Mikacenic"
dataDir <- file.path(baseDir, "geoData/GSE151928_healthy_BAL/")
plotDir <- file.path(baseDir, "geoData/GSE151928_healthy_BAL/plots")
#sftsPlotDir <- file.path(plotDir, "SFTS10X")

geoDir <- dataDir

geoFolders <- list.files(file.path(geoDir),
                         pattern = "*GSM*",
                         full.names = T)

geoFolders <- geoFolders[!(str_detect(geoFolders, ".gz$"))]

counts <- lapply(geoFolders, read.csv, header = TRUE)

for(i in 1:9){
  temp <- counts[[i]]
  rownames(temp)<- temp[,1]
  counts[[i]] <- temp[,-1]
}

expressionMatrices <- lapply(geoFolders, Read10X)

seuratObjects <- lapply(counts,
                        function(x) CreateSeuratObject(counts = x,
                                                       min.features = 100,
                                                       min.cells = 3))

#Order sample anno to match geoFolders
fileSampleOrder <- geoFolders  %>%
  sub("*GSM", "") %>%
  str_remove("\\/")

sampleAnno <- sampleAnno[match(fileSampleOrder, sampleAnno$title),]

for(i in 1:length(seuratObjects)){
  seurat <- seuratObjects[[i]]
  seurat[["percent_mt"]] <- PercentageFeatureSet(seurat, pattern = "^MT-")
  seurat[["file"]] <- geoFolders[i]
  seurat[["sample"]] <- sampleAnno$title[i]
  #seurat[["diseaseState"]] <- sampleAnno$`characteristics:ch1.3`[i]
  #seurat[["illnessPhase"]] <- sampleAnno$`illness phase:ch1`[i]
  seurat[["sex"]] <- sampleAnno$`Sex:ch1`[i]
  seurat[["age"]] <- sampleAnno$`age:ch1`[i]
  AddMetaData(seurat, paste("Healthy_", i), col.name = "sampleID")
  seuratObjects[[i]] <- seurat
}


nFeatureHistograms <- lapply(seuratObjects, function(x) hist(x$nFeature_RNA, 100, xlim= c(0,5000), main = unique(x$sample) , xlab = "nFeature RNA"))

pctMitoHistograms <- lapply(seuratObjects, function(x) hist(x$percent_mt, 50, xlim= c(0,20), main = unique(x$sample), xlab = "% mitochondrial reads"))

nFeatureLow <-500 
nFeatureHigh <- 5000
pctMt <- 20

plotLayers <- list(geom_vline(xintercept = nFeatureLow),
                   geom_vline(xintercept = nFeatureHigh),
                   geom_hline(yintercept = pctMt),
                   labs(x = "Number of genes",
                        y = "% mito reads",
                        title = ""),
                   xlim(c(0, 7000)),
                   ylim(c(0, 100)),
                   theme(legend.position = 'none'))

featureScatterPlots <- lapply(seuratObjects, function(x) FeatureScatter(x, 
                                                                        feature1 = "nFeature_RNA", 
                                                                        feature2 = "percent_mt") +
                                plotLayers +
                                ggtitle(unique(x$sample)))

seuratQC <- NULL

for(i in 1:length(seuratObjects)){
  seurat <- seuratObjects[[i]]
  
  seuratQC[[i]] <- subset(seurat, 
                          subset = nFeature_RNA > nFeatureLow & nFeature_RNA < nFeatureHigh #& percent_mt < pctMt)
  )
}

#Tally how many cells pass QC
sampleAnno$nCells <- NA
sampleAnno$nCellsQC <- NA

for(i in 1:length(seuratObjects)){
  
  selectedSample <- unique(seuratObjects[[i]]$sample) %>%
    as.character()
  
  sampleAnno$nCells[sampleAnno$title == selectedSample] <- dim(seuratObjects[[i]])[2]
  
  sampleAnno$nCellsQC[sampleAnno$title == selectedSample] <- dim(seuratQC[[i]])[2]
  
}

sampleAnno$pctPass <- round(sampleAnno$nCellsQC/sampleAnno$nCells*100,2)

sampleAnno %>%
  dplyr::select(title, nCells, nCellsQC, pctPass) %>%
  dplyr::rename(Sample = title,
                `Number of cells` = nCells,
                `Number of high quality cells` = nCellsQC,
                `Percent of cells that pass QC` = pctPass) %>%
  kable(row.names = F,
        caption = "Quality analysis summary") %>%
  kable_styling(full_width = F, 
                position = "left")


#Merge and visualize without integration
seuratMerged <- merge(seuratQC[[1]], 
                      y = c(seuratQC[2:10]),
                      add.cell.ids = fileSampleOrder,
                      merge.data = TRUE)

seuratMerged <- NormalizeData(object = seuratMerged, verbose = FALSE)
seuratMerged <- FindVariableFeatures(object = seuratMerged, 
                                     selection.method = "vst", nfeatures = 2000, verbose = FALSE)

seuratMerged <- ScaleData(object = seuratMerged, verbose = FALSE)
seuratMerged <- RunPCA(object = seuratMerged, npcs = 30, verbose = FALSE)
#ElbowPlot(seuratMerged)

seuratMerged <- FindNeighbors(seuratMerged, dims = 1:15)
seuratMerged <- FindClusters(seuratMerged, resolution = 0.5)

seuratMerged <- RunUMAP(object = seuratMerged, reduction = "pca", 
                        dims = 1:15)

gMergedSample <- DimPlot(object = seuratMerged, 
                         reduction = "umap", 
                         group.by = "sample",
                         shuffle = TRUE)+
  labs(x = "UMAP 1", y ="UMAP 2") 

FeaturePlot(object = seuratMerged, 
            features  = c("CD14"),
            order = TRUE,
            ncol = 2) &
  labs(x = "UMAP 1",
       y = "UMAP 2")&
  theme(aspect.ratio = 1)

subset(x = seuratMerged, idents = c("0","1","2","3","5","7","8","10","11"))
       
healthysplit <- SplitObject(healthySubset, split.by = "sample")
 ventsplit <- SplitObject(MonocytesIntegratedViaCluster, split.by = "sample")
allSplit <- c(ventsplit, healthysplit)
rm(ventsplit)
rm(healthysplit)

features <- SelectIntegrationFeatures(object.list = allSplit)
allSplit <- lapply(X = allSplit, FUN = function(x) {
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
})

fastIntAnchors <- FindIntegrationAnchors(object.list = allSplit, 
                                         anchor.features = features,
                                         reduction = "rpca")

save(fastIntAnchors,
     file = file.path(dataDir, "220526_integration_fastIntAnchors.RData"))

allIntegrated <- IntegrateData(anchorset = fastIntAnchors)

DefaultAssay(allIntegrated) <- 'RNA'
allIntegrated <- NormalizeData(allIntegrated) %>% ScaleData() %>% RunPCA(
                                        features = rownames(visitMarkersRNA),
                                        assay = "RNA",
                                        reduction.name = "visitRNAPCA")

DefaultAssay(allIntegrated) <- 'FB'
allIntegrated <- NormalizeData(allIntegrated) %>% ScaleData() %>% RunPCA(
                                        features = rownames(visitMarkersFB),
                                        assay = "FB",
                                        reduction.name = "visitFBPCA")

allIntegrated <- FindMultiModalNeighbors(allIntegrated,
                                               reduction.list = list("visitFBPCA", "visitRNAPCA"),
                                               dims.list = list(1:20, 1:10),
                                               modality.weight.name = list("BestIntegratedFB.weight", "BestIntegratedRNA.weight"))

allIntegrated <- RunUMAP(allIntegrated, 
                               nn.name = "weightedvisit.nn",
                               reduction.name = "umap", 
                               reduction.key = "UMAP_")

allIntegrated <- RunUMAP(allIntegrated, 
                         nn.name = "weighted.nn", 
                         reduction.name = "wnn.umap", 
                         reduction.key = "wnnUMAP_")

DefaultAssay(allIntegrated) <- 'integrated'

allIntegrated <- FindVariableFeatures(allIntegrated, assay = "integrated", selection.method = "vst")

allIntegrated <- ScaleData(allIntegrated, assay = "integrated")

allIntegrated <- RunPCA(allIntegrated, assay = "integrated", features = visitMarkersRNA, reduction.name = "integratedPCAsupervised")

DimPlot(allIntegrated, group.by = "studyGroup", reduction = "visitintPCA")


pdf("VisitUMAPTransparent_recolored.pdf", width = 6, height = 6)
  ggplot(
    data = allIntegrated@meta.data,
    mapping =
      aes(x = allIntegrated@reductions$visitRNAPCA_OLD@cell.embeddings[,1],
          y = allIntegrated@reductions$visitRNAPCA_OLD@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = "visit"), size = .25, alpha = 0.1)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Visit Timing", 
    values = c("gray","blue", "orange"),
    labels = c("Healthy", "B1", "B2")) +
 # facet_wrap(~studyGroup, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))
  dev.off()

allIntegrated <- NormalizeData(allIntegrated) %>% ScaleData() %>% RunPCA(
                                        features = rownames(visitMarkersRNA),
                                        assay = "RNA",
                                        reduction.name = "visitRNAPCA_OLD")

allIntegrated <- FindNeighbors(allIntegrated, assay ="integrated")

allIntegrated <- RunUMAP(allIntegrated, assay = "integrated", nn.name = "integratedvisit.nn", reduction.name = "integratedvisit.umap", reduction.key = "_intvistUMAP")


library(xlsx)
write.xlsx(Visit2VersusVisit1Positive, file="tables/ExpressionChangeTime.xlsx", sheetName="Visit1-Visit2 UP", row.names=TRUE)
write.xlsx(Visit2VersusVisit1Negative, file="tables/ExpressionChangeTime.xlsx", sheetName="Visit1-Visit2 DOWN", append=TRUE, row.names=TRUE)
write.xlsx(HealthyVersusVisit1Positive, file="tables/ExpressionChangeTime.xlsx", sheetName="Visit1-Healthy UP", append=TRUE, row.names=TRUE)
write.xlsx(HealthyVersusVisit1Negative, file="tables/ExpressionChangeTime.xlsx", sheetName="Visit1-Healthy DOWN", append=TRUE, row.names=TRUE)
write.xlsx(HealthyVersusVisit2Positive, file="tables/ExpressionChangeTime.xlsx", sheetName="Visit2-Healthy UP", append=TRUE, row.names=TRUE)
write.xlsx(HealthyVersusVisit2Negative, file="tables/ExpressionChangeTime.xlsx", sheetName="Visit2-Healthy DOWN", append=TRUE, row.names=TRUE)
write.xlsx(ConsistentlyPositive, file="tables/ExpressionChangeTime.xlsx", sheetName="Consistently UP", append=TRUE, row.names=TRUE)
write.xlsx(ConsistentlyNegative, file="tables/ExpressionChangeTime.xlsx", sheetName="Consistency DOWN", append=TRUE, row.names=TRUE)

AllVisit2vsVisit1 <- rbind(Visit2VersusVisit1Negative, Visit2VersusVisit1Positive) 
AllHealthyvsVisit2 <- rbind(HealthyVersusVisit2Positive, HealthyVersusVisit2Negative)
GenesForPlotting <- AllVisit2vsVisit1[rownames(AllVisit2vsVisit1) %in% rownames(AllHealthyvsVisit2),]

S100A9 <- cpmY[rownames(cpmY) %in% "S100A9",]
S100A9 <- as.data.frame(S100A9)
S100A9$Group <- factor(c("Visit 1", "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy"), levels = c("Visit 1", "Visit 2", "Healthy"))
ggplot(S100A9, aes(x = Group, y = S100A9)) + 
    geom_boxplot() + geom_dotplot(binaxis='y', stackdir='center', dotsize=1)
                       
CCL18 <- cpmY[rownames(cpmY) %in% "CCL18",]
CCL18 <- as.data.frame(CCL18)
CCL18$Group <- factor(c("Visit 1", "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy"), levels = c("Visit 1", "Visit 2", "Healthy"))

ggplot(CCL18, aes(x = Group, y = CCL18)) + 
    geom_boxplot() + geom_dotplot(binaxis='y', stackdir='center', dotsize=1)

S100A8 <- cpmY[rownames(cpmY) %in% "S100A8",]
S100A8 <- as.data.frame(S100A8)
S100A8$Group <- factor(c("Visit 1", "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy"), levels = c("Visit 1", "Visit 2", "Healthy"))

ggplot(S100A8, aes(x = Group, y = S100A8)) + 
    geom_boxplot() + geom_dotplot(binaxis='y', stackdir='center', dotsize=1)


DPEP2 <- cpmY[rownames(cpmY) %in% "DPEP2",]
DPEP2 <- as.data.frame(DPEP2)
DPEP2$Group <- factor(c("Visit 1", "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy"), levels = c("Visit 1", "Visit 2", "Healthy"))

ggplot(DPEP2, aes(x = Group, y = DPEP2)) + 
    geom_boxplot() + geom_dotplot(binaxis='y', stackdir='center', dotsize=1)

TBL2 <- cpmY[rownames(cpmY) %in% "TBL2",]
TBL2 <- as.data.frame(TBL2)
TBL2$Group <- factor(c("Visit 1", "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy"), levels = c("Visit 1", "Visit 2", "Healthy"))

ggplot(TBL2, aes(x = Group, y = TBL2)) + 
    geom_boxplot() + geom_dotplot(binaxis='y', stackdir='center', dotsize=1)

GeneList <- list()
for(i in rownames(GenesForPlotting)){
  samp <- i
  obj <- paste0(samp, "_Exp")
  gene <- avgExp$RNA[rownames(avgExp$RNA) %in% samp,]
  obj <- as.data.frame(gene)
  colnames(obj) <- samp
  obj$Group <- factor(c("Visit 1", "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Visit 1",  "Visit 2",  "Visit 1",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy"), levels = c("Visit 1", "Visit 2", "Healthy"))
  GeneList[[samp]] <- obj
}

Plots_Time <- lapply(GeneList, function(df) ggplot( df, aes(x = Group, y = df[,1])) + 
                                                 geom_boxplot()  +
                                                 geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
                                                 labs(y = paste0(colnames(df)[1]), "_avgExp"))

for(i in names(Plots_Time)){
  pathpdf = paste0(plotDir, "avgExpBox/Visit/", i, ".pdf")
  pdf(pathpdf, height = 6, width = 6)
  print(Plots_Time[[i]])
  dev.off()
}

Idents(allIntegrated) <- "studyGroup"
NoArdsvsArds <- FindMarkers(allIntegrated, ident.1 = "No ARDS", ident.2 = "ARDS", logfc.threshold = .8)
HealthyvsNoArds <- FindMarkers(allIntegrated, ident.1 = "Healthy", ident.2 = "No ARDS", logfc.threshold = .8)
HealthyvsARDS <- FindMarkers(allIntegrated, ident.1 = "Healthy", ident.2 = "No ARDS", logfc.threshold = .8)

write.xlsx(Visit2VersusVisit1Positive, file="tables/ExpressionChangeStatus.xlsx", sheetName="ARDS-No ARDS UP", row.names=TRUE)
write.xlsx(Visit2VersusVisit1Negative, file="tables/ExpressionChangeStatus.xlsx", sheetName="ARDS-No ARDS DOWN", append=TRUE, row.names=TRUE)
write.xlsx(HealthyVersusVisit1Positive, file="tables/ExpressionChangeStatus.xlsx", sheetName="ARDS-Healthy UP", append=TRUE, row.names=TRUE)
write.xlsx(HealthyVersusVisit1Negative, file="tables/ExpressionChangeStatus.xlsx", sheetName="ARDS-Healthy DOWN", append=TRUE, row.names=TRUE)
write.xlsx(HealthyVersusVisit2Positive, file="tables/ExpressionChangeStatus.xlsx", sheetName="No ARDS-Healthy UP", append=TRUE, row.names=TRUE)
write.xlsx(HealthyVersusVisit2Negative, file="tables/ExpressionChangeStatus.xlsx", sheetName="No ARDS-Healthy DOWN", append=TRUE, row.names=TRUE)


write.xlsx(Visit2VersusVisit1Positive, file="tables/ExpressionChangeTime.xlsx", sheetName="Visit1-Visit2 UP", row.names=TRUE)
write.xlsx(Visit2VersusVisit1Negative, file="tables/ExpressionChangeTime.xlsx", sheetName="Visit1-Visit2 DOWN", append=TRUE, row.names=TRUE)
write.xlsx(HealthyVersusVisit1Positive, file="tables/ExpressionChangeTime.xlsx", sheetName="Visit1-Healthy UP", append=TRUE, row.names=TRUE)
write.xlsx(HealthyVersusVisit1Negative, file="tables/ExpressionChangeTime.xlsx", sheetName="Visit1-Healthy DOWN", append=TRUE, row.names=TRUE)
write.xlsx(HealthyVersusVisit2Positive, file="tables/ExpressionChangeTime.xlsx", sheetName="Visit2-Healthy UP", append=TRUE, row.names=TRUE)
write.xlsx(HealthyVersusVisit2Negative, file="tables/ExpressionChangeTime.xlsx", sheetName="Visit2-Healthy DOWN", append=TRUE, row.names=TRUE)
write.xlsx(ConsistentlyPositive, file="tables/ExpressionChangeTime.xlsx", sheetName="Consistently UP", append=TRUE, row.names=TRUE)
write.xlsx(ConsistentlyNegative, file="tables/ExpressionChangeTime.xlsx", sheetName="Consistency DOWN", append=TRUE, row.names=TRUE)

GenesForPlotting <- NoArdsvsArds[rownames(NoArdsvsArds) %in% rownames(HealthvsNoArds),]

GeneList <- list()
for(i in rownames(GenesForPlotting)){
  samp <- i
  obj <- paste0(samp, "_Exp")
  gene <- avgExp$RNA[rownames(avgExp$RNA) %in% samp,]
  obj <- as.data.frame(gene)
  colnames(obj) <- samp
  obj$Group <- factor(c("No ARDS", "No ARDS",  "No ARDS",  "No ARDS",  "No ARDS",  "ARDS",  "ARDS",  "ARDS",  "ARDS",  "ARDS",  "ARDS",  "No ARDS",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy",  "Healthy"), levels = c("ARDS", "No ARDS", "Healthy"))
  GeneList[[samp]] <- obj
}
group_status <- factor(c(1,1,1,1,1,2,2,2,2,2,2,1,3,3,3,3,3,3,3,3,3))
Plots_Status <- lapply(GeneList, function(df) ggplot( df, aes(x = Group, y = df[,1])) + 
                                                 geom_boxplot()  +
                                                 geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
                                                 labs(y = paste0(colnames(df)[1]), "_avgExp"))

for(i in names(Plots_Status)){
  pathpdf = paste0(plotDir, "avgExpBox/Status/", i, ".pdf")
  pdf(pathpdf, height = 6, width = 6)
  print(Plots_Status[[i]])
  dev.off()
}

for_psuedobulk <- list()
for(i in levels(allIntegrated)){
  temp <- subset(allIntegrated, idents = i)
  name <- i
  temp <- temp@assays$RNA@counts
  temp <- Matrix::rowSums(temp)
  temp <- as.data.frame(temp)
  colnames(temp) <- name
  for_psuedobulk[[i]] <- as.matrix(temp)
}

for(i in targets){
  name <- i
 pseudobulk <- transform(merge(pseudobulk,for_psuedobulk$name,by=0,all=TRUE), row.names=Row.names, Row.names=NULL)
}


group_time <- factor(c(1,2,1,1,2,1,1,2,1,1,2,1,3,3,3,3,3,3,3,3,3))
group_status <- factor(c(1,1,1,1,1,2,2,2,2,2,2,1,3,3,3,3,3,3,3,3,3))
y <- DGEList(pseudobulk, group = group_time)
design <- model.matrix(~0+group_time, data=y$samples)
keep <- filterByExpr(y)
y <- y[keep,,keep.lib.sizes=FALSE]
y <- calcNormFactors(y)
y <- estimateDisp(y,design)
colnames(design) <- c("v1", "v2", "Healthy")
contr.matrix <- makeContrasts(Visit2vsVisit1 = v2-v1,
                              HealthyvsVisit2 = Healthy-v2,
                              HealthyvsVisit1 = Healthy-v1,
                              levels=design)

v<- voom(y, design, plot=TRUE)

pdf(file.path(plotDir, "MFLPlots/VoomVisit.pdf"), height = 6, width =12)
voom(y, design, plot=TRUE)
dev.off()

fit <- lmFit(v, design)
fit_c <- contrasts.fit(fit, contr.matrix)
fit_c_eb <- eBayes(fit_c)



qlf.v2v1 <- glmQLFTest(fit, contrast=Visit2VsVisit1)
genesv2v1 <- topTags(qlf.v2v1, n = 20000, p.value = 1)
qlf.HealthyV2 <- glmQLFTest(fit, contrast=HealthyVsVisit2)
genesHealthyV2 <- topTags(qlf.HealthyV2, n= 1000, p.value = .05)
qlf.HealthyV1 <- glmQLFTest(fit, contrast=HealthyVsVisit1)
genesHealthyV1 <- topTags(qlf.HealthyV1, n= 1000, p.value = .05)

clusterAverages <- AverageExpression(allIntegrated, 
                                    assays = "RNA",
                                    group.by = "seurat_clusters")

Idents(allIntegrated) <- "seurat_clusters"
clusterMarkers <- FindAllMarkers(allIntegrated, assay ="RNA")
```

```{r Basic Clustering}

MonocytesIntegratedViaCluster <- FindVariableFeatures(MonocytesIntegratedViaCluster, selection.method = "vst", nfeatures = 2000)
MonocytesIntegratedViaCluster <- ScaleData(MonocytesIntegratedViaCluster, assay = "RNA", verbose = FALSE) %>% RunPCA(assay = "RNA", reduction.name = "pca_MonoRNA", reduction.key = "pca_MonoRNA_")




MonocytesIntegratedViaCluster <- FindNeighbors(MonocytesIntegratedViaCluster, dims = 1:13, graph.name = "MonoRNA.nn")
MonocytesIntegratedViaCluster <- FindClusters(MonocytesIntegratedViaCluster, resolution = 0.5, graph.name = "MonoRNA.nn")
MonocytesIntegratedViaCluster <- MonocytesIntegratedViaCluster %>% AddMetaData(Idents(MonocytesIntegratedViaCluster), col.name = "Cluster.5")
MonocytesIntegratedViaCluster <- FindClusters(MonocytesIntegratedViaCluster, resolution = 0.3, graph.name = "MonoRNA.nn")
MonocytesIntegratedViaCluster <- MonocytesIntegratedViaCluster %>% AddMetaData(Idents(MonocytesIntegratedViaCluster), col.name = "Cluster.3")
MonocytesIntegratedViaCluster <- FindClusters(MonocytesIntegratedViaCluster, resolution = 0.2, graph.name = "MonoRNA.nn")
MonocytesIntegratedViaCluster <- MonocytesIntegratedViaCluster %>% AddMetaData(Idents(MonocytesIntegratedViaCluster), col.name = "Cluster.2")
MonocytesIntegratedViaCluster <- FindClusters(MonocytesIntegratedViaCluster, resolution = 0.1, graph.name = "MonoRNA.nn")
MonocytesIntegratedViaCluster <- MonocytesIntegratedViaCluster %>% AddMetaData(Idents(MonocytesIntegratedViaCluster), col.name = "Cluster.1")
MonocytesIntegratedViaCluster <- FindClusters(MonocytesIntegratedViaCluster, resolution = 0.15, graph.name = "MonoRNA.nn")
MonocytesIntegratedViaCluster <- MonocytesIntegratedViaCluster %>% AddMetaData(Idents(MonocytesIntegratedViaCluster), col.name = "Cluster.15")
MonocytesIntegratedViaCluster <- FindClusters(MonocytesIntegratedViaCluster, resolution = 0.25, graph.name = "MonoRNA.nn")
MonocytesIntegratedViaCluster <- MonocytesIntegratedViaCluster %>% AddMetaData(Idents(MonocytesIntegratedViaCluster), col.name = "Cluster.25")


cluster50Umap <- ggplot(
    data = MonocytesIntegratedViaCluster@meta.data,
    mapping =
        aes(x = MonocytesIntegratedViaCluster@reductions$MonoRNA.umap@cell.embeddings[,1],
            y = MonocytesIntegratedViaCluster@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = MonocytesIntegratedViaCluster@meta.data$Cluster.5), size = .4, alpha = 0.8),
    dpi = 300) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(MonocytesIntegratedViaCluster@meta.data$Cluster.5))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

cluster30Umap <- ggplot(
    data = MonocytesIntegratedviaCluster@meta.data,
    mapping =
        aes(x = MonocytesIntegratedviaCluster@reductions$MonoRNA.umap@cell.embeddings[,1],
            y = MonocytesIntegratedviaCluster@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = MonocytesIntegratedviaCluster@meta.data$Cluster.3), size = .4, alpha = 0.8),
    dpi = 300) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(MonocytesIntegratedviaCluster@meta.data$Cluster.3))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

cluster25Umap <-ggplot(
    data = MonocytesIntegratedViaCluster@meta.data,
    mapping =
        aes(x = MonocytesIntegratedViaCluster@reductions$MonoRNA.umap@cell.embeddings[,1],
            y = MonocytesIntegratedViaCluster@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = MonocytesIntegratedViaCluster@meta.data$Cluster.25), size = .4, alpha = 0.8),
    dpi = 300) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(MonocytesIntegratedViaCluster@meta.data$Cluster.25))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

cluster20Umap <-ggplot(
    data = MonocytesIntegratedViaCluster@meta.data,
    mapping =
        aes(x = MonocytesIntegratedViaCluster@reductions$MonoRNA.umap@cell.embeddings[,1],
            y = MonocytesIntegratedViaCluster@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = MonocytesIntegratedViaCluster@meta.data$Cluster.2), size = .4, alpha = 0.8),
    dpi = 300) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(MonocytesIntegratedViaCluster@meta.data$Cluster.2))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

cluster15Umap <- ggplot(
    data = MonocytesIntegratedViaCluster@meta.data,
    mapping =
        aes(x = MonocytesIntegratedViaCluster@reductions$MonoRNA.umap@cell.embeddings[,1],
            y = MonocytesIntegratedViaCluster@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = MonocytesIntegratedViaCluster@meta.data$Cluster.15), size = .4, alpha = 0.8),
    dpi = 300) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(MonocytesIntegratedViaCluster@meta.data$Cluster.15))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

for(i in 1:length(levels(MonocytesIntegratedViaCluster@meta.data$Cluster.25))){
    cluster <- i
    temp <- Markets25[Markets25$cluster == cluster,][1:5,]
    frame25 <- rbind(frame, temp)
}

frame15 <- data.frame()
for(i in 1:length(levels(MonocytesIntegratedViaCluster@meta.data$Cluster.15)) -1){
    cluster <- i
    temp <- Markers15[Markers15$cluster == cluster,][1:5,]
    frame15 <- rbind(frame15, temp)
}

frame30 <- data.frame()
for(i in 1:length(levels(MonocytesIntegratedViaCluster@meta.data$Cluster.3)) -1){
    cluster <- i
    temp <- Markers30[Markers30$cluster == cluster,][1:5,]
    frame30 <- rbind(frame30, temp)
}

frame50 <- data.frame()
for(i in 1:length(levels(MonocytesIntegratedViaCluster@meta.data$Cluster.5)) -1){
    cluster <- i
    temp <- Markers50[Markers50$cluster == cluster,][1:5,]
    frame50 <- rbind(frame50, temp)
}

frame15Dot <- DotPlot(MonocytesIntegratedViaCluster, features = sub(pattern = "\\..", "", rownames(frame15)), group.by = "Cluster.15", assay = "RNA") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 8))

frame25Dot <- DotPlot(MonocytesIntegratedViaCluster, features = sub(pattern = "\\..", "", rownames(frame25)), group.by = "Cluster.25", assay = "RNA") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 8))

frame30Dot <- DotPlot(MonocytesIntegratedviaCluster, features = sub(pattern = "\\..", "", rownames(frame30)), group.by = "Cluster.3", assay = "RNA") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 8))

frame50Dot <- DotPlot(MonocytesIntegratedViaCluster, features = unique(sub(pattern = "\\..", "", rownames(frame50))), group.by = "Cluster.5", assay = "RNA") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 8))

pdf("plots/FiguresJuly112022/Cluster15.pdf", width = 16, height = 8)
  cluster15Umap + frame15Dot
dev.off()

pdf("plots/FiguresJuly112022/Cluster25.pdf", width = 16, height = 8)
  cluster25Umap + frame25Dot
dev.off()

pdf("plots/FiguresJuly112022/Cluster30.pdf", width = 16, height = 8)
  cluster30Umap + frame30Dot
dev.off()

pdf("plots/FiguresJuly112022/Cluster50.pdf", width = 16, height = 8)
  cluster50Umap + frame50Dot
dev.off()


clusterCounts15 <- MonocytesIntegratedViaCluster@meta.data %>%
  dplyr::group_by(sample, Cluster.15, studyGroup) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(sample) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(sample)

gClusterCounts15 <- clusterCounts15 %>%
  ggplot(aes(x = Cluster.15,
             y = pctCells,
             color = studyGroup,
             fill = sample)) +
  geom_col(position = "fill") +
  scale_color_manual(values = studyGroupColors) + 
  scale_fill_manual(values = big_colorblind_pal(12)) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "Cluster",
       y = "% of cells",
       color = "Study group",
       fill = "Sample") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) + NoLegend()

pdf("plots/FiguresJuly112022/Cluster15BoxSplit.pdf", width = 10, height = 6)
  print(gClusterCounts15)
dev.off()

clusterCounts15Merge <- MonocytesIntegratedViaCluster@meta.data %>%
  dplyr::group_by(Cluster.15, studyGroup) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(studyGroup) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(studyGroup)

gClusterCounts15Merge <- clusterCounts15Merge %>%
  ggplot(aes(x = Cluster.15,
             y = pctCells,
             fill = studyGroup)) +
  geom_col(position = "fill") +
  scale_fill_manual(values = studyGroupColors) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "Cluster",
       y = "% of cells",
       fill = "Study Group") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) + NoLegend()

pdf("plots/FiguresJuly112022/Cluster15BoxMerged.pdf", width = 10, height = 6)
  print(gClusterCounts15Merge)
dev.off()

clusterCounts25 <- MonocytesIntegratedViaCluster@meta.data %>%
  dplyr::group_by(sample, Cluster.25, studyGroup) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(sample) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(sample)

gClusterCounts25 <- clusterCounts25 %>%
  ggplot(aes(x = Cluster.25,
             y = pctCells,
             color = studyGroup,
             fill = sample)) +
  geom_col(position = "fill") +
  scale_color_manual(values = studyGroupColors) + 
  scale_fill_manual(values = big_colorblind_pal(12)) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "Cluster",
       y = "% of cells",
       color = "Study group",
       fill = "Sample") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) + NoLegend()

pdf("plots/FiguresJuly112022/Cluster25BoxSplit.pdf", width = 10, height = 6)
  print(gClusterCounts25)
dev.off()

clusterCounts25Merge <- MonocytesIntegratedViaCluster@meta.data %>%
  dplyr::group_by(Cluster.25, studyGroup) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(studyGroup) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(studyGroup)

gClusterCounts25Merge <- clusterCounts25Merge %>%
  ggplot(aes(x = Cluster.25,
             y = pctCells,
             fill = studyGroup)) +
  geom_col(position = "fill") +
  scale_fill_manual(values = studyGroupColors) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "Cluster",
       y = "% of cells",
       fill = "Study Group") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) + NoLegend()

pdf("plots/FiguresJuly112022/Cluster25BoxMerged.pdf", width = 10, height = 6)
  print(gClusterCounts25Merge)
dev.off()

clusterCounts30 <- MonocytesIntegratedViaCluster@meta.data %>%
  dplyr::group_by(sample, Cluster.3, studyGroup) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(sample) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(sample)

gClusterCounts30 <- clusterCounts30 %>%
  ggplot(aes(x = Cluster.3,
             y = pctCells,
             color = studyGroup,
             fill = sample)) +
  geom_col(position = "fill") +
  scale_color_manual(values = studyGroupColors) + 
  scale_fill_manual(values = big_colorblind_pal(12)) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "Cluster",
       y = "% of cells",
       color = "Study group",
       fill = "Sample") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) + NoLegend()

pdf("plots/FiguresJuly112022/Cluster30BoxSplit.pdf", width = 10, height = 6)
  print(gClusterCounts30)
dev.off()

clusterCounts30Merge <- MonocytesIntegratedViaCluster@meta.data %>%
  dplyr::group_by(Cluster.3, studyGroup) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(studyGroup) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(studyGroup)

gClusterCounts30Merge <- clusterCounts30Merge %>%
  ggplot(aes(x = Cluster.3,
             y = pctCells,
             fill = studyGroup)) +
  geom_col(position = "fill") +
  scale_fill_manual(values = studyGroupColors) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "Cluster",
       y = "% of cells",
       fill = "Study Group") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) + NoLegend()

pdf("plots/FiguresJuly112022/Cluster30BoxMerged.pdf", width = 10, height = 6)
  print(gClusterCounts30Merge)
dev.off()

clusterCounts50 <- MonocytesIntegratedViaCluster@meta.data %>%
  dplyr::group_by(sample, Cluster.5, studyGroup) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(sample) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(sample)

gClusterCounts50 <- clusterCounts50 %>%
  ggplot(aes(x = Cluster.5,
             y = pctCells,
             color = studyGroup,
             fill = sample)) +
  geom_col(position = "fill") +
  scale_color_manual(values = studyGroupColors) + 
  scale_fill_manual(values = big_colorblind_pal(12)) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "Cluster",
       y = "% of cells",
       color = "Study group",
       fill = "Sample") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) + NoLegend()

pdf("plots/FiguresJuly112022/Cluster50BoxSplit.pdf", width = 10, height = 6)
  print(gClusterCounts50)
dev.off()

clusterCounts50Merge <- MonocytesIntegratedViaCluster@meta.data %>%
  dplyr::group_by(Cluster.5, studyGroup) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(studyGroup) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(studyGroup)

gClusterCounts50Merge <- clusterCounts50Merge %>%
  ggplot(aes(x = Cluster.5,
             y = pctCells,
             fill = studyGroup)) +
  geom_col(position = "fill") +
  scale_fill_manual(values = studyGroupColors) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "Cluster",
       y = "% of cells",
       fill = "Study Group") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) + NoLegend()

pdf("plots/FiguresJuly112022/Cluster50BoxMerged.pdf", width = 10, height = 6)
  print(gClusterCounts50Merge)
dev.off()


ARDSMarkersExpression <- ggplot(data = MeanExpMatrix, aes(x=averageExpressionOverall, y=avg_log2FC, color = mycolor1)) +
    geom_point(alpha=0.7, size=1, shape = 19) + 
  scale_color_manual(name = NULL, 
                     values = c('black' = 'gray', 'red' = 'red', 'blue' = 'darkcyan'), 
                     labels = NULL, 
                     breaks = NULL) + 
  geom_text_repel(data=subsetMeanExpMatrix, label = subsetMeanExpMatrix$gene, max.overlaps = Inf, min.segment.length = unit(0, 'lines'), size = 3, col = "black") +
        labs(x = "Overall Expression",
         y = "Log 2 Fold Change",
         title = "ARDS Markers") +
    theme(plot.title = element_text(hjust = .5)) +
    NoLegend()

ARDSvsNoARDSUMAP <- ggplot(
    data = MonocytesIntegratedviaCluster@meta.data,
    mapping =
        aes(x = MonocytesIntegratedviaCluster@reductions$MonoRNA.umap@cell.embeddings[,1],
            y = MonocytesIntegratedviaCluster@reductions$MonoRNA.umap@cell.embeddings[,2])) +
    ggrastr::rasterise(
        geom_point(mapping = aes_string(color = MonocytesIntegratedviaCluster@meta.data$studyGroupAdj), size = .4, alpha = 0.8),
        dpi = 300) +
    labs(x = "UMAP 1", y = "UMAP 2") +
    scale_color_manual(
        "Study Group", 
        values = studyGroupColors,
        labels = c("ARDS", "No ARDS")) +
    guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

pdf("plots/FiguresJuly112022/Cluster50BoxMerged.pdf", width = 10, height = 6)
  print(gClusterCounts50Merge)
dev.off()

pdf("plots/FiguresJuly112022/ARDSMAplot.pdf", width = 12, height = 6)
  print(ARDSMarkersExpression)
dev.off()
```

```{r PDL stuff}

#Order of operations = Pull FB from Seurat, then rename PDL1 to something I can use, then reintergrate as a new assay, normalize, scale, then use nested conditionals to set a new ident for PDL1 expression and find markers, build a UMAP based on those markers for all genes, and then feature plot PDL1.

counts <- MonocytesIntegratedviaCluster@meta.data$assays$FB$counts
rowname(counts[2,]) <- "PDL1"
FB2 <- CreateAssayObject(counts =  counts)
MonocytesIntegratedviaCluster[["FB2"]] <- FB2


MonocytesIntegratedViaCluster <- NormalizeData(MonocytesIntegratedViaCluster,
                                               assay = "RNA") %>%
                                ScaleData(features = rownames(PDL1Markers)) %>%
                                RunPCA(features = rownames(PDL1Markers)[1:100],
                                        assay = "RNA",
                                        reduction.name = "PDL1PCA")

MonocytesIntegratedViaCluster <- FindNeighbors(MonocytesIntegratedViaCluster, 
                               assay ="RNA", 
                               reduction = "PDL1PCA",
                               graph.name = "PDL1.nn", 
                              dims = 1:30)

MonocytesIntegratedViaCluster <- FindClusters(MonocytesIntegratedViaCluster, resolution = 0.3, graph.name = "PDL1.nn")

MonocytesIntegratedViaCluster <- RunUMAP(MonocytesIntegratedViaCluster,
                                         assay = "RNA",
                                         reduction ="PDL1PCA",
                                         dims= 1:30,
                                         reduction.name = "PDL1.umap", 
                                         reduction.key = "_PDL1UMAP")

pdf("plots/FiguresJuly112022//PDL1.pdf", width = 12, height =8)
PDL1Umap + PDL1Features
dev.off()
```

```{r ARDS Markers}

SignificantMeanExp <- MeanExpMatrix[MeanExpMatrix$avg_log2FC > .25 | MeanExpMatrix$avg_log2FC < -.25,]


#Extract putative gene names and synonymize gene names; future need to use Mygene API to look up aliases -- even after this converstion, more than half of the genes didn't map to ENTREZ and had to be manually adjusted to their actual gene names
enrichmentSet <- SignificantMeanExp$Row.names %>% unique() %>% gsub("\\.1", "") %>% gsub(" \\(.*", "")
backgroundSet <- MeanExpMatrix$Row.names %>% unique()


# Convert gene names to ENTREZ ID
enrichmentSetConverted <- bitr(enrichmentSet, fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = org.Hs.eg.db )

ARDSMarkerGoana <- goana(enrichmentSetConverted$ENTREZID, universe = backgroundSetConvert$ENTREZID) %>% topGO(ontology = "BP")

write.xlsx(ARDSMarkerGoana, file= "tables/ARDSMarkersGoana.xlsx", sheetName = "TopGO ARDSvsNON", row.names =  TRUE)
```

```{r time by cluster30}
FacetVisitUMAP <- ggplot(
    data = MonocytesIntegratedviaCluster@meta.data,
    mapping =
        aes(x = MonocytesIntegratedviaCluster@reductions$MonoRNA.umap@cell.embeddings[,1],
            y = MonocytesIntegratedviaCluster@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = MonocytesIntegratedviaCluster@meta.data$Cluster.3), size = .2, alpha = 0.8),
    dpi = 300) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  facet_wrap(~visitReal, nrow = 3) +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(MonocytesIntegratediaCluster@meta.data$Cluster.3))) +
  guides(color = guide_legend(override.aes = list(size = .2, alpha = 1)))

VisitCounts <- MonocytesIntegratedViaCluster@meta.data %>%
  dplyr::group_by(Cluster.3, visitReal, sample) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(visitReal) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(visitReal)

VisitCountsBox <- VisitCounts %>%
  ggplot(aes(x = Cluster.3,
             y = pctCells,
             fill = visitReal)) +
  geom_col(position = "fill") +
  scale_fill_manual(values = c("Gray", "Blue", "Orange")) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "Cluster",
       y = "% of cells",
       fill = "Visit", ) +
  theme(axis.text.x=element_text(angle=60, hjust=1)) + NoLegend()


pdf("plots/TimeUmap.pdf", height = 10, width = 6)
FacetVisitUMAP
dev.off()

pdf("plots/TimeBox.pdf", height = 6, width = 6)
VisitCountsBox
dev.off()

VisitUMAPCluster3 <- ggplot(
    data = MonocytesIntegratedviaCluster@meta.data,
    mapping =
        aes(x = MonocytesIntegratedviaCluster@reductions$umap_visit@cell.embeddings[,1],
            y = MonocytesIntegratedviaCluster@reductions$umap_visit@cell.embeddings[,2])) +
    ggrastr::rasterise(
        geom_point(mapping = aes_string(color = MonocytesIntegratedviaCluster@meta.data$Cluster.3), size = .4, alpha = 0.8),
        dpi = 300) +
    labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(MonocytesIntegratedviaCluster@meta.data$Cluster.3))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

VisitUMAPCluster3Facet <-  ggplot(
    data = MonocytesIntegratedviaCluster@meta.data,
    mapping =
        aes(x = MonocytesIntegratedviaCluster@reductions$umap_visit@cell.embeddings[,1],
            y = MonocytesIntegratedviaCluster@reductions$umap_visit@cell.embeddings[,2])) +
    ggrastr::rasterise(
        geom_point(mapping = aes_string(color = MonocytesIntegratedviaCluster@meta.data$Cluster.3), size = .4, alpha = 0.8),
        dpi = 300) +
    labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(MonocytesIntegratedviaCluster@meta.data$Cluster.3))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1))) + facet_wrap(~Cluster.3)

Linegraph <-list()
for(i in unique(lineGraphData$Cluster.3)){
  name <- i
  Linegraph[[name]] <- ggplot(lineGraphData %>% dplyr::filter(Cluster.3 == name) %>% dplyr::select(visitReal, sample, nCells), aes(x=visitReal, y=nCells, group=sample)) + geom_line() + geom_point() + scale_x_discrete(name = NULL, labels = c("Visit 1", "Visit 2")) + labs(title = paste0("Cluster ", name))
}

ggarrange(plotlist = Linegraph)

VisitLinegraph <- ggarrange(plotlist = Linegraph)

pdf("plots/FiguresJuly112022/LineGraphVisitnCell.pdf", width = 10, height =10)
VisitLinegraph
dev.off()

pdf("plots/FiguresJuly112022/FacetvisitUMAP.pdf", width = 6, height =6)
FacetVisitUMAP
dev.off()

pdf("plots/FiguresJuly112022/VisitBoxProportions.pdf", width = 6, height =8)
VisitCountsBox
dev.off()

pdf("plots/FiguresJuly112022/SupervisedUMAP_Visits.pdf", width = 6, height =6)
VisitUMAPCluster3
dev.off()

pdf("plots/FiguresJuly112022/SupervisedUMAP_Visits_Facet.pdf", width = 10, height =10)
VisitUMAPCluster3Facet
dev.off()

```

```{r cleanups}

frame30long <- data.frame()
for(i in 1:length(levels(MonocytesIntegratedviaCluster@meta.data$Cluster.3)) -1){
    cluster <- i
    temp <- Markers30[Markers30$cluster == cluster,][1:20,]
    frame30long <- rbind(frame30long, temp)
}

write.xlsx(frame30long, file = "tables/Cluster30MarkersLongform.xlsx")

HmpMatrix <- frame30long %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  #dplyr::slice_min(p_val_adj, n = 4) %>%
  #dplyr::slice_max(avg_log2FC, n = 4) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

selectedFeatures <- unique(HmpMatrix$gene)

sampleAverages <- AverageExpression(MonocytesIntegratedviaCluster, 
                                    assays = "RNA",
                                    features = selectedFeatures,
                                    group.by = "studyGroup",
                                    return.seurat = T) 

avgExprMat <- sampleAverages[["RNA"]]@data %>% as.matrix()

scaledreducedExprMat <- t(scale(t(reducedMatrix)))


sampleAnno <- rownames(sampleAverages@meta.data)

columnAnno <- rowAnnotation(B = sampleAnno,
                              col = list(B = sampleColors),
                              simple_anno_size = unit(0.25, "cm"),
                            show_legend = FALSE)

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf(file.path(plotDir, 
              "Heatmap_Integrated_TopClusterMarkers.pdf"),
    height = 4,
    width = 10)

Heatmap(t(scaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = FALSE,
        column_order = GeneOrder,
        show_column_dend = TRUE,
        row_order = c(sampleAnno[grep("v1", sampleAnno)], sampleAnno[grep("v2", sampleAnno)]),
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = FALSE,
        show_row_dend = TRUE,
        left_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())

studyAverage <- AverageExpression(MonocytesIntegratedviaCluster, 
                                    assays = "RNA",
                                    features = selectedFeatures,
                                    group.by = "studyGroup",
                                    return.seurat = T) 

reducedSampleAnno <- sampleAnno[-columns]

reducedColumnAnno <- rowAnnotation(B = reducedSampleAnno,
                              col = list(B = sampleColors),
                              simple_anno_size = unit(0.25, "cm"),
                            show_legend = FALSE)

pdf("plots/HeatmapB1B2Reduced.pdf", height = 8, width = 6)
Heatmap(t(scaledreducedExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = FALSE,
        column_order = GeneOrder,
        show_column_dend = TRUE,
        row_order = c(reducedSampleAnno[grep("v1", reducedSampleAnno)], reducedSampleAnno[grep("v2", reducedSampleAnno)]),
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = FALSE,
        show_row_dend = TRUE,
        left_annotation = reducedColumnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()

```

```{r subset visit1 ARDS umap}
monocytesVisit1 <- subset(x = MonocytesIntegratedviaCluster, idents = "v1")

Visit1OnlyARDS <- ggplot(
    data = monocytesVisit1@meta.data,
    mapping =
        aes(x = monocytesVisit1@reductions$MonoRNA.umap@cell.embeddings[,1],
            y = monocytesVisit1@reductions$MonoRNA.umap@cell.embeddings[,2])) +
    ggrastr::rasterise(
        geom_point(mapping = aes_string(color = monocytesVisit1@meta.data$studyGroupAdj), size = .4, alpha = 0.8),
        dpi = 300) +
    labs(x = "UMAP 1", y = "UMAP 2") +
    scale_color_manual(
        "Study Group", 
        values = studyGroupColors,
        labels = c("ARDS", "No ARDS")) +
    guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

pdf("plots/ARDSUMAP_visit1Only.pdf", height = 6, width = 6)
Visit1OnlyARDS
dev.off()
```
```{r Dotplot FB and average expression}

FB_Average_Exp <- AverageExpression(MonocytesIntegratedviaCluster, assays = "FB", return.seurat = FALSE, group.by = "Cluster.3")

FBTotalDotplot <-DotPlot(MonocytesIntegratedviaCluster, features = rownames(FB_Average_Exp$FB), group.by = "Cluster.3", assay = "FB") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 4))

pdf("plots/FB_Total_Dotplot.pdf", width = 10, height = 6)
FBTotalDotplot
dev.off()

reducedFBDotplot <- DotPlot(MonocytesIntegratedviaCluster, features = reducedFBList$Description, group.by = "Cluster.3", assay = "FB") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 4))

pdf("plots/FB_Reduced_Dotplot.pdf", width = 10, height = 6)
reducedFBDotplot
dev.off()

top5Markers <- DotPlot(MonocytesIntegratedviaCluster, features = unique(MarkerList), group.by = "Cluster.3", assay = "FB2") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 8)) + scale_x_discrete(labels = rownameReorder)

pdf("plots/FB_top5_Dotplot.pdf", width = 10, height = 6)
top5Markers
dev.off()

write.xlsx(FB_Average_Exp, file = "tables/Total_FB_Expressio_Matrix.xlsx")

```

```{r selected cell types modified}



seuratIntegrated@meta.data$celltypeAdj <- seuratIntegrated@meta.data$predicted.celltype.l1

seuratIntegrated@meta.data$celltypeAdj[(seuratIntegrated@reductions$umap@cell.embeddings[,2] > 10)] <- "pDC"

seuratIntegrated@meta.data$celltypeAdj[(seuratIntegrated@reductions$umap@cell.embeddings[,2] < -5 & seuratIntegrated@reductions$umap@cell.embeddings[,1] <0)] <- "Cycling"

celltypeAdjusted <- DimPlot(object = seuratIntegrated, 
                        reduction = "umap", 
                        group = "celltypeAdj",
                        label = FALSE,
                        repel = FALSE) +
           scale_color_manual(values = big_colorblind_pal(12)[-c(1,2)]) +
           labs(x = "UMAP 1", 
                y ="UMAP 2", 
                color = "",
                title = NULL) +
           theme(aspect.ratio=1)

pdf("plots/cellTypeUMAP_adjusted.pdf", width = 5, height = 5)
celltypeAdjusted
dev.off()

ARDSvsNoARDSUMAP <- ggplot(
data = MonocytesIntegratedviaCluster@meta.data,
mapping =
aes(x = MonocytesIntegratedviaCluster@reductions$MonoRNA.umap@cell.embeddings[,1],
y = MonocytesIntegratedviaCluster@reductions$MonoRNA.umap@cell.embeddings[,2])) +
ggrastr::rasterise(
geom_point(mapping = aes_string(color = MonocytesIntegratedviaCluster@meta.data$studyGroupAdj), size = .5, alpha = 0.8),
dpi = 300) +
labs(x = "UMAP 1", y = "UMAP 2") +
scale_color_manual(
"Study Group",
values = studyGroupColors,
labels = c("ARDS", "No ARDS")) +
guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

pdf("plots/Go-Term_DotChart.pdf", height = 6, width = 6)
dotchart(head(pvalueFigure$X.log.10.P.DE,8), labels = str_to_title(head(pvalueFigure$Term,10)), pch = 21, pt.cex = rev(sizes), main = "GO-terms of DE between ARDS and non-ARDS Monocytes", xlab = "-log10 P-value", bg = "limegreen",  groups = head(-pvalueFigure$X.log.10.P.DE,8), cex = .75)
legend(x = "bottomright", legend=c("50", "100", "150"), cex = 1.4, title= "Number of DE Genes", text.font = 4, pch = 21, pt.cex = c(1, 2, 3), bg = "white")
dev.off()

MonocytesIntegratedviaCluster <-RunUMAP(MonocytesIntegratedviaCluster, reduction = "FBBestPCA", dims = 1:30, reduction.name = "FBBestUMAP")

FeaturePlot(MonocytesIntegratedviaCluster, feature = "CD14.1", reduction = "FBBestUMAP", pt.size = 1.5) + ggplot(
data = MonocytesIntegratedviaCluster@meta.data,
mapping =
aes(x = MonocytesIntegratedviaCluster@reductions$FBBestUMAP@cell.embeddings[,1],
y = MonocytesIntegratedviaCluster@reductions$FBBestUMAP@cell.embeddings[,2])) +
ggrastr::rasterise(
geom_point(mapping = aes_string(color = MonocytesIntegratedviaCluster@meta.data$Cluster.3), size = .5, alpha = 0.8),
dpi = 300) +
labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(MonocytesIntegratedviaCluster@meta.data$Cluster.3))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))
```

```{r density plots}

CD163Full <- list()
for(cluster in 0:8){
  x <- cluster
  y <- cluster + 1
  CD163Full[y] <- as.data.frame(MonocytesIntegratedviaCluster@assays$FB2@data["CD163.1", MonocytesIntegratedviaCluster@meta.data$Cluster.3 == x])
}

CD163Overlay <- ggplot() + geom_density(aes(CD163), color = clusterColors[1], data = CD163C0) + geom_density(aes(CD163), color = clusterColors[2], data = CD163C1) + geom_density(aes(CD163), color = clusterColors[3], data = CD163C2) + geom_density(aes(CD163), color = clusterColors[4], data = CD163C3) + geom_density(aes(CD163), color = clusterColors[5], data = CD163C4) + geom_density(aes(CD163), color = clusterColors[6], data = CD163C5) + geom_density(aes(CD163), color = clusterColors[7], data = CD163C6) + geom_density(aes(CD163), color = clusterColors[8], data = CD163C7) + geom_density(aes(CD163), color = clusterColors[9], data = CD163C8) + labs(title = "CD163", x = "Expression")


PDL1Full <- list()
for(cluster in 0:8){
  x <- cluster
  y <- cluster + 1
  PDL1Full[y] <- as.data.frame(MonocytesIntegratedviaCluster@assays$FB2@data["PDL1", MonocytesIntegratedviaCluster@meta.data$Cluster.3 == x])
}

for(x in 1:9){
  cluster <- x -1
  obj <- paste0("PDL", get("cluster"))
  assign(obj, as.data.frame(PDL1Full[x]))
}
colnames(PDL0) <- "PDL1"
colnames(PDL1) <- "PDL1"
colnames(PDL2) <- "PDL1"
colnames(PDL3) <- "PDL1"
colnames(PDL4) <- "PDL1"
colnames(PDL5) <- "PDL1"
colnames(PDL6) <- "PDL1"
colnames(PDL7) <- "PDL1"
colnames(PDL8) <- "PDL1"



PDL1Overlay <- ggplot() + geom_density(aes(PDL1), color = clusterColors[1], data = PDL0) + geom_density(aes(PDL1), color = clusterColors[2], data = PDL1) + geom_density(aes(PDL1), color = clusterColors[3], data = PDL2) + geom_density(aes(PDL1), color = clusterColors[4], data = PDL3) + geom_density(aes(PDL1), color = clusterColors[5], data = PDL4) + geom_density(aes(PDL1), color = clusterColors[6], data = PDL5) + geom_density(aes(PDL1), color = clusterColors[7], data = PDL6) + geom_density(aes(PDL1), color = clusterColors[8], data = PDL7) + geom_density(aes(PDL1), color = clusterColors[9], data = PDL8) + labs(title = "PDL1", x = "Expression")

LOX1Full <- list()
for(cluster in 0:8){
  x <- cluster
  y <- cluster + 1
  LOX1Full[y] <- as.data.frame(MonocytesIntegratedviaCluster@assays$FB2@data["LOX-1", MonocytesIntegratedviaCluster@meta.data$Cluster.3 == x])
}

objlist <- c()
for(x in 1:9){
  cluster <- x -1
  obj <- paste0("LOX1C", get("cluster"))
  objlist <- append(objlist, paste(obj))
  assign(obj, as.data.frame(LOX1Full[x]))
}

colnames(LOX1C0) <- "LOX-1"
colnames(LOX1C1) <- "LOX-1"
colnames(LOX1C2) <- "LOX-1"
colnames(LOX1C3) <- "LOX-1"
colnames(LOX1C4) <- "LOX-1"
colnames(LOX1C5) <- "LOX-1"
colnames(LOX1C6) <- "LOX-1"
colnames(LOX1C7) <- "LOX-1"
colnames(LOX1C8) <- "LOX-1"

LOX1Overlay <- ggplot() + geom_density(aes(`LOX-1`), color = clusterColors[1], data = LOX1C0) + geom_density(aes(`LOX-1`), color = clusterColors[2], data = LOX1C1) + geom_density(aes(`LOX-1`), color = clusterColors[3], data = LOX1C2) + geom_density(aes(`LOX-1`), color = clusterColors[4], data = LOX1C3) + geom_density(aes(`LOX-1`), color = clusterColors[5], data = LOX1C4) + geom_density(aes(`LOX-1`), color = clusterColors[6], data = LOX1C5) + geom_density(aes(`LOX-1`), color = clusterColors[7], data = LOX1C6) + geom_density(aes(`LOX-1`), color = clusterColors[8], data = LOX1C7) + geom_density(aes(`LOX-1`), color = clusterColors[9], data = LOX1C8) + labs(title = "LOX-1", x = "Expression")

CD123Full <- list()
for(cluster in 0:8){
  x <- cluster
  y <- cluster + 1
  CD123Full[y] <- as.data.frame(MonocytesIntegratedviaCluster@assays$FB2@data["CD123", MonocytesIntegratedviaCluster@meta.data$Cluster.3 == x])
}

objlist <- c()
for(x in 1:9){
  cluster <- x -1
  obj <- paste0("CD123C", get("cluster"))
  objlist <- append(objlist, paste(obj))
  assign(obj, as.data.frame(CD123Full[x]))
}

colnames(CD123C0) <- "CD123"
colnames(CD123C1) <- "CD123"
colnames(CD123C2) <- "CD123"
colnames(CD123C3) <- "CD123"
colnames(CD123C4) <- "CD123"
colnames(CD123C5) <- "CD123"
colnames(CD123C6) <- "CD123"
colnames(CD123C7) <- "CD123"
colnames(CD123C8) <- "CD123"

CD123Overlay <- ggplot() + geom_density(aes(CD123), color = clusterColors[1], data = CD123C0) + geom_density(aes(CD123), color = clusterColors[2], data = CD123C1) + geom_density(aes(CD123), color = clusterColors[3], data = CD123C2) + geom_density(aes(CD123), color = clusterColors[4], data = CD123C3) + geom_density(aes(CD123), color = clusterColors[5], data = CD123C4) + geom_density(aes(CD123), color = clusterColors[6], data = CD123C5) + geom_density(aes(CD123), color = clusterColors[7], data = CD123C6) + geom_density(aes(CD123), color = clusterColors[8], data = CD123C7) + geom_density(aes(CD123), color = clusterColors[9], data = CD123C8) + labs(title = "CD123", x = "Expression")

CD35Full <- list()
for(cluster in 0:8){
  x <- cluster
  y <- cluster + 1
  CD35Full[y] <- as.data.frame(MonocytesIntegratedviaCluster@assays$FB2@data["CD35", MonocytesIntegratedviaCluster@meta.data$Cluster.3 == x])
}

objlist <- c()
for(x in 1:9){
  cluster <- x -1
  obj <- paste0("CD35C", get("cluster"))
  objlist <- append(objlist, paste(obj))
  assign(obj, as.data.frame(CD35Full[x]))
}

colnames(CD35C0) <- "CD35"
colnames(CD35C1) <- "CD35"
colnames(CD35C2) <- "CD35"
colnames(CD35C3) <- "CD35"
colnames(CD35C4) <- "CD35"
colnames(CD35C5) <- "CD35"
colnames(CD35C6) <- "CD35"
colnames(CD35C7) <- "CD35"
colnames(CD35C8) <- "CD35"

CD35Overlay <- ggplot() + geom_density(aes(CD35), color = clusterColors[1], data = CD35C0) + geom_density(aes(CD35), color = clusterColors[2], data = CD35C1) + geom_density(aes(CD35), color = clusterColors[3], data = CD35C2) + geom_density(aes(CD35), color = clusterColors[4], data = CD35C3) + geom_density(aes(CD35), color = clusterColors[5], data = CD35C4) + geom_density(aes(CD35), color = clusterColors[6], data = CD35C5) + geom_density(aes(CD35), color = clusterColors[7], data = CD35C6) + geom_density(aes(CD35), color = clusterColors[8], data = CD35C7) + geom_density(aes(CD35), color = clusterColors[9], data = CD35C8) + labs(title = "CD35", x = "Expression")

CD86.1Full <- list()
for(cluster in 0:8){
  x <- cluster
  y <- cluster + 1
  CD86.1Full[y] <- as.data.frame(MonocytesIntegratedviaCluster@assays$FB2@data["CD86.1", MonocytesIntegratedviaCluster@meta.data$Cluster.3 == x])
}

objlist <- c()
for(x in 1:9){
  cluster <- x -1
  obj <- paste0("CD86.1C", get("cluster"))
  objlist <- append(objlist, paste(obj))
  assign(obj, as.data.frame(CD86.1Full[x]))
}

colnames(CD86.1C0) <- "CD86.1"
colnames(CD86.1C1) <- "CD86.1"
colnames(CD86.1C2) <- "CD86.1"
colnames(CD86.1C3) <- "CD86.1"
colnames(CD86.1C4) <- "CD86.1"
colnames(CD86.1C5) <- "CD86.1"
colnames(CD86.1C6) <- "CD86.1"
colnames(CD86.1C7) <- "CD86.1"
colnames(CD86.1C8) <- "CD86.1"

CD86Overlay <- ggplot() + geom_density(aes(CD86.1), color = clusterColors[1], data = CD86.1C0) + geom_density(aes(CD86.1), color = clusterColors[2], data = CD86.1C1) + geom_density(aes(CD86.1), color = clusterColors[3], data = CD86.1C2) + geom_density(aes(CD86.1), color = clusterColors[4], data = CD86.1C3) + geom_density(aes(CD86.1), color = clusterColors[5], data = CD86.1C4) + geom_density(aes(CD86.1), color = clusterColors[6], data = CD86.1C5) + geom_density(aes(CD86.1), color = clusterColors[7], data = CD86.1C6) + geom_density(aes(CD86.1), color = clusterColors[8], data = CD86.1C7) + geom_density(aes(CD86.1), color = clusterColors[9], data = CD86.1C8) + labs(title = "CD86", x = "Expression")

CD1cFull <- list()
for(cluster in 0:8){
  x <- cluster
  y <- cluster + 1
  CD1cFull[y] <- as.data.frame(MonocytesIntegratedviaCluster@assays$FB2@data["CD1c", MonocytesIntegratedviaCluster@meta.data$Cluster.3 == x])
}

objlist <- c()
for(x in 1:9){
  cluster <- x -1
  obj <- paste0("CD1cC", get("cluster"))
  objlist <- append(objlist, paste(obj))
  assign(obj, as.data.frame(CD1cFull[x]))
}

colnames(CD1cC0) <- "CD1c"
colnames(CD1cC1) <- "CD1c"
colnames(CD1cC2) <- "CD1c"
colnames(CD1cC3) <- "CD1c"
colnames(CD1cC4) <- "CD1c"
colnames(CD1cC5) <- "CD1c"
colnames(CD1cC6) <- "CD1c"
colnames(CD1cC7) <- "CD1c"
colnames(CD1cC8) <- "CD1c"

CD1cOverlay <- ggplot() + geom_density(aes(CD1c), color = clusterColors[1], data = CD1cC0) + geom_density(aes(CD1c), color = clusterColors[2], data = CD1cC1) + geom_density(aes(CD1c), color = clusterColors[3], data = CD1cC2) + geom_density(aes(CD1c), color = clusterColors[4], data = CD1cC3) + geom_density(aes(CD1c), color = clusterColors[5], data = CD1cC4) + geom_density(aes(CD1c), color = clusterColors[6], data = CD1cC5) + geom_density(aes(CD1c), color = clusterColors[7], data = CD1cC6) + geom_density(aes(CD1c), color = clusterColors[8], data = CD1cC7) + geom_density(aes(CD1c), color = clusterColors[9], data = CD1cC8) + labs(title = "CD1c", x = "Expression")

CD44.1Full <- list()
for(cluster in 0:8){
  x <- cluster
  y <- cluster + 1
  CD44.1Full[y] <- as.data.frame(MonocytesIntegratedviaCluster@assays$FB2@data["CD44.1", MonocytesIntegratedviaCluster@meta.data$Cluster.3 == x])
}

objlist <- c()
for(x in 1:9){
  cluster <- x -1
  obj <- paste0("CD44.1C", get("cluster"))
  objlist <- append(objlist, paste(obj))
  assign(obj, as.data.frame(CD44.1Full[x]))
}

colnames(CD44.1C0) <- "CD44.1"
colnames(CD44.1C1) <- "CD44.1"
colnames(CD44.1C2) <- "CD44.1"
colnames(CD44.1C3) <- "CD44.1"
colnames(CD44.1C4) <- "CD44.1"
colnames(CD44.1C5) <- "CD44.1"
colnames(CD44.1C6) <- "CD44.1"
colnames(CD44.1C7) <- "CD44.1"
colnames(CD44.1C8) <- "CD44.1"

CD44.1Overlay <- ggplot() + geom_density(aes(CD44.1), color = clusterColors[1], data = CD44.1C0) + geom_density(aes(CD44.1), color = clusterColors[2], data = CD44.1C1) + geom_density(aes(CD44.1), color = clusterColors[3], data = CD44.1C2) + geom_density(aes(CD44.1), color = clusterColors[4], data = CD44.1C3) + geom_density(aes(CD44.1), color = clusterColors[5], data = CD44.1C4) + geom_density(aes(CD44.1), color = clusterColors[6], data = CD44.1C5) + geom_density(aes(CD44.1), color = clusterColors[7], data = CD44.1C6) + geom_density(aes(CD44.1), color = clusterColors[8], data = CD44.1C7) + geom_density(aes(CD44.1), color = clusterColors[9], data = CD44.1C8) + labs(title = "CD441", x = "Expression")

CD1cFull <- list()
for(cluster in 0:8){
  x <- cluster
  y <- cluster + 1
  CD1cFull[y] <- as.data.frame(MonocytesIntegratedviaCluster@assays$FB2@data["CD1c", MonocytesIntegratedviaCluster@meta.data$Cluster.3 == x])
}

objlist <- c()
for(x in 1:9){
  cluster <- x -1
  obj <- paste0("CD1cC", get("cluster"))
  objlist <- append(objlist, paste(obj))
  assign(obj, as.data.frame(CD1cFull[x]))
}

colnames(CD1cC0) <- "CD1c"
colnames(CD1cC1) <- "CD1c"
colnames(CD1cC2) <- "CD1c"
colnames(CD1cC3) <- "CD1c"
colnames(CD1cC4) <- "CD1c"
colnames(CD1cC5) <- "CD1c"
colnames(CD1cC6) <- "CD1c"
colnames(CD1cC7) <- "CD1c"
colnames(CD1cC8) <- "CD1c"

CD1cOverlay <- ggplot() + geom_density(aes(CD1c), color = clusterColors[1], data = CD1cC0) + geom_density(aes(CD1c), color = clusterColors[2], data = CD1cC1) + geom_density(aes(CD1c), color = clusterColors[3], data = CD1cC2) + geom_density(aes(CD1c), color = clusterColors[4], data = CD1cC3) + geom_density(aes(CD1c), color = clusterColors[5], data = CD1cC4) + geom_density(aes(CD1c), color = clusterColors[6], data = CD1cC5) + geom_density(aes(CD1c), color = clusterColors[7], data = CD1cC6) + geom_density(aes(CD1c), color = clusterColors[8], data = CD1cC7) + geom_density(aes(CD1c), color = clusterColors[9], data = CD1cC8) + labs(title = "CD1c", x = "Expression")

CD71Full <- list()
for(cluster in 0:8){
  x <- cluster
  y <- cluster + 1
  CD71Full[y] <- as.data.frame(MonocytesIntegratedviaCluster@assays$FB2@data["CD71", MonocytesIntegratedviaCluster@meta.data$Cluster.3 == x])
}

objlist <- c()
for(x in 1:9){
  cluster <- x -1
  obj <- paste0("CD71C", get("cluster"))
  objlist <- append(objlist, paste(obj))
  assign(obj, as.data.frame(CD71Full[x]))
}

colnames(CD71C0) <- "CD71"
colnames(CD71C1) <- "CD71"
colnames(CD71C2) <- "CD71"
colnames(CD71C3) <- "CD71"
colnames(CD71C4) <- "CD71"
colnames(CD71C5) <- "CD71"
colnames(CD71C6) <- "CD71"
colnames(CD71C7) <- "CD71"
colnames(CD71C8) <- "CD71"

CD71Overlay <- ggplot() + geom_density(aes(CD71), color = clusterColors[1], data = CD71C0) + geom_density(aes(CD71), color = clusterColors[2], data = CD71C1) + geom_density(aes(CD71), color = clusterColors[3], data = CD71C2) + geom_density(aes(CD71), color = clusterColors[4], data = CD71C3) + geom_density(aes(CD71), color = clusterColors[5], data = CD71C4) + geom_density(aes(CD71), color = clusterColors[6], data = CD71C5) + geom_density(aes(CD71), color = clusterColors[7], data = CD71C6) + geom_density(aes(CD71), color = clusterColors[8], data = CD71C7) + geom_density(aes(CD71), color = clusterColors[9], data = CD71C8) + labs(title = "CD71", x = "Expression")

pdf("/Users/mlawrance/Library/CloudStorage/Box-Box/P458_SLICC_Mikacenic/plots/FBExpressionFacets.pdf", height = 7, width = 7)
gridExtra::grid.arrange(CD163Overlay, LOX1Overlay, CD123Overlay, CD71Overlay, CD35Overlay, CD86Overlay, CD1cOverlay, CD44.1Overlay, PDL1Overlay)
dev.off()

CD14.1Full <- list()
for(cluster in 0:8){
  x <- cluster
  y <- cluster + 1
  CD14.1Full[y] <- as.data.frame(MonocytesIntegratedviaCluster@assays$FB2@data["CD14.1", MonocytesIntegratedviaCluster@meta.data$Cluster.3 == x])
}

objlist <- c()
for(x in 1:9){
  cluster <- x -1
  obj <- paste0("CD14.1C", get("cluster"))
  objlist <- append(objlist, paste(obj))
  assign(obj, as.data.frame(CD14.1Full[x]))
}

colnames(CD14.1C0) <- "CD14.1"
colnames(CD14.1C1) <- "CD14.1"
colnames(CD14.1C2) <- "CD14.1"
colnames(CD14.1C3) <- "CD14.1"
colnames(CD14.1C4) <- "CD14.1"
colnames(CD14.1C5) <- "CD14.1"
colnames(CD14.1C6) <- "CD14.1"
colnames(CD14.1C7) <- "CD14.1"
colnames(CD14.1C8) <- "CD14.1"

CD14.1Overlay <- ggplot() + geom_density(aes(CD14.1), color = clusterColors[1], data = CD14.1C0) + geom_density(aes(CD14.1), color = clusterColors[2], data = CD14.1C1) + geom_density(aes(CD14.1), color = clusterColors[3], data = CD14.1C2) + geom_density(aes(CD14.1), color = clusterColors[4], data = CD14.1C3) + geom_density(aes(CD14.1), color = clusterColors[5], data = CD14.1C4) + geom_density(aes(CD14.1), color = clusterColors[6], data = CD14.1C5) + geom_density(aes(CD14.1), color = clusterColors[7], data = CD14.1C6) + geom_density(aes(CD14.1), color = clusterColors[8], data = CD14.1C7) + geom_density(aes(CD14.1), color = clusterColors[9], data = CD14.1C8) + labs(title = "CD14 Expression by Cluster", x = "Expression")
```

```{r visitHeatmap reversed}

visitMarkersHmapDown <- visitMarkers %>%
  #dplyr::group_by(sample) %>%
  dplyr::filter(avg_log2FC < 0) %>%
  dplyr::slice_min(p_val_adj, n = 30) %>%
  dplyr::slice_max(avg_log2FC, n = 30) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

visitFeaturesDown <- unique(rownames(visitMarkersHmapDown))



visitAveragesDown <- AverageExpression(seuratVisit, 
                                    features = visitFeaturesDown,
                                    assays = 'RNA',
                                    group.by = "sample",
                                    use.scale = TRUE,
                                    return.seurat = T) 

visitAvgExprMatDown <- visitAveragesDown[["RNA"]]@data %>% as.matrix()

visitScaledExprMatDown <- t(scale(t(visitAvgExprMatDown)))


visitClusterAnnoDown <- rownames(visitAveragesDown@meta.data)

visitColumnAnnoDown <- rowAnnotation(samples = visitClusterAnnoDown,
                              col =  list(samples = c("SLK12v1" = "blue",
                                                      "SLK12v2" = "orange",
                                                      "SLK21v1" = "blue",
                                                      "SLK21v2" = "orange",
                                                      "SLK23v1" = "blue",
                                                      "SLK23v2" = "orange",
                                                      "SLK7v1" = "blue",
                                                      "SLK7v2" = "orange")),
                              simple_anno_size = unit(0.25, "cm"), 
                              show_legend = FALSE,
                              show_annotation_name = FALSE)

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf(file.path(plotDir, 
              "Heatmap_Integrated_visitShift_Extended_Down.pdf"),
    height = 6,
    width = 4)

Heatmap(t(visitScaledExprMatDown), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = FALSE,
        row_order = c("SLK12v2", "SLK7v2", "SLK23v2", "SLK21v2", "SLK21v1", "SLK12v1", "SLK23v1", "SLK7v1"),
        show_row_dend = TRUE,
        left_annotation = visitColumnAnnoDown,
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())
```

```{r MA plot for visit 1 and visit 2 separately}

Idents(MonocytesIntegratedviaCluster) <- "visitReal"
visit1subset <- subset(MonocytesIntegratedviaCluster, idents = c("Visit_1"))
visit2subset <- subset(MonocytesIntegratedviaCluster, idents = c("Visit_2"))

Idents(visit1subset) <- "studyGroup"

Idents(visit2subset) <- "studyGroup"

visit1MeanExp <- FindMarkers(visit1subset, ident.1 = "ARDS", ident.2 = "No ARDS", assay = "RNA", logfc.threshold = 0)
avgExpv1 <- as.data.frame(rowMeans(visit1subset@assays$RNA@data[rownames(visit1MeanExp),]))
colnames(avgExpv1) <- "averageExpressionOverall"

visit1MeanExp$gene <- rownames(visit1MeanExp)
avgExpv1$gene <- rownames(avgExpv1)
MeanExpMatrixv1 <- left_join(visit1MeanExp, avgExpv1, by = "gene")
MeanExpMatrixv1$mycolor1 <- "black"
MeanExpMatrixv1$mycolor1[MeanExpMatrixv1$avg_log2FC > .25] <- "red"
MeanExpMatrixv1$mycolor1[MeanExpMatrixv1$avg_log2FC < -.25] <- "darkcyan"

subsetMeanExpMatrixv1 <- MeanExpMatrixv1 %>% dplyr::arrange(avg_log2FC) %>% tail(., n = 10)
subsetMeanExpMatrixv1 <- MeanExpMatrixv1 %>% dplyr::arrange(avg_log2FC) %>% head(., n = 10) %>% rbind(., subsetMeanExpMatrixv1)

ARDSMarkersExpressionV1 <- ggplot(data = MeanExpMatrixv1, aes(x=averageExpressionOverall, y=avg_log2FC, color = mycolor1)) +
    geom_point(alpha=0.7, size=1, shape = 19) + 
  scale_color_manual(name = NULL, 
                     values = c('black' = 'gray', 'red' = 'red', 'darkcyan' = 'darkcyan'), 
                     labels = NULL, 
                     breaks = NULL) + 
  geom_text_repel(data=subsetMeanExpMatrixv1, label = subsetMeanExpMatrixv1$gene, max.overlaps = Inf, min.segment.length = unit(0, 'lines'), size = 3, col = "black") +
        labs(x = "Overall Expression",
         y = "Log 2 Fold Change",
         title = "ARDS Markers Visit 1") +
    theme(plot.title = element_text(hjust = .5)) +
    NoLegend()

visit2MeanExp <- FindMarkers(visit2subset, ident.1 = "ARDS", ident.2 = "No ARDS", assay = "RNA", logfc.threshold = 0)
avgExpv2 <- as.data.frame(rowMeans(visit2subset$RNA@data[rownames(visit2MeanExp),]))
colnames(avgExpv2) <- "averageExpressionOverall"

visit2MeanExp$gene <- rownames(visit2MeanExp)
avgExpv2$gene <- rownames(avgExpv2)
MeanExpMatrixv2 <- left_join(visit2MeanExp, avgExpv2, by = "gene")
MeanExpMatrixv2$mycolor1 <- "black"
MeanExpMatrixv2$mycolor1[MeanExpMatrixv2$avg_log2FC > .25] <- "red"
MeanExpMatrixv2$mycolor1[MeanExpMatrixv2$avg_log2FC < -.25] <- "darkcyan"

subsetMeanExpMatrixv2 <- MeanExpMatrixv2 %>% dplyr::arrange(avg_log2FC) %>% tail(., n = 10)
subsetMeanExpMatrixv2 <- MeanExpMatrixv2 %>% dplyr::arrange(avg_log2FC) %>% head(., n = 10) %>% rbind(., subsetMeanExpMatrixv2)

ARDSMarkersExpressionV2 <- ggplot(data = MeanExpMatrixv2, aes(x=averageExpressionOverall, y=avg_log2FC, color = mycolor1)) +
    geom_point(alpha=0.7, size=1, shape = 19) + 
  scale_color_manual(name = NULL, 
                     values = c('black' = 'gray', 'red' = 'red', 'darkcyan' = 'darkcyan'), 
                     labels = NULL, 
                     breaks = NULL) + 
  geom_text_repel(data=subsetMeanExpMatrixv2, label = subsetMeanExpMatrixv2$gene, max.overlaps = Inf, min.segment.length = unit(0, 'lines'), size = 3, col = "black") +
        labs(x = "Overall Expression",
         y = "Log 2 Fold Change",
         title = "ARDS Markers Visit 2") +
    theme(plot.title = element_text(hjust = .5)) +
    NoLegend()

pdf("plots/ARDSMarkersMAByVisit.pdf",
    width = 12,
    height = 10)
grid.arrange(ARDSMarkersExpressionV1, ARDSMarkersExpressionV2, ncol = 1)
dev.off()

```

```{r}

Idents(MonocytesIntegratedviaCluster) <- "studyGroupVisit"

ArdsTimeMarkers <- FindMarkers(MonocytesIntegratedviaCluster, ident.1 = "v2ARDS", ident.2 = "v1ARDS", assay = "RNA")

NoArdsTimeMarkers <- FindMarkers(MonocytesIntegratedviaCluster, ident.1 = "v2No_ARDS", ident.2 = "v1No_ARDS", assay = "RNA")

ChangeArdsOnly <- ArdsTimeMarkers[!rownames(ArdsTimeMarkers) %in% rownames(NoArdsTimeMarkers),]
ChangeNoArdsOnly <- NoArdsTimeMarkers[!rownames(NoArdsTimeMarkers) %in% rownames(ArdsTimeMarkers),]
ChangeBothNoArds <- NoArdsTimeMarkers[rownames(NoArdsTimeMarkers) %in% rownames(ArdsTimeMarkers),]
ChangeBothArds <- ArdsTimeMarkers[rownames(ArdsTimeMarkers) %in% rownames(NoArdsTimeMarkers),]

write.xlsx(ChangeArdsOnly, "tables/TimeSeriesArdsNoArds.xlsx", sheetName = "Ards Only")
write.xlsx(ChangeNoArdsOnly, "tables/TimeSeriesArdsNoArds.xlsx", sheetName = "No Ards Only", append = TRUE)
write.xlsx(ChangeBothNoArds, "tables/TimeSeriesArdsNoArds.xlsx", sheetName = "Both, No Ards values", append = TRUE)
write.xlsx(ChangeBothArds, "tables/TimeSeriesArdsNoArds.xlsx", sheetName = "Both, Ards values", append= TRUE)

```

```{r cell type by sample}

sampleCellType <- seuratIntegrated@meta.data %>%
  dplyr::group_by(sample, celltypeAdj, studyGroup) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(sample) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(sample)

gsampleCellType <- sampleCellType %>%
  ggplot(aes(x = sample,
             y = pctCells,
             color = studyGroup,
             fill = celltypeAdj)) +
  geom_col(position = "fill") +
  scale_color_manual(values = studyGroupColors) + 
  scale_fill_manual(values = big_colorblind_pal(12)[-c(1,2)]) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "Sample",
       y = "% of cells",
       color = "Study group",
       fill = "Cell Type") +
  theme(axis.text.x=element_text(angle=60, hjust=1))

pdf("plots/cellTypeBySample.pdf", width = 10, height = 6)
  print(gsampleCellType)
dev.off()

factor(sampleCellType$sample,
                  levels = c("B", "D", "E", "C", "A"))

cluster30Umap <- DimPlot(MonocytesIntegratedviaCluster, 
        cols = big_colorblind_pal(9), 
        reduction = "MonoRNA.umap") +
        labs(x = "UMAP 1", 
                y ="UMAP 2", 
                color = "",
                title = NULL) +
           theme(aspect.ratio=1)

```

```{r ARDS no ARDS UMAP}

studyGroupFacetWrapUMAP <- ggplot(
    data = MonocytesIntegratedviaCluster@meta.data,
    mapping =
        aes(x = MonocytesIntegratedviaCluster@reductions$MonoRNA.umap@cell.embeddings[,1],
            y = MonocytesIntegratedviaCluster@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = MonocytesIntegratedviaCluster@meta.data$Cluster.3), size = .2, alpha = 0.8),
    dpi = 300) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  facet_wrap(~studyGroup, ncol = 2) +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(MonocytesIntegratedviaCluster@meta.data$Cluster.3))) +
  guides(color = guide_legend(override.aes = list(size = .2, alpha = 1)))

pdf(file = "plots/studyGroupFacetWrapUMAP.pdf", height = 5, width = 10)
studyGroupFacetWrapUMAP
dev.off()
```

```{r heatmap B1 only ARDS markers}

ARDSB1Markers <- FindMarkers(monocytesVisit1, ident.1 = "ARDS", ident.2 = "No ARDS", assay = "RNA")

ardsMarkersHmap <- ARDSB1Markers %>%
  #dplyr::group_by("studyGroup") %>%
  #dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 100) %>%
  dplyr::slice_max(avg_log2FC, n = 100) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

ardsFeatures <- rownames(ardsMarkersHmap)



ardsAverages <- AverageExpression(monocytesVisit1, 
                                    features = ardsFeatures,
                                    assays = 'RNA',
                                    group.by = "sample",
                                    use.scale = TRUE,
                                    return.seurat = T) 

ardsAvgExprMat <- ardsAverages[["RNA"]]@data %>% as.matrix()

ardsScaledExprMat <- t(scale(t(ardsAvgExprMat)))


ardsClusterAnno <- rownames(ardsAverages@meta.data)

ardsColumnAnno <- rowAnnotation(Condition = ardsClusterAnno,
                              col =  list(Condition = c("SLK8v1" = "darkcyan",
                                                     "SLK7v1" = "red",
                                                     "SLK12v1" = "darkcyan",
                                                     "SLK14v1" = "darkcyan",
                                                     "SLK21v1" = "darkcyan",
                                                     "SLK22v1" = "red",
                                                     "SLK23v1" = "red",
                                                     "SLK3v1" = "red")),
                              simple_anno_size = unit(0.25, "cm"),
                              show_legend = FALSE)

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf(file.path(plotDir, 
              "Heatmap_ARDS_visit1_Adj_Narrow.pdf"),
    height = 4,
    width = 7)

        Heatmap(t(ardsScaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = FALSE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = FALSE,
        left_annotation = ardsColumnAnno,
        height = nrow(t(visitScaledExprMat))*unit(5, "mm"),
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())
```

```{r p458-2}

project <- "P458-2"

 libids <- getProjectLibs(project, searchType = "regex")
 
anno <- getAnno(libids)

 save(anno,
      file = file.path(annoDir,"P458-2Anno.Rdata"))

```
## Load 10x data matrix


```{r loadData10x, dependson="getProjectInfo"}
filenameP458_2_Unfiltered <-
  file.path(dataDir, "P458_2_Unfiltered.RDS")

if (file.exists(filenameP458_2_Unfiltered)) {
  data10xUnfiltered <- readRDS(filenameP458_2_Unfiltered)
} else {
  # set counts filename
  data10xUnfiltered <-
    file.path(
      "/Volumes", "Bioinformatics", "pipeline", "tenxSummaries",
      "221101-P458-2allFlowcells", "aggregated",
      paste0("P458-2_aggr_filtered", "feature_bc_matrix.h5"))
  
  inputData10x <-
    Seurat::Read10X_h5("/Volumes/Bioinformatics/pipeline/tenxSummaries/221101-P458-2allFlowcells/aggregated/P458-2_aggr_filtered_feature_bc_matrix.h5")
  
  # filter gene expression data to include only protein-coding genes
  biotypeRegex.tmp <-
    "(protein_coding)"
  inputData10x$`Gene Expression` <-
    inputData10x$`Gene Expression`[
      rownames(inputData10x$`Gene Expression`) %in% 
        (annotables::grch38 %>%
           dplyr::filter(str_detect(biotype, biotypeRegex.tmp)) %>%
           dplyr::pull(symbol)),]
  
  # create Seurat object with RNA and antibody-derived tag data
  P458_2_10xUnfiltered <-
    CreateSeuratObject(counts = inputData10x$`Gene Expression`, )
  P458_2_10xUnfiltered[["ADT"]] <- CreateAssayObject(counts = inputData10x$`Antibody Capture`)
  
  # also store the cellBarcode information, in case it gets lost from the rownames of the meta.data slot
  P458_2_10xUnfiltered <- P458_2_10xUnfiltered %>%
    AddMetaData(
      metadata = colnames(P458_2_10xUnfiltered),
      col.name = "cellBarcode")

  saveRDS(P458_2_10xUnfiltered, file = filenameP458_2_Unfiltered)
  
  rm(inputData10x)
}

rm_tmp(ask=FALSE)
```
```{add sample identities}

sampleNames <- c("10_SLK21-V2","11_SLK23-V1","12_SLK23-V2","1_SLK3","2_SLK8","3_SLK14","4_SLK22","5_SLK7-V1","6_SLK7-V2","7_SLK12-V1","8_SLK12-V2","9_SLK21-V1")

P458_2_10xUnfiltered@meta.data$sample <- "NA"
for(i in 1:12){
  name <- sampleNames[i]
  number <- i
  P458_2_10xUnfiltered@meta.data$sample[grep(paste0("-", number, "$"),rownames(P458_2_10xUnfiltered@meta.data))] <- as.character(name)
  
}

```{r loadGeneSets, dependson="setFilenamesAndGlobalVariables"}
```

# Quality control of cells using RNAseq data

## Quality metrics

We use several metrics to identify high-quality cells:
1) Total UMI counts in each cell
2) Percent mitochondrial reads for each cell
3) Percent ribosomal reads for each cell

```{r calculateQcMetrics, dependson="loadData10x"}
P458_2_10xUnfiltered <- 
  PercentageFeatureSet(P458_2_10xUnfiltered, "^MT-", col.name = "percent_mito") # percent mitochondrial reads
P458_2_10xUnfiltered <-
  PercentageFeatureSet(P458_2_10xUnfiltered, "^RP[SL]", col.name = "percent_ribo") # percent ribosomal protein reads
P458_2_10xUnfiltered <-
  PercentageFeatureSet(P458_2_10xUnfiltered, "^HB[^(P)]", col.name = "percent_hb") # percent hemoglobin reads
```

```{r setQcMetricsThresholds}
qcMetricsThresholds <-
  c(
    "min_nFeature_RNA" = 250,
    "max_nFeature_RNA" = 6000,
    "min_nCount_RNA" = 450,
    "max_nCount_RNA" = 35000,
    # "nFeature_ADT" = 50,
    # "nCount_ADT" = 500,
    "percent_mito" = 12.5,
    "percent_ribo" = 50,
    "percent_hb" = 10
  )

qcMetricsDfForPlot <-
  data.frame(
    metric = str_replace(names(qcMetricsThresholds), "^(min|max)_", ""),
    threshold = unname(qcMetricsThresholds)
  )

## plots to determine thresholds
# # nFeature_RNA
# ggplot(data10xUnfiltered@meta.data, mapping = aes(x = nFeature_RNA)) +
#   # geom_histogram(mapping = aes(color = sampleId), size = 0.5) +
#   geom_histogram(bins=200) +
#   scale_x_log10() +
#   geom_vline(xintercept = qcMetricsThresholds["min_nFeature_RNA"],
#              color = "red", linetype = "dashed") +
#   geom_vline(xintercept = qcMetricsThresholds["max_nFeature_RNA"],
#              color = "red", linetype = "dashed")
# 
# # nCount_RNA
# ggplot(data10xUnfiltered@meta.data, mapping = aes(x = nCount_RNA)) +
#   # geom_histogram(mapping = aes(color = sampleId), size = 0.5) +
#   geom_histogram(bins=200) +
#   scale_x_log10() +
#   geom_vline(xintercept = qcMetricsThresholds["min_nCount_RNA"],
#              color = "red", linetype = "dashed") +
#   geom_vline(xintercept = qcMetricsThresholds["max_nCount_RNA"],
#              color = "red", linetype = "dashed")
# 
# # nFeature_ADT
# ggplot(data10xUnfiltered@meta.data, mapping = aes(x = nFeature_ADT)) +
#   geom_histogram(mapping = aes(color = sampleId), size = 0.5) +
#   # geom_histogram(bins=200) +
#   # scale_x_log10() +
#   geom_vline(xintercept = qcMetricsThresholds["nFeature_ADT"],
#              color = "red", linetype = "dashed")
#
# # nCount_ADT
# ggplot(data10xUnfiltered@meta.data, mapping = aes(x = nCount_ADT)) +
#   geom_histogram(mapping = aes(color = sampleId), size = 0.5) +
#   # geom_histogram(bins=200) +
#   scale_x_log10() +
#   geom_vline(xintercept = qcMetricsThresholds["nCount_ADT"],
#              color = "red", linetype = "dashed")
```

```{r plotQcMetricsTall, dependson="calculateQcMetrics", fig.width=8, fig.height=12}
qcMetrics.tmp <-
  c("nFeature_RNA", "nCount_RNA",
    # "nFeature_ADT", "nCount_ADT",
    "percent_mito", "percent_ribo", "percent_hb")

# Seurat version; the manual version below allows more control
# plot.tmp <-
#   VlnPlot(data10xUnfiltered, features = qcMetrics.tmp,
#           pt.size = 0.01, ncol = 2) +
#   NoLegend()
# color and modifying themes don't work very well with these ggplot2 patchwork objects

# alternate manual approach
plot.tmp <-
  P458_2_10xUnfiltered@meta.data %>%
  dplyr::select(all_of(qcMetrics.tmp)) %>%
  pivot_longer(
    cols = one_of(qcMetrics.tmp),
    names_to = "metric", values_to = "value") %>%
  mutate(metric = metric %>% factor(levels = qcMetrics.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = value)) +
  geom_violin() +
  ggrastr::rasterise(
    geom_jitter(size = 0.5), dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(metric), ncol = 2, scales = "free_y") +
  geom_hline(
    data =
      qcMetricsDfForPlot %>%
      dplyr::filter(metric %in% qcMetrics.tmp) %>%
      mutate(metric = metric %>% factor(levels = qcMetrics.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

print(plot.tmp)

outputPlots(
  plot.tmp, fileDir = "plots",
  fileStem = paste0("plotQcMetricsTallP458_2.pdf"),
  format = plotOutputFormat,
  w = 8, h = 10)



rm_tmp(ask=FALSE)

cellsKeep <-
  WhichCells(
    P458_2_10xUnfiltered,
    expression = nFeature_RNA >= qcMetricsThresholds["min_nFeature_RNA"]) %>% # keeps 2841 of 2984
  intersect(
    WhichCells(
      P458_2_10xUnfiltered,
      expression = nFeature_RNA <= qcMetricsThresholds["max_nFeature_RNA"])) %>% # keeps 2771 of 2841
  intersect(
    WhichCells(
      P458_2_10xUnfiltered,
      expression = nCount_RNA >= qcMetricsThresholds["min_nCount_RNA"])) %>% # keeps 2761 of 2771
  intersect(
    WhichCells(
      P458_2_10xUnfiltered,
      expression = nCount_RNA <= qcMetricsThresholds["max_nCount_RNA"])) %>% # keeps 2722 of 2761
  # intersect(
  #   WhichCells(
  #     data10xUnfiltered, expression = nFeature_ADT >= qcMetricsThresholds["nFeature_ADT"])) %>% # keeps ? of ?
  # intersect(
  #   WhichCells(
  #     data10xUnfiltered, expression = nCount_ADT >= qcMetricsThresholds["nCount_ADT"])) %>% # keeps ? of ?
  intersect(
    WhichCells(
      P458_2_10xUnfiltered, expression = percent_mito <= qcMetricsThresholds["percent_mito"])) %>% # keeps 2701 of 2722
  intersect(
    WhichCells(
      P458_2_10xUnfiltered, expression = percent_ribo <= qcMetricsThresholds["percent_ribo"])) %>% # keeps 2701 of 2701
  intersect(
    WhichCells(
      P458_2_10xUnfiltered, expression = percent_hb <= qcMetricsThresholds["percent_hb"])) 

cellsDrop <-
  setdiff(colnames(P458_2_10xUnfiltered), cellsKeep)

P458_2_SeuratObjFiltered <-
  subset(P458_2_10xUnfiltered, cells = cellsKeep)


```

```{r UMAP of PBMCs}

DefaultAssay(P458_2_SeuratObjFiltered) <- "RNA"

P458_2_SeuratObjFiltered <- NormalizeData(object = P458_2_SeuratObjFiltered, verbose = FALSE)
P458_2_SeuratObjFiltered <- FindVariableFeatures(object = P458_2_SeuratObjFiltered, 
        selection.method = "vst", nfeatures = 2000, verbose = FALSE)

P458_2_SeuratObjFiltered <- ScaleData(object = P458_2_SeuratObjFiltered, verbose = FALSE)
P458_2_SeuratObjFiltered <- RunPCA(object = P458_2_SeuratObjFiltered, npcs = 30, verbose = FALSE)
#ElbowPlot(seuratMerged)

P458_2_SeuratObjFiltered <- FindNeighbors(P458_2_SeuratObjFiltered, dims = 1:30)
P458_2_SeuratObjFiltered <- FindClusters(P458_2_SeuratObjFiltered, resolution = 0.2)

P458_2_SeuratObjFiltered <- RunUMAP(object = P458_2_SeuratObjFiltered, 
                        reduction = "pca",
                        dims = 1:30,
                        reduction.name = "umap")

ggplot(
    data = P458_2_SeuratObjFiltered@meta.data,
    mapping =
        aes(x = P458_2_SeuratObjFiltered@reductions$umap@cell.embeddings[,1],
            y = P458_2_SeuratObjFiltered@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = P458_2_SeuratObjFiltered@meta.data$seurat_clusters), size = .4, alpha = 0.8),
    dpi = 300) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(P458_2_SeuratObjFiltered@meta.data$seurat_clusters))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

ClustersBySamples <- ggplot(
    data = P458_2_SeuratObjFiltered@meta.data,
    mapping =
        aes(x = P458_2_SeuratObjFiltered@reductions$umap@cell.embeddings[,1],
            y = P458_2_SeuratObjFiltered@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = P458_2_SeuratObjFiltered@meta.data$seurat_clusters), size = .4, alpha = 0.8),
    dpi = 300) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(P458_2_SeuratObjFiltered@meta.data$seurat_clusters))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1))) + facet_wrap(~sample)

pdf("plots/PBMC_UMAP_sampleFacet.pdf", width = 6, height = 8)
ClustersBySamples
dev.off()

pbmc_clusterMarkers <- FindAllMarkers(P458_2_SeuratObjFiltered, assay = "RNA")
```

```{r reference matching}

 data10xMultimodalReference <-
   SeuratDisk::LoadH5Seurat("/Volumes/Bioinformatics/workspace/mlawrance/Resources/pbmc_multimodal.h5seurat")
 
 transferAnchorsData10xMultimodalReference <- Seurat::FindTransferAnchors(
   reference = data10xMultimodalReference,
   query = P458_2_SeuratObjFiltered,
   normalization.method = "SCT",
   reference.reduction = "spca",
   dims = 1:50
 )
 
 P458_2_SeuratObjFiltered <- Seurat::MapQuery(
   anchorset = transferAnchorsData10xMultimodalReference,
   query = P458_2_SeuratObjFiltered,
   reference = data10xMultimodalReference,
   refdata = list(
     celltype.l1 = "celltype.l1",
     celltype.l2 = "celltype.l2",
     celltype.l3 = "celltype.l3",
     predicted_ADT = "ADT"
   ),
   reference.reduction = "spca", 
   reduction.model = "wnn.umap"
 )
 
PredictedCellTypesPBMC <- ggplot(
    data = P458_2_SeuratObjFiltered@meta.data,
    mapping =
        aes(x = P458_2_SeuratObjFiltered@reductions$umap@cell.embeddings[,1],
            y = P458_2_SeuratObjFiltered@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = P458_2_SeuratObjFiltered@meta.data$predicted.celltype.l2), size = .4, alpha = 0.8),
    dpi = 300) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cell Type", 
    values = big_colorblind_pal(8)) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

pdf("plots/PBMC_CellPrediction.pdf", width = 8, height = 10)
PredictedCellTypesPBMC
dev.off()

HashPlot <- ggplot(
    data = P458_2_SeuratObjFiltered@meta.data,
    mapping =
        aes(x = P458_2_SeuratObjFiltered@reductions$umap@cell.embeddings[,1],
            y = P458_2_SeuratObjFiltered@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = P458_2_SeuratObjFiltered@meta.data$MonoHash), size = .4, alpha = 0.8),
    dpi = 300) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Hashed?", 
    values = big_colorblind_pal(2)) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))


pdf("plots/MonocyteHashTagPlot.pdf", height = 8, width = 10)
HashPlot
dev.off()


```

```{r HTO-Monocyte processing}
offset.tmp = 1

P458_2_SeuratObjFilteredt.tmp <- P458_2_SeuratObjFiltered %>%
  AddMetaData(
    metadata =
      log(t(as.matrix(P458_2_SeuratObjFiltered[["ADT"]]@counts + offset.tmp) ) /
            P458_2_SeuratObjFiltered$nCount_ADT * 1e6) %>%
      as.data.frame() %>%
      set_colnames(make.names(paste0("logCpm", colnames(.))))
  )

for (col.tmp in grep("logCpm", colnames(P487_2_TetramerNonZero@meta.data), value = TRUE)) {
  P487_2_TetramerNonZero@meta.data[P487_2_TetramerNonZero@meta.data[[col.tmp]] == Inf, col.tmp] <- 0
}

ggplot(
    data = P458_2_SeuratObjFiltered@meta.data,
    mapping =
        aes(x = P458_2_SeuratObjFiltered@reductions$umap@cell.embeddings[,1],
            y = P458_2_SeuratObjFiltered@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = P458_2_SeuratObjFiltered@meta.data$MonoHash), size = .4, alpha = 0.8),
    dpi = 300) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Hash", 
    values = big_colorblind_pal(2)) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

visitMarkers %>%
  #dplyr::group_by(sample) %>%
  #dplyr::filter(avg_log2FC > 0) %>%
 filter(., p_val_adj < .02)

visitFeatures <- rownames(markersForEric)

visitAverages <- AverageExpression(seuratVisit, 
                                    features = visitFeatures,
                                    assays = 'RNA',
                                    group.by = "sample",
                                    use.scale = TRUE,
                                    return.seurat = T) 

visitAvgExprMat <- visitAverages[["RNA"]]@data %>% as.matrix()

visitScaledExprMat <- t(scale(t(visitAvgExprMat)))


visitClusterAnno <- rownames(visitAverages@meta.data)

visitColumnAnno <- rowAnnotation(samples = visitClusterAnno,
                              col =  list(samples = c("SLK12v1" = "blue",
                                                      "SLK12v2" = "orange",
                                                      "SLK21v1" = "blue",
                                                      "SLK21v2" = "orange",
                                                      "SLK23v1" = "blue",
                                                      "SLK23v2" = "orange",
                                                      "SLK7v1" = "blue",
                                                      "SLK7v2" = "orange")),
                              simple_anno_size = unit(0.25, "cm"), 
                              show_legend = FALSE,
                              show_annotation_name = FALSE)

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf(file.path(plotDir, 
              "Heatmap_Integrated_visitShift_Extended_Narrow_Full.pdf"),
    height = 6,
    width = 4)

Heatmap(t(visitScaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = FALSE,
        row_order = c("SLK12v2", "SLK7v2", "SLK23v2", "SLK21v2", "SLK21v1", "SLK12v1", "SLK23v1", "SLK7v1"),
        show_row_dend = TRUE,
        left_annotation = visitColumnAnno,
        height = nrow(t(visitScaledExprMat))*unit(5, "mm"),
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())

FB_Sample_By_Cluster <- AverageExpression(MonocytesIntegratedviaCluster, assays = "FB", return.seurat = FALSE, group.by = "Cluster3Sample")

write.xlsx(FB_Sample_By_Cluster, "tables/ANOVAFBinput.xlsx", sheetName = "ClusterBySample")

MarkerListUpdated <- MarkerList[c(1,5,10,15,18,21,22,26)]

MarkerListUpdated <- append(MarkerListUpdated, "CD48.1", 5)

rownameReorderedUpdated <- rownameReorder[c(1,5,10,15,18,21,22,26)]

rownameReorderedUpdated <- append(rownameReorderedUpdated, "CD48", 5)

top5MarkersUpdated <- DotPlot(MonocytesIntegratedviaCluster, features = unique(MarkerListUpdated), group.by = "Cluster.3", assay = "FB2") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 8)) + scale_x_discrete(labels = rownameReorderedUpdated)

pdf("plots/FB_top5_Dotplot_Updated.pdf", width = 10, height = 4)
top5MarkersUpdated
dev.off()

```

```{r subsetting PBMC Monocytes and clustering}
Idents(P458_2_SeuratObjFiltered) <- "predicted.celltype.l1"
PBMCMono <- subset(P458_2_SeuratObjFiltered, idents ="Mono")

DefaultAssay(PBMCMono) <- "RNA"

PBMCMono <- NormalizeData(object = PBMCMono, verbose = FALSE)
PBMCMono <- FindVariableFeatures(object = PBMCMono, 
        selection.method = "vst", nfeatures = 2000, verbose = FALSE)

PBMCMono <- ScaleData(object = PBMCMono, verbose = FALSE)
PBMCMono <- RunPCA(object = PBMCMono, npcs = 30, verbose = FALSE)
#ElbowPlot(seuratMerged)

PBMCMono <- FindNeighbors(PBMCMono, dims = 1:30)
PBMCMono <- FindClusters(PBMCMono, resolution = 0.2)

PBMCMono <- RunUMAP(object = PBMCMono, 
                        reduction = "pca",
                        dims = 1:30,
                        reduction.name = "umap")

InitialUMAPpbmcMonoFaceted <- ggplot(
    data = PBMCMono@meta.data,
    mapping =
        aes(x = PBMCMono@reductions$umap@cell.embeddings[,1],
            y = PBMCMono@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = PBMCMono@meta.data$seurat_clusters), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(PBMCMono@meta.data$seurat_clusters))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1))) + facet_wrap(~sample)

pdf("plots/MonocyteSubsetUMAP_Unintegrated.pdf", height = 8, width = 10)
InitialUMAPpbmcMonoFaceted
dev.off()

Idents(PBMCMono) <- "seurat_clusters"

ClusterMarkersPBMCMono <- FindAllMarkers(PBMCMono, assay = "RNA")

```


```{r pbmc integration and batch correction}

pbmcSplit <- SplitObject(PBMCMono, split.by = "sample")

# normalize and identify variable features for each dataset independently
pbmcSplit <- lapply(X = pbmcSplit, FUN = function(x) {
  
  x <- NormalizeData(object = x,
                     assay = "ADT",
                     verbose = FALSE,
                     normalization.method = "CLR", 
                     margin = 2,)
  
  #VariableFeatures(x) <- citeSeqAbs
  
  
  x <- NormalizeData(object = x, assay = "RNA", verbose = FALSE)
  x <- FindVariableFeatures(object = x, 
                            selection.method = "vst", nfeatures = 200, verbose = FALSE)
  
})

```

```{r integrate}

features <- SelectIntegrationFeatures(object.list = pbmcSplit)
pbmcSplit <- lapply(X = pbmcSplit, FUN = function(x) {
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
})


fastIntAnchors <- FindIntegrationAnchors(object.list = pbmcSplit, 
                                         anchor.features = features,
                                         reduction = "rpca")

pbmcIntegrated <- IntegrateData(anchorset = fastIntAnchors)

DefaultAssay(pbmcIntegrated) <- "integrated"

```

```{r batchCorrected PBMCs}

pbmcIntegrated <- NormalizeData(object = pbmcIntegrated, verbose = FALSE)
pbmcIntegrated <- FindVariableFeatures(object = pbmcIntegrated, 
        selection.method = "vst", nfeatures = 2000, verbose = FALSE)

pbmcIntegrated <- ScaleData(object = pbmcIntegrated, verbose = FALSE)
pbmcIntegrated <- RunPCA(object = pbmcIntegrated, npcs = 30, verbose = FALSE)
#ElbowPlot(seuratMerged)

pbmcIntegrated <- FindNeighbors(pbmcIntegrated, dims = 1:30)

DefaultAssay(pbmcIntegrated) <- "integrated"
pbmcIntegrated <- FindClusters(pbmcIntegrated, resolution = 0.1)

pbmcIntegrated <- RunUMAP(object = pbmcIntegrated, 
                        reduction = "pca",
                        dims = 1:30,
                        reduction.name = "umap")

IntegratedUMAP_Pbmc <- ggplot(
    data = pbmcIntegrated@meta.data,
    mapping =
        aes(x = pbmcIntegrated@reductions$umap@cell.embeddings[,1],
            y = pbmcIntegrated@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = pbmcIntegrated@meta.data$seurat_clusters), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(18)[10:15]) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))


pdf("plots/MonocyteSubsetUMAP_Integrated.pdf", height = 8, width = 10)
IntegratedUMAP_Pbmc
dev.off()

IntegratedUMAP_Pbmc_faceted <- ggplot(
    data = pbmcIntegrated@meta.data,
    mapping =
        aes(x = pbmcIntegrated@reductions$umap@cell.embeddings[,1],
            y = pbmcIntegrated@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = pbmcIntegrated@meta.data$seurat_clusters), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(pbmcIntegrated@meta.data$seurat_clusters))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1))) +  facet_wrap(~sample)

pdf("plots/MonocyteSubsetUMAP_Integrated_Faceted.pdf", height = 8, width = 10)
IntegratedUMAP_Pbmc_faceted
dev.off()

Idents(PBMCMono) <- "seurat_clusters"

ClusterMarkerspbmcIntegrated <- FindAllMarkers(pbmcIntegrated, assay = "RNA")

ClusterMarkerspbmcIntegrated8 <- FindAllMarkers(pbmcIntegrated, assay = "RNA", logfc.threshold = .8)

write.xlsx(ClusterMarkerspbmcIntegrated, "tables/integratedClusterMarkers.xlsx")

```

```{r heatmap of top markers for PBMC integrated clusters}

pbmcHeatmap <- ClusterMarkerspbmcIntegrated  %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 5) %>%
  dplyr::slice_max(avg_log2FC, n = 5) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

pbmcFeatures <- unique((pbmcHeatmap$gene))



pbmcAverages <- AverageExpression(pbmcIntegrated, 
                                    features = pbmcFeatures,
                                    assays = 'RNA',
                                    group.by = "seurat_clusters",
                                    use.scale = TRUE,
                                    return.seurat = T) 

pbmcAvgExprMat <- pbmcAverages[["RNA"]]@data %>% as.matrix()

pbmcScaledExprMat <- t(scale(t(pbmcAvgExprMat)))


pbmcClusterAnno <- rownames(pbmcAverages@meta.data)

pbmcColumnAnno <- rowAnnotation(Cluster = pbmcClusterAnno,
                              col = list(Cluster = clusterColorsPbmc),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf(file.path(plotDir, 
              "Heatmap_pbmc_TopMarkers.pdf"),
    height = 6,
    width = 6)

        Heatmap(t(pbmcScaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = FALSE,
        show_row_dend = TRUE,
        left_annotation = pbmcColumnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        column_names_rot = 90,
        row_names_rot = 90,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())

```

```{r integrate BAL and PBMC}

integrationList <- list(BALSeurat, PBMCMono)


integrationList <- lapply(X = integrationList, FUN = function(x) {

  
  #VariableFeatures(x) <- citeSeqAbs
  
  
  x <- NormalizeData(object = x, assay = "RNA", verbose = FALSE)
  x <- FindVariableFeatures(object = x, 
                            selection.method = "vst", nfeatures = 1000, verbose = FALSE)
  
})

features <- SelectIntegrationFeatures(object.list = integrationList)

integrationList <- lapply(X = integrationList, FUN = function(x) {
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
})


fastIntAnchors <- FindIntegrationAnchors(object.list = integrationList, 
                                         anchor.features = features,
                                         reduction = "rpca")

allIntegrated <- IntegrateData(anchorset = fastIntAnchors)

DefaultAssay(allIntegrated) <- "integrated"

allIntegrated <- ScaleData(object = allIntegrated, verbose = FALSE)
allIntegrated <- RunPCA(object = allIntegrated, npcs = 30, verbose = FALSE)
#ElbowPlot(seuratMerged)

allIntegrated <- FindNeighbors(allIntegrated, dims = 1:30)

DefaultAssay(pbmcIntegrated) <- "integrated"
pbmcIntegrated <- FindClusters(pbmcIntegrated, resolution = 0.7)

allIntegrated <- RunUMAP(object = allIntegrated, 
                        reduction = "pca",
                        dims = 1:30,
                        reduction.name = "umap",
                        return.model = TRUE)

ggplot(
    data = allIntegrated@meta.data,
    mapping =
        aes(x = allIntegrated@reductions$umap@cell.embeddings[,1],
            y = allIntegrated@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = allIntegrated@meta.data$origin), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Origin", 
    values = big_colorblind_pal(n_distinct(allIntegrated@meta.data$origin))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))


ggplot(
    data = allIntegrated@meta.data[allIntegrated@meta.data$origin == "BAL",],
    mapping =
        aes(x = allIntegrated@reductions$umap@cell.embeddings[,1][allIntegrated@meta.data$origin == "BAL"],
            y = allIntegrated@reductions$umap@cell.embeddings[,2][allIntegrated@meta.data$origin == "BAL"])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = allIntegrated@meta.data$Cluster.3[allIntegrated@meta.data$origin == "BAL"]), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Origin", 
    values = big_colorblind_pal(n_distinct(allIntegrated@meta.data$Cluster.3))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

ggplot(
    data = allIntegrated@meta.data[allIntegrated@meta.data$Cluster.3 == "3"|allIntegrated@meta.data$Cluster.3 == "4" | allIntegrated@meta.data$Cluster.3 == "PBMC",],
    mapping =
        aes(x = allIntegrated@reductions$umap@cell.embeddings[,1][allIntegrated@meta.data$Cluster.3 == "3"|allIntegrated@meta.data$Cluster.3 == "4"| allIntegrated@meta.data$Cluster.3 == "PBMC"],
            y = allIntegrated@reductions$umap@cell.embeddings[,2][allIntegrated@meta.data$Cluster.3 == "3"|allIntegrated@meta.data$Cluster.3 == "4"| allIntegrated@meta.data$Cluster.3 == "PBMC"])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = allIntegrated@meta.data$Cluster.3[allIntegrated@meta.data$Cluster.3 == "3"|allIntegrated@meta.data$Cluster.3 == "4" | allIntegrated@meta.data$Cluster.3 == "PBMC"]), size = 1, alpha = 0.2),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Origin", 
    values = originColors,
    labels = c("Inflammatory Monocyte", "Mature Macrophage", "PBMC derived Macrophage")) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1))) + facet_wrap(~Cluster.3)

clusterReplace <- replace_na(as.character(allIntegrated@meta.data$cluster), '-5') %>% as.numeric(.) + 9
clusterReplace[clusterReplace == 4] <- "BAL"
clusterReplace <- factor(clusterReplace)

allIntegrated@meta.data$cluster <- clusterReplace

ggplot(
    data = allIntegrated@meta.data,
    mapping =
        aes(x = allIntegrated@reductions$umap@cell.embeddings[,1],
            y = allIntegrated@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = allIntegrated@meta.data$clusterMerged), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = mergedClusterColors) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1))) + facet_wrap(~origin)

allIntegrated <- FindClusters(allIntegrated, resolution = .1)

```
```{r find biological groupings}

topMarkersBAL <- frame30long %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 5) %>%
  dplyr::slice_max(avg_log2FC, n = 5) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 
```

```{r figures for presentation}

Figure1 <- ggplot(
    data = BALSeurat@meta.data,
    mapping =
        aes(x = BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,1],
            y = BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = BALSeurat@meta.data$Cluster.3), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cell Category", 
    values = big_colorblind_pal(9),
    labels = names(mergedClusterColors[1:9])) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))


IntegratedUMAP_Pbmc

InteGratedPBMCHeatmap <-         Heatmap(t(pbmcScaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = FALSE,
        show_row_dend = TRUE,
        left_annotation = pbmcColumnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        column_names_rot = 90,
        row_names_rot = 90,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))


Figure3 <- FeaturePlot(pbmcIntegrated, reduction = "umap", features = "CD14.1", pt.size = 1) + labs(title = "CD14 Expression")

Figure4 <- FeaturePlot(pbmcIntegrated, reduction = "umap", features = "CD16", pt.size = 1) + labs(title = "CD16 Expression")

Figure5 <- ggplot(
    data = allIntegrated@meta.data,
    mapping =
        aes(x = allIntegrated@reductions$umap@cell.embeddings[,1],
            y = allIntegrated@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = allIntegrated@meta.data$origin), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Origin", 
    values = c("orange", "gray")) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

ggplot(
    data = allIntegrated@meta.data,
    mapping =
        aes(x = allIntegrated@reductions$umap@cell.embeddings[,1],
            y = allIntegrated@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = allIntegrated@meta.data$clusterMerged), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster Merged", 
    values = mergedClusterColors) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

Figure6 <-ggplot(
    data = allIntegrated@meta.data,
    mapping =
        aes(x = allIntegrated@reductions$umap@cell.embeddings[,1],
            y = allIntegrated@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = allIntegrated@meta.data$seurat_cluster), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(allIntegrated@meta.data$seurat_cluster))) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

Figure7 <- ggplot(
    data = allIntegrated@meta.data[allIntegrated@meta.data$Cluster.3 == "3"|allIntegrated@meta.data$Cluster.3 == "4" | allIntegrated@meta.data$Cluster.3 == "PBMC",],
    mapping =
        aes(x = allIntegrated@reductions$umap@cell.embeddings[,1][allIntegrated@meta.data$Cluster.3 == "3"|allIntegrated@meta.data$Cluster.3 == "4"| allIntegrated@meta.data$Cluster.3 == "PBMC"],
            y = allIntegrated@reductions$umap@cell.embeddings[,2][allIntegrated@meta.data$Cluster.3 == "3"|allIntegrated@meta.data$Cluster.3 == "4"| allIntegrated@meta.data$Cluster.3 == "PBMC"])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = allIntegrated@meta.data$Cluster.3[allIntegrated@meta.data$Cluster.3 == "3"|allIntegrated@meta.data$Cluster.3 == "4" | allIntegrated@meta.data$Cluster.3 == "PBMC"]), size = 1, alpha = 0.2),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Origin", 
    values = originColors,
    labels = c("Mature Macrophage", "Inflammatory Macrophage", "PBMC derived Macrophage")) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1))) + facet_wrap(~Cluster.3)


allIntegratedHeatmap <- ClusterMarkersallIntegrated  %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 5) %>%
  dplyr::slice_max(avg_log2FC, n = 5) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

allIntegratedFeatures <- unique((allIntegratedHeatmap$gene))



allIntegratedAverages <- AverageExpression(allIntegrated, 
                                    features = allIntegratedFeatures,
                                    assays = 'RNA',
                                    group.by = "seurat_clusters",
                                    use.scale = TRUE,
                                    return.seurat = T) 

allIntegratedAvgExprMat <- allIntegratedAverages[["RNA"]]@data %>% as.matrix()

allIntegratedScaledExprMat <- t(scale(t(allIntegratedAvgExprMat)))


allIntegratedClusterAnno <- rownames(allIntegratedAverages@meta.data)

allIntegratedColumnAnno <- rowAnnotation(Cluster = allIntegratedClusterAnno,
                              col = list(Cluster = clusterColorsallIntegrated),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  


      PanelB <- Heatmap(t(allIntegratedScaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        left_annotation = allIntegratedColumnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        column_names_rot = 90,
        row_names_rot = 90,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))

        
        clusterCounts <- allIntegrated@meta.data %>%
  dplyr::group_by(seurat_clusters, origin) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(seurat_clusters) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(seurat_clusters)

originBarplot <- clusterCounts %>%
  ggplot(aes(x = seurat_clusters,
             y = pctCells,
             fill = origin)) +
  geom_col() +
  scale_fill_manual(values = originColorsCoarse) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "",
       y = "% of cells",
       fill = "Origin") +
  theme(axis.text.x=element_text(angle=60, hjust=1))


        clusterCounts2 <- allIntegrated@meta.data %>%
  dplyr::group_by(seurat_clusters, clusterMerged, origin) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(seurat_clusters) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(seurat_clusters)
        
        specificBarplot <- clusterCounts2 %>%
  ggplot(aes(x = seurat_clusters,
             y = pctCells,
             color = origin,
             fill = clusterMerged)) +
  geom_col() +
  scale_color_manual(values = originColorsCoarse) +
  scale_fill_manual(values = mergedClusterColors) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "",
       y = "% of cells",
       fill = "Original Cluster",
       color = "Origin") +
  theme(axis.text.x=element_text(angle=60, hjust=1))
        
        FigureSplit <- ggplot(
    data = BALSeurat@meta.data[BALSeurat@meta.data$Cluster.3 == "3" | BALSeurat@meta.data$Cluster.3 == "4",],
    mapping =
        aes(x = BALSeurat@reductions$umap_visit@cell.embeddings[,1][BALSeurat@meta.data$Cluster.3 == "3" | BALSeurat@meta.data$Cluster.3 == "4"],
            y = BALSeurat@reductions$umap_visit@cell.embeddings[,2][BALSeurat@meta.data$Cluster.3 == "3" | BALSeurat@meta.data$Cluster.3 == "4"])) +
    ggrastr::rasterise(
        geom_point(mapping = aes_string(color = BALSeurat@meta.data$Cluster.3[BALSeurat@meta.data$Cluster.3 == "3" | BALSeurat@meta.data$Cluster.3 == "4"]), size = .4, alpha = 0.8),
        dpi = 300) +
    labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(n_distinct(BALSeurat@meta.data$Cluster.3,))[5:6],
    labels = c("Inflammatory Macrophage", "Mature Macrophage")) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1))) + facet_wrap(~Cluster.3)
        
        
pdf("FiguresSys/Figure1.pdf", height = 8, width = 12)
Figure1
dev.off()

pdf("FiguresSys/Figure1Split.pdf", height = 8, width = 12)
FigureSplit
dev.off()

pdf("FiguresSys/Figure2.pdf", height = 8, width = 12)
IntegratedUMAP_Pbmc
dev.off()

pdf("FiguresSys/Figure2Facet.pdf", height = 8, width = 12)
IntegratedUMAP_Pbmc + facet_wrap(~sample)
dev.off()

pdf("FiguresSys/Figure2Heatmap.pdf", height = 10, width = 10)
InteGratedPBMCHeatmap
dev.off()

pdf("FiguresSys/Figure3.pdf", height = 16, width = 10)
Figure3+Figure4
dev.off()

pdf("FiguresSys/Figure4.pdf", height = 8, width = 12)
Figure5
dev.off()

pdf("FiguresSys/Figure6.pdf", height = 8, width = 12)
Figure7
dev.off()

pdf("FiguresSys/Figure7.pdf", height = 8, width = 12)
mergedClusterFigure
dev.off()

pdf("FiguresSys/Figure8.pdf", height = 8, width = 12)
Figure6
dev.off()

pdf("FiguresSys/Figure8Heatmap.pdf", height = 8, width = 12)
Figure6Heatmap
dev.off()

pdf("FiguresSys/Figure9.pdf", height = 12, width = 12)
originBarplot
dev.off()

pdf("FiguresSys/Figure9.pdf", height = 12, width = 24)
specificBarplot
dev.off()
```

```{r reference mapping}

BALSeurat <- RunUMAP(BALSeurat,
                                         assay = "integrated",
                                         reduction ="pca",
                                         dims= 1:30,
                                         reduction.name = "MonoRNA.umap", 
                                         reduction.key = "_MonoRNA.umap",
        return.model = TRUE)

overlayAnchors <- FindTransferAnchors(reference = BALSeurat, query = pbmcIntegrated,
    dims = 1:30, reference.reduction = "pca")

PBMC.query <- MapQuery(anchorset = overlayAnchors, reference = BALSeurat, query = pbmcIntegrated, refdata = "Cluster.3", reference.reduction = "pca", reduction.model = "MonoRNA.umap")

DimPlot(BALSeurat, reduction = "MonoRNA.umap", group.by = "Cluster.3", label = TRUE, label.size = 3,
    repel = TRUE) + NoLegend() + ggtitle("BAL Clusters") +

DimPlot(PBMC.query, reduction = "ref.umap", group.by = "seurat_clusters", label = TRUE, label.size = 3,
    repel = TRUE) + NoLegend() + ggtitle("PBMC Cluster annotations")


QueryMappedFigure <- ggplot(
    data = PBMC.query@meta.data,
    mapping =
        aes(x = PBMC.query@reductions$ref.umap@cell.embeddings[,1],
            y = PBMC.query@reductions$ref.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = PBMC.query@meta.data$seurat_clusters), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  xlim(min(BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,1]), max(BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,1])) +
    ylim(min(BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,2]), max(BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  scale_color_manual(
    "PBMC Cluster", 
    values = mergedClusterColors[10:17],
    labels = names(mergedClusterColors[10:17])) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

Figure1 + QueryMappedFigure

pdf("plots/PBMCMapped.pdf", height = 8, width = 12)
QueryMappedFigure
dev.off()

predictions <- TransferData(anchorset = overlayAnchors, refdata = BALSeurat$Cluster.3,
                            dims = 1:30)
PBMC.query <- AddMetaData(PBMC.query, metadata = predictions)


QueryAnnotatedFigure <- ggplot(
    data = PBMC.query@meta.data,
    mapping =
        aes(x = PBMC.query@reductions$ref.umap@cell.embeddings[,1],
            y = PBMC.query@reductions$ref.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = PBMC.query@meta.data$Cluster.3), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  xlim(min(BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,1]), max(BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,1])) +
    ylim(min(BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,2]), max(BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  scale_color_manual(
    "BAL Cluster", 
    values = PBMCrelabledAnote,
    labels = names(mergedClusterColors[c(1,2,5,6,8,9)])) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

pdf("plots/PBMCMappedBALAnnotations.pdf", height = 8, width = 12)
QueryAnnotatedFigure
dev.off()
```

```{r}

clusterCounts <- PBMC.query@meta.data %>%
  dplyr::group_by(Cluster.3, seurat_clusters) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(Cluster.3) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(Cluster.3)

newBarplot <- clusterCounts %>%
  ggplot(aes(x = Cluster.3,
             y = pctCells,
             fill = seurat_clusters)) +
  geom_col() +
  scale_fill_manual(values = mergedClusterColors[10:15]) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "",
       y = "% of cells",
       fill = "PBMC Cluster") +
  theme(axis.text.x=element_text(angle=60, hjust=1))
```

```{r panel Figures}

ClusterMarkersallIntegrated <- FindAllMarkers(allIntegrated, assay = "RNA", logfc.threshold = .8)

allIntegratedHeatmap <- ClusterMarkersallIntegrated  %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 5) %>%
  dplyr::slice_max(avg_log2FC, n = 5) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

allIntegratedFeatures <- unique((allIntegratedHeatmap$gene))



allIntegratedAverages <- AverageExpression(allIntegrated, 
                                    features = allIntegratedFeatures,
                                    assays = 'RNA',
                                    group.by = "seurat_clusters",
                                    use.scale = TRUE,
                                    return.seurat = T) 

allIntegratedAvgExprMat <- allIntegratedAverages[["RNA"]]@data %>% as.matrix()

allIntegratedScaledExprMat <- t(scale(t(allIntegratedAvgExprMat)))


allIntegratedClusterAnno <- rownames(allIntegratedAverages@meta.data)

allIntegratedColumnAnno <- rowAnnotation(Cluster = allIntegratedClusterAnno,
                              col = list(Cluster = clusterColorsallIntegrated),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  


      PanelB <- Heatmap(t(allIntegratedScaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        left_annotation = allIntegratedColumnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        column_names_rot = 90,
        row_names_rot = 90,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))

PanelA <- Figure6

Supplemental1 <- ClusterMarkersallIntegrated  %>%
  dplyr::group_by(cluster) %>%
  dplyr::slice_min(p_val_adj, n = 50) %>%
  dplyr::slice_max(avg_log2FC, n = 50)

Supplemental1 <- ClusterMarkersallIntegrated  %>%
  dplyr::group_by(cluster) %>%
  dplyr::slice_min(p_val_adj, n = 50) %>%
  dplyr::slice_min(avg_log2FC, n = 50) %>% rbind(Supplemental1,.) %>% arrange(cluster)



allIntegrated <- NormalizeData(allIntegrated, assay = "ADT", normalization.method = "CLR")

allIntegrated <- ScaleData(object = allIntegrated, verbose = FALSE, assay = "ADT")

ClusterMarkersallIntegratedFB <- FindAllMarkers(allIntegrated, assay = "FB3")

Supplemental2 <- ClusterMarkersallIntegratedFB  %>%
  dplyr::group_by(cluster) %>%
  dplyr::slice_min(p_val_adj, n = 20) %>%
  dplyr::slice_max(avg_log2FC, n = 10)

Supplemental2 <- ClusterMarkersallIntegratedFB  %>%
  dplyr::group_by(cluster) %>%
  dplyr::slice_min(p_val_adj, n = 20) %>%
  dplyr::slice_min(avg_log2FC, n = 10) %>% rbind(Supplemental2,.) %>% arrange(cluster)

PanelC <- originBarplot

PanelD <- Figure1

PanelE <- QueryMappedFigure

PanelF <-  Heatmap(t(pbmcScaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = FALSE,
        show_row_dend = TRUE,
        left_annotation = pbmcColumnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        column_names_rot = 90,
        row_names_rot = 90,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))

Supplemental3 <- ClusterMarkerspbmcIntegrated  %>%
  dplyr::group_by(cluster) %>%
  dplyr::slice_min(p_val_adj, n = 10) %>%
  dplyr::slice_max(avg_log2FC, n = 10)

Supplemental3 <- ClusterMarkerspbmcIntegrated  %>%
  dplyr::group_by(cluster) %>%
  dplyr::slice_min(p_val_adj, n = 10) %>%
  dplyr::slice_min(avg_log2FC, n = 10) %>% rbind(Supplemental3,.) %>% arrange(cluster)



pbmcIntegrated <- NormalizeData(pbmcIntegrated, assay = "ADT", normalization.method = "CLR")

pbmcIntegrated <- ScaleData(object = pbmcIntegrated, verbose = FALSE, assay = "ADT")

ClusterMarkerspbmcIntegratedFB <- FindAllMarkers(pbmcIntegrated, assay = "FB3")

Supplemental4 <- ClusterMarkerspbmcIntegratedFB  %>%
  dplyr::group_by(cluster) %>%
  dplyr::slice_min(p_val_adj, n = 20) %>%
  dplyr::slice_max(avg_log2FC, n = 10)

Supplemental4 <- ClusterMarkerspbmcIntegratedFB  %>%
  dplyr::group_by(cluster) %>%
  dplyr::slice_min(p_val_adj, n = 20) %>%
  dplyr::slice_min(avg_log2FC, n = 10) %>% rbind(Supplemental4,.) %>% arrange(cluster)

write.xlsx(as.data.frame(Supplemental1), file = "tables/Supplemental1_PBMC_BAL_cluster_markers.xlsx", sheetName = "RNA Markers")
write.xlsx(as.data.frame(Supplemental2), file = "tables/Supplemental1_PBMC_BAL_cluster_markers.xlsx", sheetName = "FB Markers", append = TRUE)
write.xlsx(as.data.frame(Supplemental1_Cluster6_additional), file = "tables/Supplemental1_PBMC_BAL_cluster_markers.xlsx", sheetName = "Expanded Cluster 6 RNA Markers", append = TRUE)

write.xlsx(as.data.frame(Supplemental3), file = "tables/Supplemental2_PBMC_cluster_markers.xlsx", sheetName = "RNA Markers")
write.xlsx(as.data.frame(Supplemental4), file = "tables/Supplemental2_PBMC_cluster_markers.xlsx", sheetName = "FB Markers", append = TRUE)

pdf(file = "plots/FigurePanels/PanelA.pdf", height = 8, width = 12)
PanelA
dev.off()

pdf(file = "plots/FigurePanels/PanelB.pdf", height = 8, width = 12)
PanelB
dev.off()

pdf(file = "plots/FigurePanels/PanelC.pdf", height = 8, width = 12)
PanelC
dev.off()

pdf(file = "plots/FigurePanels/PanelD.pdf", height = 8, width = 12)
PanelD
dev.off()

pdf(file = "plots/FigurePanels/PanelE.pdf", height = 8, width = 12)
PanelE
dev.off()

pdf(file = "plots/FigurePanels/PanelF.pdf", height = 8, width = 12)
PanelF
dev.off()

allIntegrated <- 
  PercentageFeatureSet(allIntegrated, "^MT-", col.name = "percent_mito", assay = "RNA") # percent mitochondrial reads
allIntegrated <-
  PercentageFeatureSet(allIntegrated, "^RP[SL]", col.name = "percent_ribo", assay = "RNA") # percent ribosomal protein reads
allIntegrated <-
  PercentageFeatureSet(allIntegrated, "^HB[^(P)]", col.name = "percent_hb", assay = "RNA") # percent hemoglobin reads

cluster6qc <-
  allIntegrated@meta.data[allIntegrated@meta.data$seurat_cluster == "6",] %>%
  dplyr::select(all_of(qcMetrics.tmp)) %>%
  pivot_longer(
    cols = one_of(qcMetrics.tmp),
    names_to = "metric", values_to = "value") %>%
  mutate(metric = metric %>% factor(levels = qcMetrics.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = value)) +
  geom_violin() +
  ggrastr::rasterise(
    geom_jitter(size = 0.5), dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(metric), ncol = 2, scales = "free_y") +
  geom_hline(
    data =
      qcMetricsDfForPlot %>%
      dplyr::filter(metric %in% qcMetrics.tmp) %>%
      mutate(metric = metric %>% factor(levels = qcMetrics.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

cluster0qc <-
  allIntegrated@meta.data[allIntegrated@meta.data$seurat_cluster == "0",] %>%
  dplyr::select(all_of(qcMetrics.tmp)) %>%
  pivot_longer(
    cols = one_of(qcMetrics.tmp),
    names_to = "metric", values_to = "value") %>%
  mutate(metric = metric %>% factor(levels = qcMetrics.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = value)) +
  geom_violin() +
  ggrastr::rasterise(
    geom_jitter(size = 0.5), dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(metric), ncol = 2, scales = "free_y") +
  geom_hline(
    data =
      qcMetricsDfForPlot %>%
      dplyr::filter(metric %in% qcMetrics.tmp) %>%
      mutate(metric = metric %>% factor(levels = qcMetrics.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

pbmc11cells <-
  pbmcIntegrated@meta.data[pbmcIntegrated@meta.data$seurat_cluster == "11",] %>%
  dplyr::select(all_of(qcMetrics.tmp)) %>%
  pivot_longer(
    cols = one_of(qcMetrics.tmp),
    names_to = "metric", values_to = "value") %>%
  mutate(metric = metric %>% factor(levels = qcMetrics.tmp)) %>%
  ggplot(mapping = aes(x = "sample", y = value)) +
  geom_violin() +
  ggrastr::rasterise(
    geom_jitter(size = 0.5), dpi = rasterResolutionDpi) +
  facet_wrap(facets = vars(metric), ncol = 2, scales = "free_y") +
  geom_hline(
    data =
      qcMetricsDfForPlot %>%
      dplyr::filter(metric %in% qcMetrics.tmp) %>%
      mutate(metric = metric %>% factor(levels = qcMetrics.tmp)),
    mapping = aes(yintercept = threshold),
    color = "red", linetype = "dashed") +
  theme(
    axis.text.x = element_blank(),
    # axis.text.x = element_text(angle=-45, hjust=0, size = 10),
    strip.text=element_text(size = rel(0.8), margin=margin(t=3, b=3))) +
  labs(x = NULL, y = NULL)

```


```{r referenced mapped heatmap}

ClusterMarkersQueryMapped <- FindAllMarkers(PBMC.query, assay = "RNA" logfc.threshold = 1)

pbmcHeatmap <- ClusterMarkersQueryMapped  %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 5) %>%
  dplyr::slice_max(avg_log2FC, n = 5) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

pbmcFeatures <- unique((pbmcHeatmap$gene))



pbmcAverages <- AverageExpression(PBMC.query, 
                                    features = pbmcFeatures,
                                    assays = 'RNA',
                                    group.by = "BALAnnotation",
                                    use.scale = TRUE,
                                    return.seurat = T) 

pbmcAvgExprMat <- pbmcAverages[["RNA"]]@data %>% as.matrix()

pbmcScaledExprMat <- t(scale(t(pbmcAvgExprMat)))


pbmcClusterAnno <- rownames(pbmcAverages@meta.data)

pbmcColumnAnno <- rowAnnotation(Cluster = pbmcClusterAnno,
                              col = list(Cluster = PBMCrelabledAnote),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf(file.path(plotDir, 
              "Heatmap_pbmc_BAL_celltypes.pdf"),
    height = 6,
    width = 8)

        Heatmap(t(pbmcScaledExprMat), 
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = FALSE,
        show_row_dend = TRUE,
        left_annotation = pbmcColumnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        column_names_rot = 90,
        row_names_rot = 0,
        column_names_gp = gpar(fontsize = 5),
        row_names_gp = gpar(fontsize = 10))
        

invisible(dev.off())
```

```{r scRNA velocity conversion and analysis}
#the first part of this is going to convert the Seurat object into a format Python can use. Rest of analysis will be done outside of R, and then imported back into R for analysis

BALSeurat$barcode <- colnames(BALSeurat) %>% gsub(".*_", "", .)
BALSeurat$UMAP_1 <- BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,1]
BALSeurat$UMAP_2 <- BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,2]
write.csv(BALSeurat@meta.data, file='metadata.csv', quote=F, row.names=F)

write.csv(BALSeurat@reductions$pca@cell.embeddings, file='pca.csv', quote=F, row.names=F)

counts_matrix <- GetAssayData(BALSeurat, assay='RNA', slot='counts')

write.table(
  data.frame('gene'=rownames(counts_matrix)),file='gene_names.csv',
  quote=F,row.names=F,col.names=F
)

read.csv("Cluster_Velocities.csv")

VelocityBoxPlot <- ggplot(boxplotVelocities, aes(x=Cluster.3, y=velocities, fill=Cluster.3)) + geom_boxplot() + labs(x = "Cell Category", y = "Velocity") + coord_flip() + scale_fill_manual(values = mergedClusterColorBox) +  NoLegend()

pdf("plots/VelocityBoxplot.pdf", height = 8, width =8)
VelocityBoxPlot
dev.off()


for(i in levels(factor(boxplotVelocities$Cluster.3))){
  print(paste("Mean for", i, "is", mean(boxplotVelocities$velocities[boxplotVelocities$Cluster.3 == i])))
}

```


```{r}

BALSeurat <- RunPCA(BALSeurat,
                                        features = featuresFB,
                                        assay = "FB",
                                        reduction.name = "FB_pca")

BALSeurat <- FindMultiModalNeighbors(BALSeurat,
                                               reduction.list = list("visitFBPCA", "visitRNAPCA"),
                                               dims.list = list(1:20, 1:10),
                                               modality.weight.name = list("BestIntegratedFB.weight", "BestIntegratedRNA.weight"))

BALSeurat <- RunUMAP(BALSeurat, 
                               reduction.name = "umap_FB", 
                               reduction.key = "UMAP_FB",
                     reduction = "FB_pca",
                     dims = 1:20)

FBUMAP_RNACluster <- ggplot(
    data = BALSeurat@meta.data,
    mapping =
        aes(x = BALSeurat@reductions$umap_FB@cell.embeddings[,1],
            y = BALSeurat@reductions$umap_FB@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = BALSeurat@meta.data$Cluster.3), size = .4, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cell Category", 
    values = big_colorblind_pal(9),
    labels = names(mergedClusterColors[1:9])) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

BALSeurat <- FindNeighbors(BALSeurat, dims = 1:30, reduction = "FB_pca")

BALSeurat <- FindClusters(BALSeurat, resolution = 0.16)

BALSeurat@meta.data$Cluster_FB <- BALSeurat@meta.data$seurat_clusters


FBUMAP_RNACluster <- ggplot(
    data = BALSeurat@meta.data,
    mapping =
        aes(x = BALSeurat@reductions$umap_FB@cell.embeddings[,1],
            y = BALSeurat@reductions$umap_FB@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = BALSeurat@meta.data$Cluster_FB), size = .8, alpha = 0.8),
    dpi = 300) +
 labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "FB Based Cluster", 
    values = big_colorblind_pal(9)) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))

       clusterCounts <- BALSeurat@meta.data %>%
  dplyr::group_by(Cluster_FB, CellIdentities) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(Cluster_FB) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(Cluster_FB)

originBarplot <- clusterCounts %>%
  ggplot(aes(x = Cluster_FB,
             y = pctCells,
             fill = CellIdentities)) +
  geom_col() +
  scale_fill_manual(values = mergedClusterColors[1:9]) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "",
       y = "% of cells",
       fill = "Cell Idenitity from RNA") +
  theme(axis.text.x=element_text(angle=60, hjust=1))

       clusterCountsSample <- BALSeurat@meta.data %>%
  dplyr::group_by(Cluster_FB, sample) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(Cluster_FB) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(Cluster_FB)
       
write.csv(clusterCountsSample, "tables/FB_ClustersBySample.csv")
```
  
```{r}

pdf("~/Documents/LymphotcytesP458.pdf", height = 8, width =8)

ggplot(
    data = seuratMerged@meta.data,
    mapping =
        aes(x = seuratMerged@reductions$umap@cell.embeddings[,1],
            y = seuratMerged@reductions$umap@cell.embeddings[,2])) +
    ggrastr::rasterise(
        geom_point(mapping = aes_string(color = seuratMerged@meta.data$predicted.celltype.l1), size = .4, alpha = 0.8),
        dpi = 300) +
    labs(x = "UMAP 1", y = "UMAP 2") +
    scale_color_manual(
        "Predicted Cell Type", 
        values = big_colorblind_pal(9),
        labels = c("B [219]", "CD4 T [1242]", "CD8 T [565]", "DC [3123]", "Monocyte [59104]", "NK [44", "other [6]", "other T [14]")) +
    guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))
dev.off()

pdf("~/Documents/samplesP458.pdf", height = 8, width =8)
ggplot(
    data = seuratMerged@meta.data,
    mapping =
        aes(x = seuratMerged@reductions$umap@cell.embeddings[,1],
            y = seuratMerged@reductions$umap@cell.embeddings[,2])) +
    ggrastr::rasterise(
        geom_point(mapping = aes_string(color = factor(seuratMerged@meta.data$sample)), size = .4, alpha = 0.8),
        dpi = 300) +
    labs(x = "UMAP 1", y = "UMAP 2") +
    scale_color_manual(
        "Predicted Cell Type", 
        values = big_colorblind_pal(12)) +
    guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))
    
```


```{r merge Lymphocytes}
#no integration here, just to see how bad the batch is

mergedLymphocytes <- merge(seaLymphocytes, 
                       y = coloradoLymphocytes,
                       add.cell.ids = c("sea", "col"),
                       merge.data = TRUE)

mergedLymphocytes <- NormalizeData(mergedLymphocytes, normalization.method = "LogNormalize")

mergedLymphocytes <- FindVariableFeatures(mergedLymphocytes, selection.method = "vst", nfeatures = 2000)

mergedLymphocytes <- ScaleData(mergedLymphocytes, features = rownames(mergedLymphocytes))

mergedLymphocytes<- RunPCA(mergedLymphocytes, features = VariableFeatures(object = mergedLymphocytes))

mergedLymphocytes <- FindNeighbors(mergedLymphocytes, dims = 1:10)
mergedLymphocytes <- FindClusters(mergedLymphocytes, resolution = 0.5)

mergedLymphocytes <- RunUMAP(mergedLymphocytes, dims = 1:10)

#batched as all hell, time to integrate instead of merge

seurat.list <- list()

DefaultAssay(coloradoLymphocytes) <- "RNA"
coloradoLymphocytes <- FindVariableFeatures(coloradoLymphocytes, selection.method = "vst", nfeatures = 2000)
seurat.list[[1]] <-  coloradoLymphocytes

seurat.list[[2]] <- seaLymphocytes


features <- SelectIntegrationFeatures(object.list = seurat.list)

lymph.anchors <- FindIntegrationAnchors(object.list = seurat.list, anchor.features = features)

lymphocytesCombined <- IntegrateData(anchorset = lymph.anchors)


DefaultAssay(lymphocytesCombined) <- "integrated"

# Run the standard workflow for visualization and clustering
lymphocytesCombined <- ScaleData(lymphocytesCombined, verbose = FALSE)
lymphocytesCombined <- RunPCA(lymphocytesCombined, npcs = 30, verbose = FALSE)
lymphocytesCombined <- RunUMAP(lymphocytesCombined, reduction = "pca", dims = 1:30)
lymphocytesCombined <- FindNeighbors(lymphocytesCombined, reduction = "pca", dims = 1:30)
lymphocytesCombined <- FindClusters(lymphocytesCombined, resolution = 0.1)

 data10xMultimodalReference <-
   SeuratDisk::LoadH5Seurat("data/pbmc_multimodal.h5seurat")
 
 lymphocytesCombined <- SCTransform(lymphocytesCombined, verbose = FALSE)
 
 anchors <- FindTransferAnchors(
  reference = data10xMultimodalReference,
  query = lymphocytesCombined,
  normalization.method = "SCT",
  reference.reduction = "spca",
  dims = 1:50
)
 
lymphocytesCombined<-  MapQuery(
  anchorset = anchors,
  query = lymphocytesCombined,
  reference = data10xMultimodalReference,
  refdata = list(
    celltype.l1 = "celltype.l1",
    celltype.l2 = "celltype.l2",
    predicted_ADT = "ADT"
  ),
  reference.reduction = "spca", 
  reduction.model = "wnn.umap"
)
```


```{r origins}
        clusterCounts <- lymphocytesCombined@meta.data %>%
  dplyr::group_by(Cluster.2, originalStudy) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(Cluster.2) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(Cluster.2)

originalStudyBarplot.2 <- clusterCounts %>%
  ggplot(aes(x = Cluster.2,
             y = pctCells,
             fill = originalStudy)) +
  geom_col() +
  scale_fill_manual(values = c("colorado" = "red", "seattle" = "darkcyan")) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "",
       y = "% of cells",
       fill = "Origin") +
  theme(axis.text.x=element_text(angle=60, hjust=1))

clusterCounts <- lymphocytesCombined@meta.data %>%
  dplyr::group_by(Cluster.1, originalStudy) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(Cluster.1) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(Cluster.1)

originalStudyBarplot.1 <- clusterCounts %>%
  ggplot(aes(x = Cluster.1,
             y = pctCells,
             fill = originalStudy)) +
  geom_col() +
  scale_fill_manual(values = c("colorado" = "red", "seattle" = "darkcyan")) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "",
       y = "% of cells",
       fill = "Origin") +
  theme(axis.text.x=element_text(angle=60, hjust=1))

clusterCounts <- lymphocytesCombined@meta.data %>%
  dplyr::group_by(seurat_clusters, originalStudy) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(seurat_clusters) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(seurat_clusters)

originalStudyBarplot.05 <- clusterCounts %>%
  ggplot(aes(x = seurat_clusters,
             y = pctCells,
             fill = originalStudy)) +
  geom_col() +
  scale_fill_manual(values = c("colorado" = "red", "seattle" = "darkcyan")) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "",
       y = "% of cells",
       fill = "Origin") +
  theme(axis.text.x=element_text(angle=60, hjust=1))
```

```{r plots}
pdf("CellTypes.pdf", height = 8, width = 10)
DimPlot(lymphocytesCombined, group.by = "predicted.celltype.l1")
dev.off()

pdf("Cluster2.pdf", height = 8, width = 10)
DimPlot(lymphocytesCombined, group.by = "Cluster.2")
dev.off()

pdf("Cluster1.pdf", height = 8, width = 10)
DimPlot(lymphocytesCombined, group.by = "Cluster.1")
dev.off()

pdf("Cluster05.pdf", height = 8, width = 10)
DimPlot(lymphocytesCombined, group.by = "seurat_clusters")
dev.off()

pdf("boxplot2.pdf", height = 8, width = 10)
originalStudyBardplot.2
dev.off()

pdf("boxplot2.pdf", height = 8, width = 10)
originalStudyBardplot.2
dev.off()

pdf("barplot1.pdf", height = 8, width = 10)
originalStudyBarplot.1
dev.off()

pdf("barplot05.pdf", height = 8, width = 10)
originalStudyBarplot.05
dev.off()

pdf("time.pdf", height = 8, width = 10)
DimPlot(lymphocytesCombined, group.by = "time")
dev.off()
```

```{r cluster_markers}

Idents(lymphocytesCombined) <- "Cluster.2"

clusterMarkers <- FindAllMarkers(lymphocytesCombined)

topMarkersHigh <- clusterMarkers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n=10) %>%
  dplyr::slice_max(avg_log2FC, n = 10) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 
  
topMarkersLow <-  clusterMarkers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC < 0) %>%
  dplyr::slice_min(p_val_adj, n=10) %>%
  dplyr::slice_max(avg_log2FC, n = 10) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 
  
  topMarkersLong <- rbind(topMarkersLow, topMarkersHigh) %>% group_by(cluster)

topMarkers <- topMarkersLong %>%
  pivot_wider(id_cols = "fc_order",
              names_from = "cluster",
              values_from = "gene")

write.csv(topMarkersLong,
          "~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/cluster2Markers.csv",
          row.names = F,
          quote = F)

#Get top all top markers in wide format

topMarkersWide <- clusterMarkers %>%
  dplyr::filter(!str_detect(gene, "Trav|Trbv")) %>%
  dplyr::mutate(cluster_id = paste0("Cluster ",cluster),
                cluster_id = factor(cluster_id, levels = mixedsort(unique(cluster_id)))) %>%
  dplyr::group_by(cluster_id) %>%
  dplyr::mutate(row_order = order(p_val_adj)) %>%
  pivot_wider(id_cols = row_order,
              names_from = "cluster_id",
              names_glue = "{cluster_id}_{.value}",
              values_from = c( "cluster_id",
                               "gene",
                               "avg_log2FC",
                               "p_val_adj"))
  
clusters <- unique(clusterMarkers$cluster) %>% mixedsort()
clusterLabs <- paste0("Cluster ",rep(clusters, each = 4))

columnLabs <- c("cluster_id",
                 "gene",
                 "avg_log2FC",
                "p_val_adj")

nClust <- length(unique(seuratIntegrated$seurat_clusters))
columnLabs <- rep(columnLabs, nClust)

columnLabs <- paste0(clusterLabs, "_", columnLabs)

topMarkersWide <- topMarkersWide %>%
  dplyr::select(columnLabs)

write.csv(topMarkersWide,
          file.path(tableDir, "SeuratIntegrated_TopMarkersByCluster.csv"),
          row.names = F,
          quote = F)

```

```{r integrated_hmap}

#Average top marker expr by cluster.

topMarkersHmap <- clusterMarkers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 4) %>%
  dplyr::slice_max(avg_log2FC, n = 4) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

selectedFeatures <- unique(topMarkersHmap$gene)

lymphocytesCombined <- NormalizeData(lymphocytesCombined, assay = "RNA", normalization.method = "CLR")

clusterAverages <- AverageExpression(lymphocytesCombined, 
                                    assays = "RNA",
                                    features = selectedFeatures,
                                    group.by = "seurat_clusters",
                                    return.seurat = T) 

avgExprMat <- clusterAverages[["RNA"]]@data %>% as.matrix()

scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- rownames(clusterAverages@meta.data)

columnAnno <- rowAnnotation(Cluster = clusterAnno,
                              col = list(Cluster = clusterColors),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/Heatmap_Integrated_TopClusterMarkers.pdf",
    height = 4,
    width = 10)

Heatmap(t(scaledExprMat),
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        left_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())

```

```{r mapping seattle onto colorado umap}

coloradoLymphocytes@meta.data$colCluster <- coloradoLymphocytes@meta.data$seurat_clusters

coloradoLymphocytes <- FindClusters(coloradoLymphocytes, resolution = 0.5)

pdf("coloradoClustersNewReUmap.pdf", width = 8, height = 8)
  ggplot(
    data = coloradoLymphocytes@meta.data,
    mapping =
      aes(x = coloradoLymphocytes@reductions$umap2@cell.embeddings[,1],
          y = coloradoLymphocytes@reductions$umap2@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = "seurat_clusters"), size = .75, alpha = 0.1)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(9)) +
 # facet_wrap(~studyGroup, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))
  dev.off()

  
  anchors <- FindTransferAnchors(reference = coloradoLymphocytes, query = seaLymphocytes, dims = 1:30, reference.reduction = "pca", k.filter = NA)
  
referencedMappedSea <-MapQuery(anchorset = anchors, reference = coloradoLymphocytes, query = seaLymphocytes,
         refdata = "seurat_clusters", reference.reduction = "pca", reduction.model = "umap2")

pdf("referenceProjectionSeattle.pdf", width = 8, height = 8)
  ggplot(
    data = referencedMappedSea@meta.data,
    mapping =
      aes(x = referencedMappedSea@reductions$ref.umap@cell.embeddings[,1],
          y = referencedMappedSea@reductions$ref.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = "predicted.id"), size = .75, alpha = 0.1)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(9)) +
 # facet_wrap(~studyGroup, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1))) + labs(title = "Seattle Samples on Denver Projection")
  dev.off()
  
  pdf("Cluster2.pdf", height = 8, width =8)
    ggplot(
    data = lymphocytesCombined@meta.data,
    mapping =
      aes(x = lymphocytesCombined@reductions$umap@cell.embeddings[,1],
          y = lymphocytesCombined@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = "Cluster.2"), size = .75, alpha = 0.1)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(10)) +
 # facet_wrap(~studyGroup, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1)))
    dev.off()
    
counts <- lymphocytesCombined@assays$integrated@data

noTRCcounts <- counts[-grep("^TR[A-B]", rownames(counts)),]

lymphocytesCombined[["noTCR"]] <- CreateAssayObject(noTRCcounts)

DefaultAssay(lymphocytesCombined) <- 'noTCR'

lymphocytesCombined <- ScaleData(lymphocytesCombined, verbose = FALSE)
lymphocytesCombined <- RunPCA(lymphocytesCombined, npcs = 30, verbose = FALSE, reduction.name = "noTCRpca")
lymphocytesCombined <- RunUMAP(lymphocytesCombined, reduction = "noTCRpca", dims = 1:30, reduction.name = "noTCRumap")
lymphocytesCombined <- FindNeighbors(lymphocytesCombined, reduction = "noTCRpca", dims = 1:30)
lymphocytesCombined <- FindClusters(lymphocytesCombined, resolution = 0.1)

  pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/clustersIntegratedUMAP.pdf", height = 10, width = 12)        

  ggplot(
    data = lymphocytesCombined@meta.data,
    mapping =
      aes(x = lymphocytesCombined@reductions$noTCRumap@cell.embeddings[,1],
          y = lymphocytesCombined@reductions$noTCRumap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = "seurat_clusters"), size = .75, alpha = 0.5)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(8)) +
 # facet_wrap(~studyGroup, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))
  dev.off()

  pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/originalStudyIntegratedUMAP.pdf", height = 10, width = 12)        
  
    ggplot(
    data = lymphocytesCombined@meta.data,
    mapping =
      aes(x = lymphocytesCombined@reductions$noTCRumap@cell.embeddings[,1],
          y = lymphocytesCombined@reductions$noTCRumap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = "originalStudy"), size = .75, alpha = 0.5)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Cluster", 
    values = big_colorblind_pal(8)) +
 # facet_wrap(~studyGroup, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))
    dev.off()
    
    
pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/cellTypesIntegratedUMAP.pdf", height = 10, width = 12)        
  
    ggplot(
    data = lymphocytesCombined@meta.data,
    mapping =
      aes(x = lymphocytesCombined@reductions$noTCRumap@cell.embeddings[,1],
          y = lymphocytesCombined@reductions$noTCRumap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = "lymphocytesCombined@meta.data$predicted.celltype.l1"), size = .75, alpha = 0.5)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_manual(
    "Study", 
    values = big_colorblind_pal(8)) +
 # facet_wrap(~studyGroup, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))
dev.off()
    
    DefaultAssay(lymphocytesCombined) <- "RNA"
    
      lymphocytesCombined[["percent_mt"]] <- PercentageFeatureSet(lymphocytesCombined, pattern = "^MT-")

    
        ggplot(
    data = lymphocytesCombined@meta.data,
    mapping =
      aes(x = lymphocytesCombined@reductions$noTCRumap@cell.embeddings[,1],
          y = lymphocytesCombined@reductions$noTCRumap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = "lymphocytesCombined@meta.data$percent_mt"), size = .75, alpha = 0.5)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_continuous(
    "Percent MT") +
 # facet_wrap(~studyGroup, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))
  
        

        
        clusterCounts <- lymphocytesCombined@meta.data %>%
  dplyr::group_by(seurat_clusters, originalStudy) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(seurat_clusters) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(seurat_clusters)

pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/clusterbyOriginalStudy.pdf", height = 10, width = 12)        
        clusterCounts %>%
  ggplot(aes(x = seurat_clusters,
             y = pctCells,
             fill = originalStudy)) +
  geom_col() +
  scale_fill_manual(values = c("colorado" = "darkred", "seattle" = "darkcyan")) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "",
       y = "% of cells",
       fill = "Origin") +
  theme(axis.text.x=element_text(angle=60, hjust=1))
dev.off()

pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/OriginalStudyByCluster.pdf", height = 10, width = 12)
                clusterCounts %>%
  ggplot(aes(x = originalStudy,
             y = scaledNCells,
             fill = seurat_clusters)) +
  geom_col() +
  scale_fill_manual(values = big_colorblind_pal(8)) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "",
       y = "% of cells",
       fill = "cluster assignment") +
  theme(axis.text.x=element_text(angle=60, hjust=1))
dev.off()

DefaultAssay(lymphocytesCombined) <- "integrated"                
Idents(lymphocytesCombined) <- "time"
timeMarkers <- FindMarkers(lymphocytesCombined, ident.1 = "0", ident.2 = "24")

timeFeatures <- timeMarkers %>% filter(p_val_adj < .05) %>% head(30) %>% rownames()

lymphocytesCombined <- RunPCA(lymphocytesCombined, npcs = 20, verbose = FALSE, reduction.name = "timepca", features = timeFeatures)
lymphocytesCombined <- RunUMAP(lymphocytesCombined, reduction = "timepca", dims = 1:20, reduction.name = "timeumap")
lymphocytesCombined <- FindNeighbors(lymphocytesCombined, reduction = "timepca", dims = 1:20)
lymphocytesCombined <- FindClusters(lymphocytesCombined, resolution = 0.1)

pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/timeProjectionUMAP.pdf", height = 10, width = 20)
        ggplot(
    data = lymphocytesCombined@meta.data,
    mapping =
      aes(x = lymphocytesCombined@reductions$timeumap@cell.embeddings[,1],
          y = lymphocytesCombined@reductions$timeumap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = "lymphocytesCombined@meta.data$time"), size = .75, alpha = 0.5)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
    scale_color_manual(
    "timePoint", 
    values = big_colorblind_pal(10))  +
  facet_wrap(~time, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))
dev.off()

        
```

```{r}
topMarkersHmap <- timeMarkers %>%
  dplyr::slice_min(p_val_adj, n = 20) %>%
  dplyr::slice_max(avg_log2FC, n = 20) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

selectedFeatures <- unique(rownames(topMarkersHmap))

lymphocytesCombined <- NormalizeData(lymphocytesCombined, assay = "RNA", normalization.method = "CLR")

clusterAverages <- AverageExpression(lymphocytesCombined, 
                                    assays = "RNA",
                                    features = selectedFeatures,
                                    group.by = "time",
                                    return.seurat = T) 

avgExprMat <- clusterAverages[["RNA"]]@data %>% as.matrix()

scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- rownames(clusterAverages@meta.data)

columnAnno <- rowAnnotation(time = clusterAnno,
                              col = list(time = timeColors),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/Heatmap_Integrated_TopClusterMarkersTIME.pdf",
    height = 4,
    width = 10)

Heatmap(t(scaledExprMat),
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        left_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())

```


```{r stats}
sheet1 <- lymphocytesCombined@meta.data %>%
  dplyr::group_by(seurat_clusters, originalStudy) %>%
  dplyr::summarize(nCells = n())

sheet2 <- lymphocytesCombined@meta.data %>%
  dplyr::group_by(seurat_clusters, originalStudy) %>%
  dplyr::summarize(nCells = n()) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(seurat_clusters) %>%
  dplyr::mutate(nCellsTot = sum(nCells),
                pctCells = nCells/nCellsTot) %>%
  dplyr::arrange(seurat_clusters)

write.xlsx(as.data.frame(sheet1), file="~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/cellNumbers.xlsx", sheetName="Cells divided Original Study", row.names=TRUE)
write.xlsx(as.data.frame(sheet2), file="~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/cellNumbers.xlsx", sheetName="Cells divided CLUSTER", append=TRUE, row.names=TRUE)

```

```{r}
        ggplot(
    data = lymphocytesCombined@meta.data,
    mapping =
      aes(x = lymphocytesCombined@reductions$timeumap@cell.embeddings[,1],
          y = lymphocytesCombined@reductions$timeumap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = lymphocytesCombined@assays$RNA["BCL6"]), size = .75, alpha = 0.5)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
    scale_color_manual(
    "timePoint", 
    values = big_colorblind_pal(10))  +
  facet_wrap(~time, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))

features <- c("CXCR6","TCF7","GATA3","CD294","PRDM1","TBX21","IL18R1","IL1RL1","CX3CR1","BCL6")

pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/selectedFeaturesCombined.pdf", height = 10, width = 12)
FeaturePlot(lymphocytesCombined, features, pt.size = 1, reduction = "noTCRumap")
dev.off()

pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/selectedFeaturesSeattle.pdf", height = 10, width = 12)
FeaturePlot(seaLymphocytes, features, pt.size = 1, reduction = "ref.umap")
dev.off()
```


```{r timescale heatmaps}

Idents(coloradoLymphocytes) <- "time"

coloradoT0Markers <- data.frame()
for(i in levels(factor(coloradoLymphocytes@meta.data$time))[-1]){
temp.tmp  <- FindMarkers(coloradoLymphocytes, ident.1 = "0", ident.2 = i, logfc.threshold = .20, assay = "RNA")
#rownames(temp.tmp) <- gsub("\\.[1-9]", "", rownames(temp.tmp))
temp.tmp$timepoint <- eval(parse(text=i))
temp.tmp <- temp.tmp[!rownames(temp.tmp) %in% rownames(coloradoT0Markers),]
coloradoT0Markers <- rbind(coloradoT0Markers, filter(temp.tmp, avg_log2FC > 0) %>% slice(1:10))
}
coloradoT0Markers$gene <- rownames(coloradoT0Markers)

topMarkersHmap <- coloradoT0Markers %>%
  dplyr::group_by(timepoint) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

selectedFeatures <- unique(topMarkersHmap$gene)

coloradoLymphocytes <- NormalizeData(coloradoLymphocytes, assay = "RNA", normalization.method = "LogNormalize")

timeAverages <- AverageExpression(coloradoLymphocytes, 
                                    assays = "RNA",
                                    features = selectedFeatures,
                                    group.by = "time",
                                    return.seurat = T) 

avgExprMat <- timeAverages[["RNA"]]@data %>% as.matrix()



scaledExprMat <- t(scale(t(avgExprMat[,-1])))


clusterAnno <- rownames(timeAverages@meta.data)[-1]

columnAnno <- columnAnnotation(Timepoint = clusterAnno,
                              col = list(Timepoint = timeColors[-1]),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/Heatmap_Integrated_TopClusterMarkers.pdf",
    height = 4,
    width = 10)

Heatmap(scaledExprMat,
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        top_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        show_column_names = FALSE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())



```


```{r }

#Psuedo Bulking of the Denver samples and then adding sample as a confounding variable

# cds_coords is a dataframe I use to keep metadata about the cells
# cds_seurat_CD8 Is the seurat object
# Here i'm doing pseudobulk DE to find cluster markers. Each pseudobulk sample is the collection of cells in a particular cluster from a particular sample in this case.
groups <- unique(cds_coords$timeBySubj)
barcodesByGroup <- lapply(groups, function(g) rownames(cds_coords)[ cds_coords$timeBySubj == g]  )
    names(barcodesByGroup) <- groups
    #  barcodesByDonorVisit <- lapply(donorVisits, function(dv) rownames(cds_coords)[cds_coords$donorVisit == dv & cds_coords$Treg]  )
    barcodesByGroup <- barcodesByGroup[ sapply(barcodesByGroup, length) >= 5]
    barcodes <- do.call(c,barcodesByGroup)
    bulk <- sapply( barcodesByGroup, function(b) rowSums(coloradoLymphocytes@assays[["RNA"]][,b])  )
    gene_filter <- function(counts_in, per_cutoff){
      #Keep genes with cpm of at least one in at least per_cutoff fraction of libraries
      #CPM normalize
      counts_cpm_norm <- as.data.frame(t(t(counts_in*10^6)/colSums(counts_in)))
      #Filter out lowly expressed genes
      keepRows <- rowSums((counts_cpm_norm) >= 1) >= per_cutoff*ncol(counts_cpm_norm)
      counts_filtered <- counts_in[keepRows,]
      return(counts_filtered)
    }
    bulk_pc_filtered <- gene_filter(bulk, 0.20)
    normalize_counts <- function(counts_in, method){
      #normalize using tmm or deconvolution
      #tmm is good for bulk RNAseq
      #deconvolution is best for large datasets of single cell RNAseq
      #deconvolution is NOT recommended for smaller datasets (less than a few hundred cells)
      if(method == "decon"){
        #Normalize using the deconvolution algorithm
        decon_norm_factors <- calculateSumFactors(as.matrix(counts_in))
        counts_norm <- as.data.frame(t(t(counts_in)/decon_norm_factors))
      }
      if(method == "tmm"){
        #Normalize using the TMM algorithm
        dge <- DGEList(counts_in)
        dge <- calcNormFactors(dge)
        counts_norm <- edgeR::cpm(dge, normalized.lib.sizes=TRUE)
      }
      return(counts_norm)
    }
    print("B")
    bulk_pc_norm <- normalize_counts(bulk_pc_filtered, "tmm")
    df <- data.frame( t(sapply( colnames(bulk_pc_norm), function(s) strsplit(s,"_")[[1]])) )
    colnames(df) <- c("time","sub")
    design_mat <-  model.matrix(~time+sub, data=df)
    vwts <- voomWithQualityWeights(bulk_pc_norm, design= design_mat, plot=T, span=0.1)
    vfit<- lmFit(vwts, design = design_mat)
    time <- colnames(design_mat)[ grepl("time",colnames(design_mat))]
    cont.matrix <- matrix(0,ncol(design_mat),length(time))
    rownames(cont.matrix) <- colnames(design_mat)
    for( i in 1:length(time) ){
      t <- time[i]
      cont.matrix[time,i] <- 0
      cont.matrix[time[i],i] <- 1
    }
    colnames(cont.matrix) <- time
    vfit_c <- contrasts.fit(vfit, cont.matrix)
    vfit_eb_c <- eBayes(vfit_c)
    #clustcomps <- lapply(1:ncol(cont.matrix), function(i) topTable (vfit_eb_c, coef = i , number=Inf, sort.by="P"))

avgExprMat <- topTags[1:5] %>% as.matrix()



scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- colnames(topTags[1:5])

columnAnno <- columnAnnotation(Timepoint = clusterAnno,
                              col = list(Timepoint = timeColors[-1]),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/psuedoBulksubjectbyTimeModelHeatmap.pdf",
    height = 4,
    width = 10)

Heatmap(scaledExprMat,
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = FALSE,
        top_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = FALSE,
        show_column_names = TRUE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())    


avgExprMat <- highestTopTags[1:5] %>% as.matrix()



scaledExprMat <- t(scale(t(avgExprMat)))


clusterAnno <- colnames(topTags[1:5])

columnAnno <- columnAnnotation(Timepoint = clusterAnno,
                              col = list(Timepoint = timeColors[-1]),
                              simple_anno_size = unit(0.25, "cm"))

# geneNameAnno <- rowAnnotation(link = anno_mark(at = match(selectedGenes, rownames(scaledExprMat)), 
#         labels = selectedGenes,
#         labels_gp = gpar(fontsize = 6),
#         padding = unit(1, "mm")))  

pdf("~/Library/CloudStorage/Box-Box/DenSea T cell Collaboration/analyses/psuedoBulksubjectbyTimeModelHeatmap.pdf",
    height = 4,
    width = 10)

Heatmap(scaledExprMat,
        name = "Gene z-score",
        clustering_distance_columns = "manhattan",
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        column_dend_height = unit(0.5, "cm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        top_annotation = columnAnno,
        #right_annotation = geneNameAnno,
        show_row_names = TRUE,
        show_column_names = TRUE,
        column_names_rot = 90,
        column_names_gp = gpar(fontsize = 8),
        row_names_gp = gpar(fontsize = 10))

invisible(dev.off())    
```    


```{r}
VolcSubset <- topTags[1:15,]

# produce the volcano plot with ggplot

  Actives <- ggplot(data = topTags,  # set the data as the table contained within the object we loaded
                    aes(x=(time24- AveExpr), y=-log(adj.P.Val), # set the x axis and y axis to logFC and FDR respectively
                        color = (time24- AveExpr)>0)) + # set the color condition to be if the FC is above or below 0 
    geom_point(alpha=0.7, size=1, shape = 19) + # define the transparency, size, and shape of the dots 
    #ylim(0,15) + #define the limits of the Y axis to range from 0 to 15 
    #xlim(-10,10) + # define the limits of the X axis to range from -10 to 10
    geom_text_repel(data=VolcSubset, # set up the labeling of the top 15 genes I subsetted earlier
                  label = rownames(VolcSubset), #use the rownames (which are the gene names) as the label
                  max.overlaps = Inf, min.segment.length = unit(0, 'lines'), # set length of label lines
                  size = 3, # set size of the label
                  col = "black") + # set color of label
    scale_color_manual(values = c("darkcyan", "red"), # set color of dots for below and above 0
                  name = "", 
                  label = NULL) +
    labs(x = "Delta Timepoint 24 Exp from Time 0", # labels of the axes and the title
                  y = "-log(adj P Val)",
                  title = "Timepoint 24 vs TimePoint 0") +
    geom_hline(yintercept= -log10(.01), #set up a horizontal line to indicate a FDR of .01
                  color="black",
                  linetype="dotted",
                  size=1.0) +
    theme(plot.title = element_text(hjust = .5)) + # adjust the plot title's horizontal alignment slightly
    NoLegend() # remove the legend to clear clutter
  
  # display the chart!
  print(Actives)

```

```{r setting up Project #1 to merge}

DefaultAssay(BALSeurat) <- "FB"
specifictyScore <- list()

for(i in rownames(BALSeurat)){

specifictyScore[[i]]<- rowMeans(cbind((1 - pbeta(.925, (as.numeric(BALSeurat@assays$FB@counts[i,]) +sPrior), (as.numeric(BALSeurat@assays$FB@counts["Mouse IgG1",]) + nPrior))) * 100, (1 - pbeta(.925, (as.numeric(BALSeurat@assays$FB@counts[i,]) +sPrior), (as.numeric(BALSeurat@assays$FB@counts["Mouse IgG2a",]) + nPrior))) * 100, (1 - pbeta(.925, (as.numeric(BALSeurat@assays$FB@counts[i,]) +sPrior), (as.numeric(BALSeurat@assays$FB@counts["Mouse IgG2b",]) + nPrior))) * 100))

}

for(i in names(specifictyScore)){
BALSeurat@meta.data[paste0(i, "_score")] <- specifictyScore[i]
}

pdf("~/Library/CloudStorage/Box-Box/P458-2_SLICC_Mikacenic/plots/CD71noiseReduction.pdf", height = 8, width = 10)
    ggplot(
    data = BALSeurat@meta.data,
    mapping =
      aes(x = BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,1],
          y = BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = "BALSeurat@meta.data$CD71_score"), size = .75, alpha = 0.5)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_continuous(
    "Specificity") +
 # facet_wrap(~studyGroup, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))
    dev.off()

    pdf("~/Library/CloudStorage/Box-Box/P458-2_SLICC_Mikacenic/plots/CD163noiseReduction.pdf", height = 8, width = 10)    
        ggplot(
    data = BALSeurat@meta.data,
    mapping =
      aes(x = BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,1],
          y = BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = "BALSeurat@meta.data$CD163.1_score"), size = .75, alpha = 0.5)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_continuous(
    "Specificity") +
 # facet_wrap(~studyGroup, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))
        dev.off()
        
grobs <- list()
for(i in colnames(BALSeurat@meta.data)[grep("_score", colnames(BALSeurat@meta.data))]){
y <- i  
x <- gsub("_score", "", i)
grobs[[x]] <- ggplot(data = BALSeurat@meta.data, aes(x=Cluster.3, y=eval(parse(text=y)), fill=Cluster.3)) + geom_boxplot(outlier.shape = NA) + scale_fill_manual("Cluster", values = big_colorblind_pal(9)) + labs(x = "Cluster Identity", y = paste0(x, " Specificity")) + NoLegend() + theme_bw()
}

grobs <- list()
for(i in colnames(BALSeurat@meta.data)[grep("_score", colnames(BALSeurat@meta.data))]){
y <- i  
x <- gsub("_score", "", i)
grobs[[x]] <- ggplot(
    data = BALSeurat@meta.data,
    mapping =
      aes(x = BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,1],
          y = BALSeurat@reductions$MonoRNA.umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = i), size = .75, alpha = 0.5)) +
  labs(x = "UMAP 1", y = "UMAP 2", title = paste0(i, " Score")) +
  scale_color_continuous(
    "Specificity") +
 # facet_wrap(~studyGroup, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))
}

pdf("~/Library/CloudStorage/Box-Box/P458-2_SLICC_Mikacenic/plots/noiseReductionTotal.pdf", height = 12, width = 8)
grid.arrange(grobs$CD71, grobs$CD163, grobs$CD123, grobs$CD48, grobs$CD86, grobs$CD44, grobs$CD1c, grobs$LOX_1, grobs$CD35, grobs$CD206MMR, grobs$CD14)
dev.off()
```

```{r}

MTDataframe <- data.frame()
for(i in seuratQC){
  entry <- cbind(unique(i@meta.data$sample), count(i@meta.data$percent_mt >= 0), count(i@meta.data$percent_mt > 12.5), count(i@meta.data$percent_mt > 12.5)/count(i@meta.data$percent_mt > 0))
  MTDataframe <- rbind(MTDataframe, entry)
}

seuratMerged@meta.data$mitoStatus <- "Low"
seuratMerged@meta.data$mitoStatus[seuratMerged@meta.data$percent_mt > 12.5] <- "High"

Idents(seuratMerged) <- "mitoStatus"

FindMarkers(seuratMerged, ident.1 = "High")


write.xlsx(x = HighMitoCells[-grep("^MT",rownames(HighMitoCells)),] %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n = 20) %>%
  dplyr::slice_max(avg_log2FC, n = 20) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)), file = "~/Library/CloudStorage/Box-Box/P458-2_SLICC_Mikacenic/tables/highMitoMarkers30.xlsx", sheetName = "Markers")


  ggplot(
    data = seuratMerged@meta.data,
    mapping =
      aes(x = seuratMerged@reductions$umap@cell.embeddings[,1],
          y = seuratMerged@reductions$umap@cell.embeddings[,2])) +
  ggrastr::rasterise(
    geom_point(mapping = aes_string(color = "seuratMerged@meta.data$percent_mt"), size = .75, alpha = 0.5)) +
  labs(x = "UMAP 1", y = "UMAP 2") +
  scale_color_continuous(
    "Mitochondrial Read %") +
 # facet_wrap(~studyGroup, nrow = 2) +
  guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))
  
  seuratQC <- NULL

for(i in 1:length(seuratObjects)){
  seurat <- seuratObjects[[i]]
    
  seuratQC[[i]] <- subset(seurat, 
                      subset = nFeature_RNA > nFeatureLow & nFeature_RNA < nFeatureHigh)

}
  
  nFeatureHistograms <- lapply(seuratQC, function(x) hist(x$nFeature_RNA, 200, xlim= c(0,5000), main = unique(x$sample), xlab = "nFeature RNA"))

  par(mfrow=c(4,3))
lapply(seuratQC, function(x) {hist(x$nFeature_RNA, 50, xlim= c(0,6000), main = unique(x$sample) , xlab = "nFeature RNA")
  abline(v = nFeatureHigh, col = "red")
  abline(v = nFeatureLow, col = "red")})

for(i in colnames(BALSeurat@meta.data)[grep("_score", colnames(BALSeurat@meta.data))]){
y <- i  
x <- gsub("_score", "", i)
grobs_Viol[[x]] <- ggplot(data = BALSeurat@meta.data, aes(x=Cluster.3, y=eval(parse(text=y)), fill=Cluster.3)) + geom_violin(outlier.shape = NA) + scale_fill_manual("Cluster", values = big_colorblind_pal(9)) + labs(x = "Cluster Identity", y = paste0(x, " Specificity")) + NoLegend() + theme_bw()
}


for(i in rownames(BALSeurat)){
specifictyScore[[i]]<- rowMeans(cbind((1 - pbeta(.925, (as.numeric(BALSeurat@assays$FB@counts[i,]) +sPrior), (as.numeric(BALSeurat@assays$FB@counts["Mouse IgG1",]) + nPrior))) * 100, (1 - pbeta(.925, (as.numeric(BALSeurat@assays$FB@counts[i,]) +sPrior), (as.numeric(BALSeurat@assays$FB@counts["Mouse IgG2a",]) + nPrior))) * 100, (1 - pbeta(.925, (as.numeric(BALSeurat@assays$FB@counts[i,]) +sPrior), (as.numeric(BALSeurat@assays$FB@counts["Mouse IgG2b",]) + nPrior))) * 100))

}

for(i in names(specifictyScore)){
BALSeurat@meta.data[paste0(i, "_score")] <- specifictyScore[i]
}

FBcountsfordsb <- as.matrix(BALSeurat@assays$FB2@counts)


adt_norm <- ModelNegativeADTnorm(cell_protein_matrix = FBcountsfordsb,
                                denoise.counts = TRUE,
                                use.isotype.control = TRUE,
                                isotype.control.name.vec = rownames(FBcountsfordsb)[32:34]
                                )

BALSeurat[["normADT"]] <- CreateAssayObject(counts = adt_norm)

DefaultAssay(BALSeurat) <- "normADT"

  BALSeurat <- NormalizeData(object = BALSeurat, assay = "normADT", verbose = FALSE, normalization.method = "LogNormalize")
  BALSeurat <- ScaleData(BALSeurat, features = rownames(BALSeurat@assays$normADT), verbose = FALSE)
  
ggplot(as.data.frame(t(BALSeurat@assays$FB2@scale.data[BALSeurat@meta.data$Cluster.3 == 3])), aes(x= CD71, y = CD274 (B7-H1, PD-L1)))


p1<-FeatureScatter(subset(BALSeurat, idents = 3), feature1 = "CD169", feature2 = "CD71", cols = big_colorblind_pal(9)[4], plot.cor = FALSE) + NoLegend() + stat_cor(method = "pearson", label.x = 0, label.y = 7.5)

p2<-FeatureScatter(subset(BALSeurat, idents = 3), feature1 = "fb3_PDL1", feature2 = "CD71", cols = big_colorblind_pal(9)[4], plot.cor = FALSE) + NoLegend() + stat_cor(method = "pearson", label.x = 0, label.y = 7.5) + labs(x = "PD-L1")

p3<-FeatureScatter(subset(BALSeurat, idents = 3), feature1 = "fb3_PDL1", feature2 = "CD169", cols = big_colorblind_pal(9)[4], plot.cor = FALSE) + NoLegend() + stat_cor(method = "pearson", label.x = 0, label.y = 7.5) + labs(x = "PD-L1")

pdf("~/Library/CloudStorage/Box-Box/P458-2_SLICC_Mikacenic/plots/ScatterplotsWithCorrelation.pdf", height = 4, width = 10)
p1 + p2 + p3
dev.off()
```
